// rclnodejs/generated/safety_cage_interfaces/safety_cage_interfaces__msg__VehicleIdentity.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const VehicleIdentityRefStruct = StructType({
  serial_number: primitiveTypes.string,
  ip_address: primitiveTypes.string,
});
const VehicleIdentityRefArray = ArrayType(VehicleIdentityRefStruct);
const VehicleIdentityRefStructArray = StructType({
  data: VehicleIdentityRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VehicleIdentityWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VehicleIdentityRefStruct(other._refObject.toObject());
      this._serial_numberIntialized = true;
      this._wrapperFields.serial_number =  new StringWrapper(other._wrapperFields.serial_number);
      this._ip_addressIntialized = true;
      this._wrapperFields.ip_address =  new StringWrapper(other._wrapperFields.ip_address);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VehicleIdentityRefStruct();
    this._serial_numberIntialized = false;
    this._wrapperFields.serial_number = new StringWrapper();
    this._ip_addressIntialized = false;
    this._wrapperFields.ip_address = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new VehicleIdentityWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VehicleIdentityArrayWrapper;
  }
  static get ArrayType() {
    return VehicleIdentityArrayWrapper;
  }
  static get refObjectArrayType() {
    return VehicleIdentityRefStructArray
  }
  static get refObjectType() {
    return VehicleIdentityRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._serial_numberIntialized) {
        throw new TypeError('Invalid argument: serial_number in VehicleIdentity');
      }
      if (!this._ip_addressIntialized) {
        throw new TypeError('Invalid argument: ip_address in VehicleIdentity');
      }
    }
    if (own) {
      this._wrapperFields.serial_number.freeze(own, checkConsistency);
    }
    this._refObject.serial_number = this._wrapperFields.serial_number.refObject;
    if (own) {
      this._wrapperFields.ip_address.freeze(own, checkConsistency);
    }
    this._refObject.ip_address = this._wrapperFields.ip_address.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._serial_numberIntialized = true;
    this._wrapperFields.serial_number.data = refObject.serial_number.data;
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.data = refObject.ip_address.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.serial_number);
    StringWrapper.freeStruct(refObject.ip_address);
  }
  static destoryRawROS(msg) {
    VehicleIdentityWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'safety_cage_interfaces', subFolder: 'msg', interfaceName: 'VehicleIdentity'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get serial_number() {
    if (!this._serial_numberIntialized) {
      return undefined;
    }
    return this._wrapperFields.serial_number.data;
  }
  set serial_number(value) {
    this._serial_numberIntialized = true;
    this._wrapperFields.serial_number.data = value;
  }
  get ip_address() {
    if (!this._ip_addressIntialized) {
      return undefined;
    }
    return this._wrapperFields.ip_address.data;
  }
  set ip_address(value) {
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new VehicleIdentityRefStruct(refObject.toObject());
    this._serial_numberIntialized = true;
    this._wrapperFields.serial_number.copyRefObject(this._refObject.serial_number);
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.copyRefObject(this._refObject.ip_address);
  }
  copy(other) {
    this._refObject = new VehicleIdentityRefStruct(other._refObject.toObject());
    this._serial_numberIntialized = true;
    this._wrapperFields.serial_number.copy(other._wrapperFields.serial_number);
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.copy(other._wrapperFields.ip_address);
  }
  static get classType() {
    return VehicleIdentityWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"serial_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ip_address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"safety_cage_interfaces","type":"VehicleIdentity","stringUpperBound":null,"isPrimitiveType":false},"msgName":"VehicleIdentity"};
  }
  hasMember(name) {
    let memberNames = ["serial_number","ip_address"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VehicleIdentityArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VehicleIdentityWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VehicleIdentityWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleIdentityArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VehicleIdentityArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VehicleIdentityRefArray(size);
    this._refObject = new VehicleIdentityRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VehicleIdentityWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VehicleIdentityArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VehicleIdentityArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VehicleIdentityWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VehicleIdentityWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VehicleIdentityArrayWrapper;
  }
}
module.exports = VehicleIdentityWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "serial_number",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ip_address",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "safety_cage_interfaces",
    "type": "VehicleIdentity",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "VehicleIdentity"
}
*/
