// rclnodejs/generated/safety_cage_interfaces/safety_cage_interfaces__srv__UploadMesh_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UploadMesh_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  message: primitiveTypes.string,
  file_name: primitiveTypes.string,
  frame_id: primitiveTypes.string,
});
const UploadMesh_ResponseRefArray = ArrayType(UploadMesh_ResponseRefStruct);
const UploadMesh_ResponseRefStructArray = StructType({
  data: UploadMesh_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UploadMesh_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UploadMesh_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._messageIntialized = true;
      this._wrapperFields.message =  new StringWrapper(other._wrapperFields.message);
      this._file_nameIntialized = true;
      this._wrapperFields.file_name =  new StringWrapper(other._wrapperFields.file_name);
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UploadMesh_ResponseRefStruct();
    this._successIntialized = false;
    this._messageIntialized = false;
    this._wrapperFields.message = new StringWrapper();
    this._file_nameIntialized = false;
    this._wrapperFields.file_name = new StringWrapper();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new UploadMesh_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UploadMesh_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return UploadMesh_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return UploadMesh_ResponseRefStructArray
  }
  static get refObjectType() {
    return UploadMesh_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in UploadMesh_Response');
      }
      if (!this._messageIntialized) {
        throw new TypeError('Invalid argument: message in UploadMesh_Response');
      }
      if (!this._file_nameIntialized) {
        throw new TypeError('Invalid argument: file_name in UploadMesh_Response');
      }
      if (!this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in UploadMesh_Response');
      }
    }
    if (own) {
      this._wrapperFields.message.freeze(own, checkConsistency);
    }
    this._refObject.message = this._wrapperFields.message.refObject;
    if (own) {
      this._wrapperFields.file_name.freeze(own, checkConsistency);
    }
    this._refObject.file_name = this._wrapperFields.file_name.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._messageIntialized = true;
    this._wrapperFields.message.data = refObject.message.data;
    this._file_nameIntialized = true;
    this._wrapperFields.file_name.data = refObject.file_name.data;
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message);
    StringWrapper.freeStruct(refObject.file_name);
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destoryRawROS(msg) {
    UploadMesh_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'safety_cage_interfaces', subFolder: 'srv', interfaceName: 'UploadMesh_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get message() {
    if (!this._messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.message.data;
  }
  set message(value) {
    this._messageIntialized = true;
    this._wrapperFields.message.data = value;
  }
  get file_name() {
    if (!this._file_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.file_name.data;
  }
  set file_name(value) {
    this._file_nameIntialized = true;
    this._wrapperFields.file_name.data = value;
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UploadMesh_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copyRefObject(this._refObject.message);
    this._file_nameIntialized = true;
    this._wrapperFields.file_name.copyRefObject(this._refObject.file_name);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
  }
  copy(other) {
    this._refObject = new UploadMesh_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._messageIntialized = true;
    this._wrapperFields.message.copy(other._wrapperFields.message);
    this._file_nameIntialized = true;
    this._wrapperFields.file_name.copy(other._wrapperFields.file_name);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
  }
  static get classType() {
    return UploadMesh_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"file_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"safety_cage_interfaces","type":"UploadMesh_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UploadMesh_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","message","file_name","frame_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UploadMesh_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UploadMesh_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UploadMesh_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UploadMesh_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UploadMesh_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UploadMesh_ResponseRefArray(size);
    this._refObject = new UploadMesh_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UploadMesh_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UploadMesh_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UploadMesh_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UploadMesh_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UploadMesh_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UploadMesh_ResponseArrayWrapper;
  }
}
module.exports = UploadMesh_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "file_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "safety_cage_interfaces",
    "type": "UploadMesh_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UploadMesh_Response"
}
*/
