// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__CartesianTrajectoryPoint.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const AccelWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Accel.js');
const CartesianTrajectoryPointRefStruct = StructType({
  time_from_start: DurationWrapper.refObjectType,
  pose: PoseWrapper.refObjectType,
  twist: TwistWrapper.refObjectType,
  acceleration: AccelWrapper.refObjectType,
});
const CartesianTrajectoryPointRefArray = ArrayType(CartesianTrajectoryPointRefStruct);
const CartesianTrajectoryPointRefStructArray = StructType({
  data: CartesianTrajectoryPointRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CartesianTrajectoryPointWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CartesianTrajectoryPointRefStruct(other._refObject.toObject());
      this._wrapperFields.time_from_start =  new DurationWrapper(other._wrapperFields.time_from_start);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.twist =  new TwistWrapper(other._wrapperFields.twist);
      this._wrapperFields.acceleration =  new AccelWrapper(other._wrapperFields.acceleration);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CartesianTrajectoryPointRefStruct();
    this._wrapperFields.time_from_start = new DurationWrapper();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.twist = new TwistWrapper();
    this._wrapperFields.acceleration = new AccelWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CartesianTrajectoryPointWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CartesianTrajectoryPointArrayWrapper;
  }
  static get ArrayType() {
    return CartesianTrajectoryPointArrayWrapper;
  }
  static get refObjectArrayType() {
    return CartesianTrajectoryPointRefStructArray
  }
  static get refObjectType() {
    return CartesianTrajectoryPointRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.time_from_start.freeze(own, checkConsistency);
    this._refObject.time_from_start = this._wrapperFields.time_from_start.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.twist.freeze(own, checkConsistency);
    this._refObject.twist = this._wrapperFields.twist.refObject;
    this._wrapperFields.acceleration.freeze(own, checkConsistency);
    this._refObject.acceleration = this._wrapperFields.acceleration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.time_from_start.copyRefObject(refObject.time_from_start);
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.twist.copyRefObject(refObject.twist);
    this._wrapperFields.acceleration.copyRefObject(refObject.acceleration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    DurationWrapper.freeStruct(refObject.time_from_start);
    PoseWrapper.freeStruct(refObject.pose);
    TwistWrapper.freeStruct(refObject.twist);
    AccelWrapper.freeStruct(refObject.acceleration);
  }
  static destoryRawROS(msg) {
    CartesianTrajectoryPointWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'CartesianTrajectoryPoint'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time_from_start() {
    return this._wrapperFields.time_from_start;
  }
  set time_from_start(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.time_from_start.copy(value);
    } else {
      this._wrapperFields.time_from_start.copy(new DurationWrapper(value));
    }
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get twist() {
    return this._wrapperFields.twist;
  }
  set twist(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.twist.copy(value);
    } else {
      this._wrapperFields.twist.copy(new TwistWrapper(value));
    }
  }
  get acceleration() {
    return this._wrapperFields.acceleration;
  }
  set acceleration(value) {
    if (value instanceof AccelWrapper) {
      this._wrapperFields.acceleration.copy(value);
    } else {
      this._wrapperFields.acceleration.copy(new AccelWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CartesianTrajectoryPointRefStruct(refObject.toObject());
    this._wrapperFields.time_from_start.copyRefObject(this._refObject.time_from_start);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.twist.copyRefObject(this._refObject.twist);
    this._wrapperFields.acceleration.copyRefObject(this._refObject.acceleration);
  }
  copy(other) {
    this._refObject = new CartesianTrajectoryPointRefStruct(other._refObject.toObject());
    this._wrapperFields.time_from_start.copy(other._wrapperFields.time_from_start);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.twist.copy(other._wrapperFields.twist);
    this._wrapperFields.acceleration.copy(other._wrapperFields.acceleration);
  }
  static get classType() {
    return CartesianTrajectoryPointWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time_from_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twist","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acceleration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Accel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"CartesianTrajectoryPoint","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CartesianTrajectoryPoint"};
  }
  hasMember(name) {
    let memberNames = ["time_from_start","pose","twist","acceleration"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CartesianTrajectoryPointArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CartesianTrajectoryPointWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CartesianTrajectoryPointWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryPointArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianTrajectoryPointArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CartesianTrajectoryPointRefArray(size);
    this._refObject = new CartesianTrajectoryPointRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CartesianTrajectoryPointWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CartesianTrajectoryPointArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CartesianTrajectoryPointArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CartesianTrajectoryPointWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CartesianTrajectoryPointWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CartesianTrajectoryPointArrayWrapper;
  }
}
module.exports = CartesianTrajectoryPointWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "time_from_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twist",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "acceleration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Accel",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "CartesianTrajectoryPoint",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CartesianTrajectoryPoint"
}
*/
