// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__V2xTriggerConfiguration.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TriggerConfigurationWrapper = require('../../generated/atos_interfaces/atos_interfaces__msg__TriggerConfiguration.js');
const V2xWrapper = require('../../generated/atos_interfaces/atos_interfaces__msg__V2x.js');
const V2xTriggerConfigurationRefStruct = StructType({
  trigger_configuration: TriggerConfigurationWrapper.refObjectType,
  v2x_message: V2xWrapper.refObjectType,
});
const V2xTriggerConfigurationRefArray = ArrayType(V2xTriggerConfigurationRefStruct);
const V2xTriggerConfigurationRefStructArray = StructType({
  data: V2xTriggerConfigurationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class V2xTriggerConfigurationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new V2xTriggerConfigurationRefStruct(other._refObject.toObject());
      this._wrapperFields.trigger_configuration =  new TriggerConfigurationWrapper(other._wrapperFields.trigger_configuration);
      this._wrapperFields.v2x_message =  new V2xWrapper(other._wrapperFields.v2x_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new V2xTriggerConfigurationRefStruct();
    this._wrapperFields.trigger_configuration = new TriggerConfigurationWrapper();
    this._wrapperFields.v2x_message = new V2xWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new V2xTriggerConfigurationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new V2xTriggerConfigurationArrayWrapper;
  }
  static get ArrayType() {
    return V2xTriggerConfigurationArrayWrapper;
  }
  static get refObjectArrayType() {
    return V2xTriggerConfigurationRefStructArray
  }
  static get refObjectType() {
    return V2xTriggerConfigurationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.trigger_configuration.freeze(own, checkConsistency);
    this._refObject.trigger_configuration = this._wrapperFields.trigger_configuration.refObject;
    this._wrapperFields.v2x_message.freeze(own, checkConsistency);
    this._refObject.v2x_message = this._wrapperFields.v2x_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.trigger_configuration.copyRefObject(refObject.trigger_configuration);
    this._wrapperFields.v2x_message.copyRefObject(refObject.v2x_message);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TriggerConfigurationWrapper.freeStruct(refObject.trigger_configuration);
    V2xWrapper.freeStruct(refObject.v2x_message);
  }
  static destoryRawROS(msg) {
    V2xTriggerConfigurationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'V2xTriggerConfiguration'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trigger_configuration() {
    return this._wrapperFields.trigger_configuration;
  }
  set trigger_configuration(value) {
    if (value instanceof TriggerConfigurationWrapper) {
      this._wrapperFields.trigger_configuration.copy(value);
    } else {
      this._wrapperFields.trigger_configuration.copy(new TriggerConfigurationWrapper(value));
    }
  }
  get v2x_message() {
    return this._wrapperFields.v2x_message;
  }
  set v2x_message(value) {
    if (value instanceof V2xWrapper) {
      this._wrapperFields.v2x_message.copy(value);
    } else {
      this._wrapperFields.v2x_message.copy(new V2xWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new V2xTriggerConfigurationRefStruct(refObject.toObject());
    this._wrapperFields.trigger_configuration.copyRefObject(this._refObject.trigger_configuration);
    this._wrapperFields.v2x_message.copyRefObject(this._refObject.v2x_message);
  }
  copy(other) {
    this._refObject = new V2xTriggerConfigurationRefStruct(other._refObject.toObject());
    this._wrapperFields.trigger_configuration.copy(other._wrapperFields.trigger_configuration);
    this._wrapperFields.v2x_message.copy(other._wrapperFields.v2x_message);
  }
  static get classType() {
    return V2xTriggerConfigurationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trigger_configuration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"atos_interfaces","type":"TriggerConfiguration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"v2x_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"atos_interfaces","type":"V2x","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"V2xTriggerConfiguration","stringUpperBound":null,"isPrimitiveType":false},"msgName":"V2xTriggerConfiguration"};
  }
  hasMember(name) {
    let memberNames = ["trigger_configuration","v2x_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class V2xTriggerConfigurationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof V2xTriggerConfigurationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new V2xTriggerConfigurationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to V2xTriggerConfigurationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to V2xTriggerConfigurationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new V2xTriggerConfigurationRefArray(size);
    this._refObject = new V2xTriggerConfigurationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new V2xTriggerConfigurationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof V2xTriggerConfigurationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "V2xTriggerConfigurationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      V2xTriggerConfigurationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return V2xTriggerConfigurationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return V2xTriggerConfigurationArrayWrapper;
  }
}
module.exports = V2xTriggerConfigurationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trigger_configuration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "atos_interfaces",
        "type": "TriggerConfiguration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "v2x_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "atos_interfaces",
        "type": "V2x",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "V2xTriggerConfiguration",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "V2xTriggerConfiguration"
}
*/
