// rclnodejs/generated/atos_interfaces/atos_interfaces__srv__GetObjectControlState_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetObjectControlState_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  state: primitiveTypes.uint8,
});
const GetObjectControlState_ResponseRefArray = ArrayType(GetObjectControlState_ResponseRefStruct);
const GetObjectControlState_ResponseRefStructArray = StructType({
  data: GetObjectControlState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetObjectControlState_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetObjectControlState_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._stateIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetObjectControlState_ResponseRefStruct();
    this._refObject.success = true;
    this._successIntialized = true;
    this._stateIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GetObjectControlState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetObjectControlState_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetObjectControlState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetObjectControlState_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetObjectControlState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetObjectControlState_Response');
      }
      if (!this._stateIntialized) {
        throw new TypeError('Invalid argument: state in GetObjectControlState_Response');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._stateIntialized = true;
    this._refObject.state = refObject.state;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GetObjectControlState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'srv', interfaceName: 'GetObjectControlState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get state() {
    if (!this._stateIntialized) {
      return undefined;
    }
    return this._refObject.state;
  }
  set state(value) {
    this._stateIntialized = true;
    this._refObject.state = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetObjectControlState_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._stateIntialized = true;
  }
  copy(other) {
    this._refObject = new GetObjectControlState_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._stateIntialized = true;
  }
  static get classType() {
    return GetObjectControlState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"UNDEFINED","value":0},{"type":"uint8","name":"IDLE","value":1},{"type":"uint8","name":"INITIALIZED","value":2},{"type":"uint8","name":"CONNECTED","value":3},{"type":"uint8","name":"ARMED","value":4},{"type":"uint8","name":"RUNNING","value":5},{"type":"uint8","name":"REMOTECTRL","value":6},{"type":"uint8","name":"ERROR","value":7},{"type":"uint8","name":"ABORTING","value":8}],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true},{"name":"state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"GetObjectControlState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetObjectControlState_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","state"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetObjectControlState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetObjectControlState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetObjectControlState_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectControlState_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectControlState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetObjectControlState_ResponseRefArray(size);
    this._refObject = new GetObjectControlState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetObjectControlState_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetObjectControlState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetObjectControlState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetObjectControlState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetObjectControlState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetObjectControlState_ResponseArrayWrapper;
  }
}
// Define constants (9 in total)
Object.defineProperty(GetObjectControlState_ResponseWrapper, "UNDEFINED", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "IDLE", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "INITIALIZED", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "CONNECTED", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "ARMED", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "RUNNING", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "REMOTECTRL", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "ERROR", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(GetObjectControlState_ResponseWrapper, "ABORTING", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = GetObjectControlState_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "UNDEFINED",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "IDLE",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "INITIALIZED",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "CONNECTED",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "ARMED",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "RUNNING",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "REMOTECTRL",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "ERROR",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "ABORTING",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": true
    },
    {
      "name": "state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "GetObjectControlState_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetObjectControlState_Response"
}
*/
