// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ManoeuvreCommand.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ManoeuvreCommandRefStruct = StructType({
  manoeuvre: primitiveTypes.int8,
  object_ip: primitiveTypes.uint32,
});
const ManoeuvreCommandRefArray = ArrayType(ManoeuvreCommandRefStruct);
const ManoeuvreCommandRefStructArray = StructType({
  data: ManoeuvreCommandRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ManoeuvreCommandWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ManoeuvreCommandRefStruct(other._refObject.toObject());
      this._manoeuvreIntialized = true;
      this._object_ipIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ManoeuvreCommandRefStruct();
    this._manoeuvreIntialized = false;
    this._object_ipIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ManoeuvreCommandWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ManoeuvreCommandArrayWrapper;
  }
  static get ArrayType() {
    return ManoeuvreCommandArrayWrapper;
  }
  static get refObjectArrayType() {
    return ManoeuvreCommandRefStructArray
  }
  static get refObjectType() {
    return ManoeuvreCommandRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._manoeuvreIntialized) {
        throw new TypeError('Invalid argument: manoeuvre in ManoeuvreCommand');
      }
      if (!this._object_ipIntialized) {
        throw new TypeError('Invalid argument: object_ip in ManoeuvreCommand');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._manoeuvreIntialized = true;
    this._refObject.manoeuvre = refObject.manoeuvre;
    this._object_ipIntialized = true;
    this._refObject.object_ip = refObject.object_ip;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ManoeuvreCommandWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ManoeuvreCommand'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get manoeuvre() {
    if (!this._manoeuvreIntialized) {
      return undefined;
    }
    return this._refObject.manoeuvre;
  }
  set manoeuvre(value) {
    this._manoeuvreIntialized = true;
    this._refObject.manoeuvre = value;
  }
  get object_ip() {
    if (!this._object_ipIntialized) {
      return undefined;
    }
    return this._refObject.object_ip;
  }
  set object_ip(value) {
    this._object_ipIntialized = true;
    this._refObject.object_ip = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ManoeuvreCommandRefStruct(refObject.toObject());
    this._manoeuvreIntialized = true;
    this._object_ipIntialized = true;
  }
  copy(other) {
    this._refObject = new ManoeuvreCommandRefStruct(other._refObject.toObject());
    this._manoeuvreIntialized = true;
    this._object_ipIntialized = true;
  }
  static get classType() {
    return ManoeuvreCommandWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"manoeuvre","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"object_ip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ManoeuvreCommand","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ManoeuvreCommand"};
  }
  hasMember(name) {
    let memberNames = ["manoeuvre","object_ip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ManoeuvreCommandArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ManoeuvreCommandWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ManoeuvreCommandWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ManoeuvreCommandArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ManoeuvreCommandArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ManoeuvreCommandRefArray(size);
    this._refObject = new ManoeuvreCommandRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ManoeuvreCommandWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ManoeuvreCommandArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ManoeuvreCommandArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ManoeuvreCommandWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ManoeuvreCommandWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ManoeuvreCommandArrayWrapper;
  }
}
module.exports = ManoeuvreCommandWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "manoeuvre",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "object_ip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ManoeuvreCommand",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ManoeuvreCommand"
}
*/
