// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__V2x.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const V2xRefStruct = StructType({
  message_type: primitiveTypes.string,
  event_id: primitiveTypes.string,
  cause_code: primitiveTypes.uint8,
  detection_time: primitiveTypes.uint64,
  altitude: primitiveTypes.int32,
  latitude: primitiveTypes.int32,
  longitude: primitiveTypes.int32,
});
const V2xRefArray = ArrayType(V2xRefStruct);
const V2xRefStructArray = StructType({
  data: V2xRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class V2xWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new V2xRefStruct(other._refObject.toObject());
      this._message_typeIntialized = true;
      this._wrapperFields.message_type =  new StringWrapper(other._wrapperFields.message_type);
      this._event_idIntialized = true;
      this._wrapperFields.event_id =  new StringWrapper(other._wrapperFields.event_id);
      this._cause_codeIntialized = true;
      this._detection_timeIntialized = true;
      this._altitudeIntialized = true;
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new V2xRefStruct();
    this._message_typeIntialized = false;
    this._wrapperFields.message_type = new StringWrapper();
    this._event_idIntialized = false;
    this._wrapperFields.event_id = new StringWrapper();
    this._cause_codeIntialized = false;
    this._detection_timeIntialized = false;
    this._altitudeIntialized = false;
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new V2xWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new V2xArrayWrapper;
  }
  static get ArrayType() {
    return V2xArrayWrapper;
  }
  static get refObjectArrayType() {
    return V2xRefStructArray
  }
  static get refObjectType() {
    return V2xRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._message_typeIntialized) {
        throw new TypeError('Invalid argument: message_type in V2x');
      }
      if (!this._event_idIntialized) {
        throw new TypeError('Invalid argument: event_id in V2x');
      }
      if (!this._cause_codeIntialized) {
        throw new TypeError('Invalid argument: cause_code in V2x');
      }
      if (!this._detection_timeIntialized) {
        throw new TypeError('Invalid argument: detection_time in V2x');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in V2x');
      }
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in V2x');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in V2x');
      }
    }
    if (own) {
      this._wrapperFields.message_type.freeze(own, checkConsistency);
    }
    this._refObject.message_type = this._wrapperFields.message_type.refObject;
    if (own) {
      this._wrapperFields.event_id.freeze(own, checkConsistency);
    }
    this._refObject.event_id = this._wrapperFields.event_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._message_typeIntialized = true;
    this._wrapperFields.message_type.data = refObject.message_type.data;
    this._event_idIntialized = true;
    this._wrapperFields.event_id.data = refObject.event_id.data;
    this._cause_codeIntialized = true;
    this._refObject.cause_code = refObject.cause_code;
    this._detection_timeIntialized = true;
    this._refObject.detection_time = refObject.detection_time;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.message_type);
    StringWrapper.freeStruct(refObject.event_id);
  }
  static destoryRawROS(msg) {
    V2xWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'V2x'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get message_type() {
    if (!this._message_typeIntialized) {
      return undefined;
    }
    return this._wrapperFields.message_type.data;
  }
  set message_type(value) {
    this._message_typeIntialized = true;
    this._wrapperFields.message_type.data = value;
  }
  get event_id() {
    if (!this._event_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.event_id.data;
  }
  set event_id(value) {
    this._event_idIntialized = true;
    this._wrapperFields.event_id.data = value;
  }
  get cause_code() {
    if (!this._cause_codeIntialized) {
      return undefined;
    }
    return this._refObject.cause_code;
  }
  set cause_code(value) {
    this._cause_codeIntialized = true;
    this._refObject.cause_code = value;
  }
  get detection_time() {
    if (!this._detection_timeIntialized) {
      return undefined;
    }
    return this._refObject.detection_time;
  }
  set detection_time(value) {
    this._detection_timeIntialized = true;
    this._refObject.detection_time = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  copyRefObject(refObject) {
    this._refObject = new V2xRefStruct(refObject.toObject());
    this._message_typeIntialized = true;
    this._wrapperFields.message_type.copyRefObject(this._refObject.message_type);
    this._event_idIntialized = true;
    this._wrapperFields.event_id.copyRefObject(this._refObject.event_id);
    this._cause_codeIntialized = true;
    this._detection_timeIntialized = true;
    this._altitudeIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
  }
  copy(other) {
    this._refObject = new V2xRefStruct(other._refObject.toObject());
    this._message_typeIntialized = true;
    this._wrapperFields.message_type.copy(other._wrapperFields.message_type);
    this._event_idIntialized = true;
    this._wrapperFields.event_id.copy(other._wrapperFields.event_id);
    this._cause_codeIntialized = true;
    this._detection_timeIntialized = true;
    this._altitudeIntialized = true;
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
  }
  static get classType() {
    return V2xWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"message_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"event_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cause_code","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"detection_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"V2x","stringUpperBound":null,"isPrimitiveType":false},"msgName":"V2x"};
  }
  hasMember(name) {
    let memberNames = ["message_type","event_id","cause_code","detection_time","altitude","latitude","longitude"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class V2xArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof V2xWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new V2xWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to V2xArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to V2xArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new V2xRefArray(size);
    this._refObject = new V2xRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new V2xWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof V2xArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "V2xArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      V2xWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return V2xWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return V2xArrayWrapper;
  }
}
module.exports = V2xWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "message_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "event_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cause_code",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "detection_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "V2x",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "V2x"
}
*/
