// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ControlSignalPercentage.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ATOSHeaderWrapper = require('../../generated/atos_interfaces/atos_interfaces__msg__ATOSHeader.js');
const ControlSignalPercentageRefStruct = StructType({
  atos_header: ATOSHeaderWrapper.refObjectType,
  throttle: primitiveTypes.float64,
  brake: primitiveTypes.float64,
  steering_angle: primitiveTypes.float64,
});
const ControlSignalPercentageRefArray = ArrayType(ControlSignalPercentageRefStruct);
const ControlSignalPercentageRefStructArray = StructType({
  data: ControlSignalPercentageRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ControlSignalPercentageWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ControlSignalPercentageRefStruct(other._refObject.toObject());
      this._wrapperFields.atos_header =  new ATOSHeaderWrapper(other._wrapperFields.atos_header);
      this._throttleIntialized = true;
      this._brakeIntialized = true;
      this._steering_angleIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ControlSignalPercentageRefStruct();
    this._wrapperFields.atos_header = new ATOSHeaderWrapper();
    this._throttleIntialized = false;
    this._brakeIntialized = false;
    this._steering_angleIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ControlSignalPercentageWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ControlSignalPercentageArrayWrapper;
  }
  static get ArrayType() {
    return ControlSignalPercentageArrayWrapper;
  }
  static get refObjectArrayType() {
    return ControlSignalPercentageRefStructArray
  }
  static get refObjectType() {
    return ControlSignalPercentageRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._throttleIntialized) {
        throw new TypeError('Invalid argument: throttle in ControlSignalPercentage');
      }
      if (!this._brakeIntialized) {
        throw new TypeError('Invalid argument: brake in ControlSignalPercentage');
      }
      if (!this._steering_angleIntialized) {
        throw new TypeError('Invalid argument: steering_angle in ControlSignalPercentage');
      }
    }
    this._wrapperFields.atos_header.freeze(own, checkConsistency);
    this._refObject.atos_header = this._wrapperFields.atos_header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.atos_header.copyRefObject(refObject.atos_header);
    this._throttleIntialized = true;
    this._refObject.throttle = refObject.throttle;
    this._brakeIntialized = true;
    this._refObject.brake = refObject.brake;
    this._steering_angleIntialized = true;
    this._refObject.steering_angle = refObject.steering_angle;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ATOSHeaderWrapper.freeStruct(refObject.atos_header);
  }
  static destoryRawROS(msg) {
    ControlSignalPercentageWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ControlSignalPercentage'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get atos_header() {
    return this._wrapperFields.atos_header;
  }
  set atos_header(value) {
    if (value instanceof ATOSHeaderWrapper) {
      this._wrapperFields.atos_header.copy(value);
    } else {
      this._wrapperFields.atos_header.copy(new ATOSHeaderWrapper(value));
    }
  }
  get throttle() {
    if (!this._throttleIntialized) {
      return undefined;
    }
    return this._refObject.throttle;
  }
  set throttle(value) {
    this._throttleIntialized = true;
    this._refObject.throttle = value;
  }
  get brake() {
    if (!this._brakeIntialized) {
      return undefined;
    }
    return this._refObject.brake;
  }
  set brake(value) {
    this._brakeIntialized = true;
    this._refObject.brake = value;
  }
  get steering_angle() {
    if (!this._steering_angleIntialized) {
      return undefined;
    }
    return this._refObject.steering_angle;
  }
  set steering_angle(value) {
    this._steering_angleIntialized = true;
    this._refObject.steering_angle = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ControlSignalPercentageRefStruct(refObject.toObject());
    this._wrapperFields.atos_header.copyRefObject(this._refObject.atos_header);
    this._throttleIntialized = true;
    this._brakeIntialized = true;
    this._steering_angleIntialized = true;
  }
  copy(other) {
    this._refObject = new ControlSignalPercentageRefStruct(other._refObject.toObject());
    this._wrapperFields.atos_header.copy(other._wrapperFields.atos_header);
    this._throttleIntialized = true;
    this._brakeIntialized = true;
    this._steering_angleIntialized = true;
  }
  static get classType() {
    return ControlSignalPercentageWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"atos_header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"atos_interfaces","type":"ATOSHeader","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"throttle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"brake","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"steering_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ControlSignalPercentage","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ControlSignalPercentage"};
  }
  hasMember(name) {
    let memberNames = ["atos_header","throttle","brake","steering_angle"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ControlSignalPercentageArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ControlSignalPercentageWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ControlSignalPercentageWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ControlSignalPercentageArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ControlSignalPercentageArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ControlSignalPercentageRefArray(size);
    this._refObject = new ControlSignalPercentageRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ControlSignalPercentageWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ControlSignalPercentageArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ControlSignalPercentageArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ControlSignalPercentageWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ControlSignalPercentageWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ControlSignalPercentageArrayWrapper;
  }
}
module.exports = ControlSignalPercentageWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "atos_header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "atos_interfaces",
        "type": "ATOSHeader",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "throttle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "brake",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "steering_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ControlSignalPercentage",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ControlSignalPercentage"
}
*/
