// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__TriggerConfiguration.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TriggerConfigurationRefStruct = StructType({
  trigger_id: primitiveTypes.uint16,
  trigger_type: primitiveTypes.uint16,
  treo_send_count: primitiveTypes.uint16,
  treo_send_interval: primitiveTypes.uint16,
  trigger_comparison_type: primitiveTypes.uint8,
  trigger_param_sign: primitiveTypes.uint8,
  trigger_type_parameter1: primitiveTypes.uint32,
  trigger_type_parameter2: primitiveTypes.uint32,
  trigger_type_parameter3: primitiveTypes.uint32,
  trigger_type_parameter4: primitiveTypes.uint32,
  trigger_type_parameter5: primitiveTypes.uint32,
  ip: primitiveTypes.uint32,
});
const TriggerConfigurationRefArray = ArrayType(TriggerConfigurationRefStruct);
const TriggerConfigurationRefStructArray = StructType({
  data: TriggerConfigurationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TriggerConfigurationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TriggerConfigurationRefStruct(other._refObject.toObject());
      this._trigger_idIntialized = true;
      this._trigger_typeIntialized = true;
      this._treo_send_countIntialized = true;
      this._treo_send_intervalIntialized = true;
      this._trigger_comparison_typeIntialized = true;
      this._trigger_param_signIntialized = true;
      this._trigger_type_parameter1Intialized = true;
      this._trigger_type_parameter2Intialized = true;
      this._trigger_type_parameter3Intialized = true;
      this._trigger_type_parameter4Intialized = true;
      this._trigger_type_parameter5Intialized = true;
      this._ipIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TriggerConfigurationRefStruct();
    this._trigger_idIntialized = false;
    this._trigger_typeIntialized = false;
    this._treo_send_countIntialized = false;
    this._treo_send_intervalIntialized = false;
    this._trigger_comparison_typeIntialized = false;
    this._trigger_param_signIntialized = false;
    this._trigger_type_parameter1Intialized = false;
    this._trigger_type_parameter2Intialized = false;
    this._trigger_type_parameter3Intialized = false;
    this._trigger_type_parameter4Intialized = false;
    this._trigger_type_parameter5Intialized = false;
    this._ipIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TriggerConfigurationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TriggerConfigurationArrayWrapper;
  }
  static get ArrayType() {
    return TriggerConfigurationArrayWrapper;
  }
  static get refObjectArrayType() {
    return TriggerConfigurationRefStructArray
  }
  static get refObjectType() {
    return TriggerConfigurationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._trigger_idIntialized) {
        throw new TypeError('Invalid argument: trigger_id in TriggerConfiguration');
      }
      if (!this._trigger_typeIntialized) {
        throw new TypeError('Invalid argument: trigger_type in TriggerConfiguration');
      }
      if (!this._treo_send_countIntialized) {
        throw new TypeError('Invalid argument: treo_send_count in TriggerConfiguration');
      }
      if (!this._treo_send_intervalIntialized) {
        throw new TypeError('Invalid argument: treo_send_interval in TriggerConfiguration');
      }
      if (!this._trigger_comparison_typeIntialized) {
        throw new TypeError('Invalid argument: trigger_comparison_type in TriggerConfiguration');
      }
      if (!this._trigger_param_signIntialized) {
        throw new TypeError('Invalid argument: trigger_param_sign in TriggerConfiguration');
      }
      if (!this._trigger_type_parameter1Intialized) {
        throw new TypeError('Invalid argument: trigger_type_parameter1 in TriggerConfiguration');
      }
      if (!this._trigger_type_parameter2Intialized) {
        throw new TypeError('Invalid argument: trigger_type_parameter2 in TriggerConfiguration');
      }
      if (!this._trigger_type_parameter3Intialized) {
        throw new TypeError('Invalid argument: trigger_type_parameter3 in TriggerConfiguration');
      }
      if (!this._trigger_type_parameter4Intialized) {
        throw new TypeError('Invalid argument: trigger_type_parameter4 in TriggerConfiguration');
      }
      if (!this._trigger_type_parameter5Intialized) {
        throw new TypeError('Invalid argument: trigger_type_parameter5 in TriggerConfiguration');
      }
      if (!this._ipIntialized) {
        throw new TypeError('Invalid argument: ip in TriggerConfiguration');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = refObject.trigger_id;
    this._trigger_typeIntialized = true;
    this._refObject.trigger_type = refObject.trigger_type;
    this._treo_send_countIntialized = true;
    this._refObject.treo_send_count = refObject.treo_send_count;
    this._treo_send_intervalIntialized = true;
    this._refObject.treo_send_interval = refObject.treo_send_interval;
    this._trigger_comparison_typeIntialized = true;
    this._refObject.trigger_comparison_type = refObject.trigger_comparison_type;
    this._trigger_param_signIntialized = true;
    this._refObject.trigger_param_sign = refObject.trigger_param_sign;
    this._trigger_type_parameter1Intialized = true;
    this._refObject.trigger_type_parameter1 = refObject.trigger_type_parameter1;
    this._trigger_type_parameter2Intialized = true;
    this._refObject.trigger_type_parameter2 = refObject.trigger_type_parameter2;
    this._trigger_type_parameter3Intialized = true;
    this._refObject.trigger_type_parameter3 = refObject.trigger_type_parameter3;
    this._trigger_type_parameter4Intialized = true;
    this._refObject.trigger_type_parameter4 = refObject.trigger_type_parameter4;
    this._trigger_type_parameter5Intialized = true;
    this._refObject.trigger_type_parameter5 = refObject.trigger_type_parameter5;
    this._ipIntialized = true;
    this._refObject.ip = refObject.ip;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TriggerConfigurationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'TriggerConfiguration'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trigger_id() {
    if (!this._trigger_idIntialized) {
      return undefined;
    }
    return this._refObject.trigger_id;
  }
  set trigger_id(value) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = value;
  }
  get trigger_type() {
    if (!this._trigger_typeIntialized) {
      return undefined;
    }
    return this._refObject.trigger_type;
  }
  set trigger_type(value) {
    this._trigger_typeIntialized = true;
    this._refObject.trigger_type = value;
  }
  get treo_send_count() {
    if (!this._treo_send_countIntialized) {
      return undefined;
    }
    return this._refObject.treo_send_count;
  }
  set treo_send_count(value) {
    this._treo_send_countIntialized = true;
    this._refObject.treo_send_count = value;
  }
  get treo_send_interval() {
    if (!this._treo_send_intervalIntialized) {
      return undefined;
    }
    return this._refObject.treo_send_interval;
  }
  set treo_send_interval(value) {
    this._treo_send_intervalIntialized = true;
    this._refObject.treo_send_interval = value;
  }
  get trigger_comparison_type() {
    if (!this._trigger_comparison_typeIntialized) {
      return undefined;
    }
    return this._refObject.trigger_comparison_type;
  }
  set trigger_comparison_type(value) {
    this._trigger_comparison_typeIntialized = true;
    this._refObject.trigger_comparison_type = value;
  }
  get trigger_param_sign() {
    if (!this._trigger_param_signIntialized) {
      return undefined;
    }
    return this._refObject.trigger_param_sign;
  }
  set trigger_param_sign(value) {
    this._trigger_param_signIntialized = true;
    this._refObject.trigger_param_sign = value;
  }
  get trigger_type_parameter1() {
    if (!this._trigger_type_parameter1Intialized) {
      return undefined;
    }
    return this._refObject.trigger_type_parameter1;
  }
  set trigger_type_parameter1(value) {
    this._trigger_type_parameter1Intialized = true;
    this._refObject.trigger_type_parameter1 = value;
  }
  get trigger_type_parameter2() {
    if (!this._trigger_type_parameter2Intialized) {
      return undefined;
    }
    return this._refObject.trigger_type_parameter2;
  }
  set trigger_type_parameter2(value) {
    this._trigger_type_parameter2Intialized = true;
    this._refObject.trigger_type_parameter2 = value;
  }
  get trigger_type_parameter3() {
    if (!this._trigger_type_parameter3Intialized) {
      return undefined;
    }
    return this._refObject.trigger_type_parameter3;
  }
  set trigger_type_parameter3(value) {
    this._trigger_type_parameter3Intialized = true;
    this._refObject.trigger_type_parameter3 = value;
  }
  get trigger_type_parameter4() {
    if (!this._trigger_type_parameter4Intialized) {
      return undefined;
    }
    return this._refObject.trigger_type_parameter4;
  }
  set trigger_type_parameter4(value) {
    this._trigger_type_parameter4Intialized = true;
    this._refObject.trigger_type_parameter4 = value;
  }
  get trigger_type_parameter5() {
    if (!this._trigger_type_parameter5Intialized) {
      return undefined;
    }
    return this._refObject.trigger_type_parameter5;
  }
  set trigger_type_parameter5(value) {
    this._trigger_type_parameter5Intialized = true;
    this._refObject.trigger_type_parameter5 = value;
  }
  get ip() {
    if (!this._ipIntialized) {
      return undefined;
    }
    return this._refObject.ip;
  }
  set ip(value) {
    this._ipIntialized = true;
    this._refObject.ip = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TriggerConfigurationRefStruct(refObject.toObject());
    this._trigger_idIntialized = true;
    this._trigger_typeIntialized = true;
    this._treo_send_countIntialized = true;
    this._treo_send_intervalIntialized = true;
    this._trigger_comparison_typeIntialized = true;
    this._trigger_param_signIntialized = true;
    this._trigger_type_parameter1Intialized = true;
    this._trigger_type_parameter2Intialized = true;
    this._trigger_type_parameter3Intialized = true;
    this._trigger_type_parameter4Intialized = true;
    this._trigger_type_parameter5Intialized = true;
    this._ipIntialized = true;
  }
  copy(other) {
    this._refObject = new TriggerConfigurationRefStruct(other._refObject.toObject());
    this._trigger_idIntialized = true;
    this._trigger_typeIntialized = true;
    this._treo_send_countIntialized = true;
    this._treo_send_intervalIntialized = true;
    this._trigger_comparison_typeIntialized = true;
    this._trigger_param_signIntialized = true;
    this._trigger_type_parameter1Intialized = true;
    this._trigger_type_parameter2Intialized = true;
    this._trigger_type_parameter3Intialized = true;
    this._trigger_type_parameter4Intialized = true;
    this._trigger_type_parameter5Intialized = true;
    this._ipIntialized = true;
  }
  static get classType() {
    return TriggerConfigurationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trigger_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"treo_send_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"treo_send_interval","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_comparison_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_param_sign","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type_parameter1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type_parameter2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type_parameter3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type_parameter4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trigger_type_parameter5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"TriggerConfiguration","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TriggerConfiguration"};
  }
  hasMember(name) {
    let memberNames = ["trigger_id","trigger_type","treo_send_count","treo_send_interval","trigger_comparison_type","trigger_param_sign","trigger_type_parameter1","trigger_type_parameter2","trigger_type_parameter3","trigger_type_parameter4","trigger_type_parameter5","ip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TriggerConfigurationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TriggerConfigurationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TriggerConfigurationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggerConfigurationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggerConfigurationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TriggerConfigurationRefArray(size);
    this._refObject = new TriggerConfigurationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TriggerConfigurationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TriggerConfigurationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TriggerConfigurationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TriggerConfigurationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TriggerConfigurationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TriggerConfigurationArrayWrapper;
  }
}
module.exports = TriggerConfigurationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trigger_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "treo_send_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "treo_send_interval",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_comparison_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_param_sign",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type_parameter1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type_parameter2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type_parameter3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type_parameter4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trigger_type_parameter5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "TriggerConfiguration",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TriggerConfiguration"
}
*/
