// rclnodejs/generated/atos_interfaces/atos_interfaces__srv__GetObjectIds_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const GetObjectIds_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  ids: UInt32Wrapper.refObjectArrayType,
});
const GetObjectIds_ResponseRefArray = ArrayType(GetObjectIds_ResponseRefStruct);
const GetObjectIds_ResponseRefStructArray = StructType({
  data: GetObjectIds_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetObjectIds_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetObjectIds_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._wrapperFields.ids = UInt32Wrapper.createArray();
      this._wrapperFields.ids.copy(other._wrapperFields.ids);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetObjectIds_ResponseRefStruct();
    this._refObject.success = true;
    this._successIntialized = true;
    this._wrapperFields.ids = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetObjectIds_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetObjectIds_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetObjectIds_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetObjectIds_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetObjectIds_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetObjectIds_Response');
      }
    }
    this._wrapperFields.ids.freeze(own, checkConsistency);
    this._refObject.ids = this._wrapperFields.ids.refObject;
    if (own) {
      this._wrapperFields.ids.fill(Uint32Array.from([]));
      this._wrapperFields.ids.freeze(own, checkConsistency);
      this._refObject.ids = this._wrapperFields.ids.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._wrapperFields.ids.copyRefObject(refObject.ids);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ids.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.ids);
      if (UInt32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ids, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetObjectIds_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'srv', interfaceName: 'GetObjectIds_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get ids() {
    return this._wrapperFields['ids'].data;
  }
  set ids(value) {
    this._wrapperFields['ids'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetObjectIds_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._wrapperFields.ids.copyRefObject(this._refObject.ids);
  }
  copy(other) {
    this._refObject = new GetObjectIds_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._wrapperFields.ids.copy(other._wrapperFields.ids);
  }
  static get classType() {
    return GetObjectIds_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true},{"name":"ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"GetObjectIds_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetObjectIds_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","ids"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetObjectIds_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetObjectIds_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetObjectIds_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectIds_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectIds_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetObjectIds_ResponseRefArray(size);
    this._refObject = new GetObjectIds_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetObjectIds_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetObjectIds_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetObjectIds_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetObjectIds_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetObjectIds_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetObjectIds_ResponseArrayWrapper;
  }
}
module.exports = GetObjectIds_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": true
    },
    {
      "name": "ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "GetObjectIds_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetObjectIds_Response"
}
*/
