// rclnodejs/generated/atos_interfaces/atos_interfaces__srv__GetObjectTrajectory_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CartesianTrajectoryWrapper = require('../../generated/atos_interfaces/atos_interfaces__msg__CartesianTrajectory.js');
const GetObjectTrajectory_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  id: primitiveTypes.uint32,
  trajectory: CartesianTrajectoryWrapper.refObjectType,
});
const GetObjectTrajectory_ResponseRefArray = ArrayType(GetObjectTrajectory_ResponseRefStruct);
const GetObjectTrajectory_ResponseRefStructArray = StructType({
  data: GetObjectTrajectory_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetObjectTrajectory_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetObjectTrajectory_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._idIntialized = true;
      this._wrapperFields.trajectory =  new CartesianTrajectoryWrapper(other._wrapperFields.trajectory);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetObjectTrajectory_ResponseRefStruct();
    this._refObject.success = true;
    this._successIntialized = true;
    this._idIntialized = false;
    this._wrapperFields.trajectory = new CartesianTrajectoryWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetObjectTrajectory_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetObjectTrajectory_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetObjectTrajectory_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetObjectTrajectory_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetObjectTrajectory_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetObjectTrajectory_Response');
      }
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in GetObjectTrajectory_Response');
      }
    }
    this._wrapperFields.trajectory.freeze(own, checkConsistency);
    this._refObject.trajectory = this._wrapperFields.trajectory.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._wrapperFields.trajectory.copyRefObject(refObject.trajectory);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    CartesianTrajectoryWrapper.freeStruct(refObject.trajectory);
  }
  static destoryRawROS(msg) {
    GetObjectTrajectory_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'srv', interfaceName: 'GetObjectTrajectory_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get trajectory() {
    return this._wrapperFields.trajectory;
  }
  set trajectory(value) {
    if (value instanceof CartesianTrajectoryWrapper) {
      this._wrapperFields.trajectory.copy(value);
    } else {
      this._wrapperFields.trajectory.copy(new CartesianTrajectoryWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetObjectTrajectory_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._idIntialized = true;
    this._wrapperFields.trajectory.copyRefObject(this._refObject.trajectory);
  }
  copy(other) {
    this._refObject = new GetObjectTrajectory_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._idIntialized = true;
    this._wrapperFields.trajectory.copy(other._wrapperFields.trajectory);
  }
  static get classType() {
    return GetObjectTrajectory_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"atos_interfaces","type":"CartesianTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"GetObjectTrajectory_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetObjectTrajectory_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","id","trajectory"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetObjectTrajectory_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetObjectTrajectory_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetObjectTrajectory_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectTrajectory_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectTrajectory_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetObjectTrajectory_ResponseRefArray(size);
    this._refObject = new GetObjectTrajectory_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetObjectTrajectory_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetObjectTrajectory_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetObjectTrajectory_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetObjectTrajectory_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetObjectTrajectory_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetObjectTrajectory_ResponseArrayWrapper;
  }
}
module.exports = GetObjectTrajectory_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": true
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "atos_interfaces",
        "type": "CartesianTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "GetObjectTrajectory_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetObjectTrajectory_Response"
}
*/
