// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ReadyToArm.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ReadyToArmRefStruct = StructType({
  ready_to_arm: primitiveTypes.uint8,
});
const ReadyToArmRefArray = ArrayType(ReadyToArmRefStruct);
const ReadyToArmRefStructArray = StructType({
  data: ReadyToArmRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ReadyToArmWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ReadyToArmRefStruct(other._refObject.toObject());
      this._ready_to_armIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ReadyToArmRefStruct();
    this._ready_to_armIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ReadyToArmWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ReadyToArmArrayWrapper;
  }
  static get ArrayType() {
    return ReadyToArmArrayWrapper;
  }
  static get refObjectArrayType() {
    return ReadyToArmRefStructArray
  }
  static get refObjectType() {
    return ReadyToArmRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._ready_to_armIntialized) {
        throw new TypeError('Invalid argument: ready_to_arm in ReadyToArm');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._ready_to_armIntialized = true;
    this._refObject.ready_to_arm = refObject.ready_to_arm;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ReadyToArmWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ReadyToArm'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ready_to_arm() {
    if (!this._ready_to_armIntialized) {
      return undefined;
    }
    return this._refObject.ready_to_arm;
  }
  set ready_to_arm(value) {
    this._ready_to_armIntialized = true;
    this._refObject.ready_to_arm = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ReadyToArmRefStruct(refObject.toObject());
    this._ready_to_armIntialized = true;
  }
  copy(other) {
    this._refObject = new ReadyToArmRefStruct(other._refObject.toObject());
    this._ready_to_armIntialized = true;
  }
  static get classType() {
    return ReadyToArmWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"NOT_READY","value":0},{"type":"uint8","name":"READY_TO_ARM","value":1},{"type":"uint8","name":"NOT_READY_NO_TRAJ","value":2},{"type":"uint8","name":"NOT_READY_NO_OSEM","value":3},{"type":"uint8","name":"UNAVALIABLE","value":255}],"fields":[{"name":"ready_to_arm","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ReadyToArm","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ReadyToArm"};
  }
  hasMember(name) {
    let memberNames = ["ready_to_arm"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ReadyToArmArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ReadyToArmWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ReadyToArmWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadyToArmArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ReadyToArmArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ReadyToArmRefArray(size);
    this._refObject = new ReadyToArmRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ReadyToArmWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ReadyToArmArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ReadyToArmArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ReadyToArmWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ReadyToArmWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ReadyToArmArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(ReadyToArmWrapper, "NOT_READY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ReadyToArmWrapper, "READY_TO_ARM", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ReadyToArmWrapper, "NOT_READY_NO_TRAJ", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ReadyToArmWrapper, "NOT_READY_NO_OSEM", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(ReadyToArmWrapper, "UNAVALIABLE", {value: 255, writable: false, enumerable: true, configurable: true});
module.exports = ReadyToArmWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "NOT_READY",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "READY_TO_ARM",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "NOT_READY_NO_TRAJ",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "NOT_READY_NO_OSEM",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "UNAVALIABLE",
      "value": 255
    }
  ],
  "fields": [
    {
      "name": "ready_to_arm",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ReadyToArm",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ReadyToArm"
}
*/
