// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ActionConfiguration.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ActionConfigurationRefStruct = StructType({
  trigger_id: primitiveTypes.uint16,
  action_id: primitiveTypes.uint16,
  action_type: primitiveTypes.uint8,
  action_sign: primitiveTypes.uint16,
  action_type_parameter1: primitiveTypes.uint32,
  action_type_parameter2: primitiveTypes.uint32,
  action_type_parameter3: primitiveTypes.uint32,
  action_type_parameter4: primitiveTypes.uint32,
  action_type_parameter5: primitiveTypes.uint32,
  ip: primitiveTypes.uint32,
});
const ActionConfigurationRefArray = ArrayType(ActionConfigurationRefStruct);
const ActionConfigurationRefStructArray = StructType({
  data: ActionConfigurationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ActionConfigurationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ActionConfigurationRefStruct(other._refObject.toObject());
      this._trigger_idIntialized = true;
      this._action_idIntialized = true;
      this._action_typeIntialized = true;
      this._action_signIntialized = true;
      this._action_type_parameter1Intialized = true;
      this._action_type_parameter2Intialized = true;
      this._action_type_parameter3Intialized = true;
      this._action_type_parameter4Intialized = true;
      this._action_type_parameter5Intialized = true;
      this._ipIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ActionConfigurationRefStruct();
    this._trigger_idIntialized = false;
    this._action_idIntialized = false;
    this._action_typeIntialized = false;
    this._action_signIntialized = false;
    this._action_type_parameter1Intialized = false;
    this._action_type_parameter2Intialized = false;
    this._action_type_parameter3Intialized = false;
    this._action_type_parameter4Intialized = false;
    this._action_type_parameter5Intialized = false;
    this._ipIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ActionConfigurationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ActionConfigurationArrayWrapper;
  }
  static get ArrayType() {
    return ActionConfigurationArrayWrapper;
  }
  static get refObjectArrayType() {
    return ActionConfigurationRefStructArray
  }
  static get refObjectType() {
    return ActionConfigurationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._trigger_idIntialized) {
        throw new TypeError('Invalid argument: trigger_id in ActionConfiguration');
      }
      if (!this._action_idIntialized) {
        throw new TypeError('Invalid argument: action_id in ActionConfiguration');
      }
      if (!this._action_typeIntialized) {
        throw new TypeError('Invalid argument: action_type in ActionConfiguration');
      }
      if (!this._action_signIntialized) {
        throw new TypeError('Invalid argument: action_sign in ActionConfiguration');
      }
      if (!this._action_type_parameter1Intialized) {
        throw new TypeError('Invalid argument: action_type_parameter1 in ActionConfiguration');
      }
      if (!this._action_type_parameter2Intialized) {
        throw new TypeError('Invalid argument: action_type_parameter2 in ActionConfiguration');
      }
      if (!this._action_type_parameter3Intialized) {
        throw new TypeError('Invalid argument: action_type_parameter3 in ActionConfiguration');
      }
      if (!this._action_type_parameter4Intialized) {
        throw new TypeError('Invalid argument: action_type_parameter4 in ActionConfiguration');
      }
      if (!this._action_type_parameter5Intialized) {
        throw new TypeError('Invalid argument: action_type_parameter5 in ActionConfiguration');
      }
      if (!this._ipIntialized) {
        throw new TypeError('Invalid argument: ip in ActionConfiguration');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = refObject.trigger_id;
    this._action_idIntialized = true;
    this._refObject.action_id = refObject.action_id;
    this._action_typeIntialized = true;
    this._refObject.action_type = refObject.action_type;
    this._action_signIntialized = true;
    this._refObject.action_sign = refObject.action_sign;
    this._action_type_parameter1Intialized = true;
    this._refObject.action_type_parameter1 = refObject.action_type_parameter1;
    this._action_type_parameter2Intialized = true;
    this._refObject.action_type_parameter2 = refObject.action_type_parameter2;
    this._action_type_parameter3Intialized = true;
    this._refObject.action_type_parameter3 = refObject.action_type_parameter3;
    this._action_type_parameter4Intialized = true;
    this._refObject.action_type_parameter4 = refObject.action_type_parameter4;
    this._action_type_parameter5Intialized = true;
    this._refObject.action_type_parameter5 = refObject.action_type_parameter5;
    this._ipIntialized = true;
    this._refObject.ip = refObject.ip;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ActionConfigurationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ActionConfiguration'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trigger_id() {
    if (!this._trigger_idIntialized) {
      return undefined;
    }
    return this._refObject.trigger_id;
  }
  set trigger_id(value) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = value;
  }
  get action_id() {
    if (!this._action_idIntialized) {
      return undefined;
    }
    return this._refObject.action_id;
  }
  set action_id(value) {
    this._action_idIntialized = true;
    this._refObject.action_id = value;
  }
  get action_type() {
    if (!this._action_typeIntialized) {
      return undefined;
    }
    return this._refObject.action_type;
  }
  set action_type(value) {
    this._action_typeIntialized = true;
    this._refObject.action_type = value;
  }
  get action_sign() {
    if (!this._action_signIntialized) {
      return undefined;
    }
    return this._refObject.action_sign;
  }
  set action_sign(value) {
    this._action_signIntialized = true;
    this._refObject.action_sign = value;
  }
  get action_type_parameter1() {
    if (!this._action_type_parameter1Intialized) {
      return undefined;
    }
    return this._refObject.action_type_parameter1;
  }
  set action_type_parameter1(value) {
    this._action_type_parameter1Intialized = true;
    this._refObject.action_type_parameter1 = value;
  }
  get action_type_parameter2() {
    if (!this._action_type_parameter2Intialized) {
      return undefined;
    }
    return this._refObject.action_type_parameter2;
  }
  set action_type_parameter2(value) {
    this._action_type_parameter2Intialized = true;
    this._refObject.action_type_parameter2 = value;
  }
  get action_type_parameter3() {
    if (!this._action_type_parameter3Intialized) {
      return undefined;
    }
    return this._refObject.action_type_parameter3;
  }
  set action_type_parameter3(value) {
    this._action_type_parameter3Intialized = true;
    this._refObject.action_type_parameter3 = value;
  }
  get action_type_parameter4() {
    if (!this._action_type_parameter4Intialized) {
      return undefined;
    }
    return this._refObject.action_type_parameter4;
  }
  set action_type_parameter4(value) {
    this._action_type_parameter4Intialized = true;
    this._refObject.action_type_parameter4 = value;
  }
  get action_type_parameter5() {
    if (!this._action_type_parameter5Intialized) {
      return undefined;
    }
    return this._refObject.action_type_parameter5;
  }
  set action_type_parameter5(value) {
    this._action_type_parameter5Intialized = true;
    this._refObject.action_type_parameter5 = value;
  }
  get ip() {
    if (!this._ipIntialized) {
      return undefined;
    }
    return this._refObject.ip;
  }
  set ip(value) {
    this._ipIntialized = true;
    this._refObject.ip = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ActionConfigurationRefStruct(refObject.toObject());
    this._trigger_idIntialized = true;
    this._action_idIntialized = true;
    this._action_typeIntialized = true;
    this._action_signIntialized = true;
    this._action_type_parameter1Intialized = true;
    this._action_type_parameter2Intialized = true;
    this._action_type_parameter3Intialized = true;
    this._action_type_parameter4Intialized = true;
    this._action_type_parameter5Intialized = true;
    this._ipIntialized = true;
  }
  copy(other) {
    this._refObject = new ActionConfigurationRefStruct(other._refObject.toObject());
    this._trigger_idIntialized = true;
    this._action_idIntialized = true;
    this._action_typeIntialized = true;
    this._action_signIntialized = true;
    this._action_type_parameter1Intialized = true;
    this._action_type_parameter2Intialized = true;
    this._action_type_parameter3Intialized = true;
    this._action_type_parameter4Intialized = true;
    this._action_type_parameter5Intialized = true;
    this._ipIntialized = true;
  }
  static get classType() {
    return ActionConfigurationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trigger_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_sign","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type_parameter1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type_parameter2","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type_parameter3","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type_parameter4","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"action_type_parameter5","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ActionConfiguration","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ActionConfiguration"};
  }
  hasMember(name) {
    let memberNames = ["trigger_id","action_id","action_type","action_sign","action_type_parameter1","action_type_parameter2","action_type_parameter3","action_type_parameter4","action_type_parameter5","ip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ActionConfigurationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ActionConfigurationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ActionConfigurationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActionConfigurationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ActionConfigurationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ActionConfigurationRefArray(size);
    this._refObject = new ActionConfigurationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ActionConfigurationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ActionConfigurationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ActionConfigurationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ActionConfigurationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ActionConfigurationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ActionConfigurationArrayWrapper;
  }
}
module.exports = ActionConfigurationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trigger_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_sign",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type_parameter1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type_parameter2",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type_parameter3",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type_parameter4",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "action_type_parameter5",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ActionConfiguration",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ActionConfiguration"
}
*/
