// rclnodejs/generated/atos_interfaces/atos_interfaces__srv__GetObjectIp_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetObjectIp_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  id: primitiveTypes.uint32,
  ip: primitiveTypes.string,
});
const GetObjectIp_ResponseRefArray = ArrayType(GetObjectIp_ResponseRefStruct);
const GetObjectIp_ResponseRefStructArray = StructType({
  data: GetObjectIp_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetObjectIp_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetObjectIp_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._idIntialized = true;
      this._ipIntialized = true;
      this._wrapperFields.ip =  new StringWrapper(other._wrapperFields.ip);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetObjectIp_ResponseRefStruct();
    this._refObject.success = true;
    this._successIntialized = true;
    this._idIntialized = false;
    this._ipIntialized = false;
    this._wrapperFields.ip = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetObjectIp_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetObjectIp_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetObjectIp_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetObjectIp_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetObjectIp_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetObjectIp_Response');
      }
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in GetObjectIp_Response');
      }
      if (!this._ipIntialized) {
        throw new TypeError('Invalid argument: ip in GetObjectIp_Response');
      }
    }
    if (own) {
      this._wrapperFields.ip.freeze(own, checkConsistency);
    }
    this._refObject.ip = this._wrapperFields.ip.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._idIntialized = true;
    this._refObject.id = refObject.id;
    this._ipIntialized = true;
    this._wrapperFields.ip.data = refObject.ip.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.ip);
  }
  static destoryRawROS(msg) {
    GetObjectIp_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'srv', interfaceName: 'GetObjectIp_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._refObject.id;
  }
  set id(value) {
    this._idIntialized = true;
    this._refObject.id = value;
  }
  get ip() {
    if (!this._ipIntialized) {
      return undefined;
    }
    return this._wrapperFields.ip.data;
  }
  set ip(value) {
    this._ipIntialized = true;
    this._wrapperFields.ip.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetObjectIp_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._idIntialized = true;
    this._ipIntialized = true;
    this._wrapperFields.ip.copyRefObject(this._refObject.ip);
  }
  copy(other) {
    this._refObject = new GetObjectIp_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._idIntialized = true;
    this._ipIntialized = true;
    this._wrapperFields.ip.copy(other._wrapperFields.ip);
  }
  static get classType() {
    return GetObjectIp_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":true},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"GetObjectIp_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetObjectIp_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","id","ip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetObjectIp_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetObjectIp_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetObjectIp_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectIp_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetObjectIp_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetObjectIp_ResponseRefArray(size);
    this._refObject = new GetObjectIp_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetObjectIp_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetObjectIp_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetObjectIp_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetObjectIp_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetObjectIp_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetObjectIp_ResponseArrayWrapper;
  }
}
module.exports = GetObjectIp_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": true
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "GetObjectIp_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetObjectIp_Response"
}
*/
