// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__TriggerEventOccurred.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TriggerEventOccurredRefStruct = StructType({
  trigger_id: primitiveTypes.uint16,
  timestamp_qmsow: primitiveTypes.uint32,
  ip: primitiveTypes.uint32,
});
const TriggerEventOccurredRefArray = ArrayType(TriggerEventOccurredRefStruct);
const TriggerEventOccurredRefStructArray = StructType({
  data: TriggerEventOccurredRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TriggerEventOccurredWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TriggerEventOccurredRefStruct(other._refObject.toObject());
      this._trigger_idIntialized = true;
      this._timestamp_qmsowIntialized = true;
      this._ipIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TriggerEventOccurredRefStruct();
    this._trigger_idIntialized = false;
    this._timestamp_qmsowIntialized = false;
    this._ipIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TriggerEventOccurredWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TriggerEventOccurredArrayWrapper;
  }
  static get ArrayType() {
    return TriggerEventOccurredArrayWrapper;
  }
  static get refObjectArrayType() {
    return TriggerEventOccurredRefStructArray
  }
  static get refObjectType() {
    return TriggerEventOccurredRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._trigger_idIntialized) {
        throw new TypeError('Invalid argument: trigger_id in TriggerEventOccurred');
      }
      if (!this._timestamp_qmsowIntialized) {
        throw new TypeError('Invalid argument: timestamp_qmsow in TriggerEventOccurred');
      }
      if (!this._ipIntialized) {
        throw new TypeError('Invalid argument: ip in TriggerEventOccurred');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = refObject.trigger_id;
    this._timestamp_qmsowIntialized = true;
    this._refObject.timestamp_qmsow = refObject.timestamp_qmsow;
    this._ipIntialized = true;
    this._refObject.ip = refObject.ip;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TriggerEventOccurredWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'TriggerEventOccurred'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get trigger_id() {
    if (!this._trigger_idIntialized) {
      return undefined;
    }
    return this._refObject.trigger_id;
  }
  set trigger_id(value) {
    this._trigger_idIntialized = true;
    this._refObject.trigger_id = value;
  }
  get timestamp_qmsow() {
    if (!this._timestamp_qmsowIntialized) {
      return undefined;
    }
    return this._refObject.timestamp_qmsow;
  }
  set timestamp_qmsow(value) {
    this._timestamp_qmsowIntialized = true;
    this._refObject.timestamp_qmsow = value;
  }
  get ip() {
    if (!this._ipIntialized) {
      return undefined;
    }
    return this._refObject.ip;
  }
  set ip(value) {
    this._ipIntialized = true;
    this._refObject.ip = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TriggerEventOccurredRefStruct(refObject.toObject());
    this._trigger_idIntialized = true;
    this._timestamp_qmsowIntialized = true;
    this._ipIntialized = true;
  }
  copy(other) {
    this._refObject = new TriggerEventOccurredRefStruct(other._refObject.toObject());
    this._trigger_idIntialized = true;
    this._timestamp_qmsowIntialized = true;
    this._ipIntialized = true;
  }
  static get classType() {
    return TriggerEventOccurredWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"trigger_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"timestamp_qmsow","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ip","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"TriggerEventOccurred","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TriggerEventOccurred"};
  }
  hasMember(name) {
    let memberNames = ["trigger_id","timestamp_qmsow","ip"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TriggerEventOccurredArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TriggerEventOccurredWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TriggerEventOccurredWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggerEventOccurredArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TriggerEventOccurredArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TriggerEventOccurredRefArray(size);
    this._refObject = new TriggerEventOccurredRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TriggerEventOccurredWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TriggerEventOccurredArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TriggerEventOccurredArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TriggerEventOccurredWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TriggerEventOccurredWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TriggerEventOccurredArrayWrapper;
  }
}
module.exports = TriggerEventOccurredWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "trigger_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "timestamp_qmsow",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ip",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "TriggerEventOccurred",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TriggerEventOccurred"
}
*/
