// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ObjectIdArray.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const ObjectIdArrayRefStruct = StructType({
  ids: UInt32Wrapper.refObjectArrayType,
});
const ObjectIdArrayRefArray = ArrayType(ObjectIdArrayRefStruct);
const ObjectIdArrayRefStructArray = StructType({
  data: ObjectIdArrayRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectIdArrayWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectIdArrayRefStruct(other._refObject.toObject());
      this._wrapperFields.ids = UInt32Wrapper.createArray();
      this._wrapperFields.ids.copy(other._wrapperFields.ids);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectIdArrayRefStruct();
    this._wrapperFields.ids = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ObjectIdArrayWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectIdArrayArrayWrapper;
  }
  static get ArrayType() {
    return ObjectIdArrayArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectIdArrayRefStructArray
  }
  static get refObjectType() {
    return ObjectIdArrayRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.ids.freeze(own, checkConsistency);
    this._refObject.ids = this._wrapperFields.ids.refObject;
    if (own) {
      this._wrapperFields.ids.fill(Uint32Array.from([]));
      this._wrapperFields.ids.freeze(own, checkConsistency);
      this._refObject.ids = this._wrapperFields.ids.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ids.copyRefObject(refObject.ids);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ids.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.ids);
      if (UInt32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ids, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ObjectIdArrayWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ObjectIdArray'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ids() {
    return this._wrapperFields['ids'].data;
  }
  set ids(value) {
    this._wrapperFields['ids'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectIdArrayRefStruct(refObject.toObject());
    this._wrapperFields.ids.copyRefObject(this._refObject.ids);
  }
  copy(other) {
    this._refObject = new ObjectIdArrayRefStruct(other._refObject.toObject());
    this._wrapperFields.ids.copy(other._wrapperFields.ids);
  }
  static get classType() {
    return ObjectIdArrayWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ObjectIdArray","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectIdArray"};
  }
  hasMember(name) {
    let memberNames = ["ids"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectIdArrayArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectIdArrayWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectIdArrayWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectIdArrayArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectIdArrayArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectIdArrayRefArray(size);
    this._refObject = new ObjectIdArrayRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectIdArrayWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectIdArrayArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectIdArrayArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectIdArrayWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectIdArrayWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectIdArrayArrayWrapper;
  }
}
module.exports = ObjectIdArrayWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ObjectIdArray",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectIdArray"
}
*/
