// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__CartesianTolerance.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const AccelWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Accel.js');
const CartesianToleranceRefStruct = StructType({
  position_error: Vector3Wrapper.refObjectType,
  orientation_error: Vector3Wrapper.refObjectType,
  twist_error: TwistWrapper.refObjectType,
  acceleration_error: AccelWrapper.refObjectType,
});
const CartesianToleranceRefArray = ArrayType(CartesianToleranceRefStruct);
const CartesianToleranceRefStructArray = StructType({
  data: CartesianToleranceRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CartesianToleranceWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CartesianToleranceRefStruct(other._refObject.toObject());
      this._wrapperFields.position_error =  new Vector3Wrapper(other._wrapperFields.position_error);
      this._wrapperFields.orientation_error =  new Vector3Wrapper(other._wrapperFields.orientation_error);
      this._wrapperFields.twist_error =  new TwistWrapper(other._wrapperFields.twist_error);
      this._wrapperFields.acceleration_error =  new AccelWrapper(other._wrapperFields.acceleration_error);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CartesianToleranceRefStruct();
    this._wrapperFields.position_error = new Vector3Wrapper();
    this._wrapperFields.orientation_error = new Vector3Wrapper();
    this._wrapperFields.twist_error = new TwistWrapper();
    this._wrapperFields.acceleration_error = new AccelWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new CartesianToleranceWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CartesianToleranceArrayWrapper;
  }
  static get ArrayType() {
    return CartesianToleranceArrayWrapper;
  }
  static get refObjectArrayType() {
    return CartesianToleranceRefStructArray
  }
  static get refObjectType() {
    return CartesianToleranceRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.position_error.freeze(own, checkConsistency);
    this._refObject.position_error = this._wrapperFields.position_error.refObject;
    this._wrapperFields.orientation_error.freeze(own, checkConsistency);
    this._refObject.orientation_error = this._wrapperFields.orientation_error.refObject;
    this._wrapperFields.twist_error.freeze(own, checkConsistency);
    this._refObject.twist_error = this._wrapperFields.twist_error.refObject;
    this._wrapperFields.acceleration_error.freeze(own, checkConsistency);
    this._refObject.acceleration_error = this._wrapperFields.acceleration_error.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.position_error.copyRefObject(refObject.position_error);
    this._wrapperFields.orientation_error.copyRefObject(refObject.orientation_error);
    this._wrapperFields.twist_error.copyRefObject(refObject.twist_error);
    this._wrapperFields.acceleration_error.copyRefObject(refObject.acceleration_error);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.position_error);
    Vector3Wrapper.freeStruct(refObject.orientation_error);
    TwistWrapper.freeStruct(refObject.twist_error);
    AccelWrapper.freeStruct(refObject.acceleration_error);
  }
  static destoryRawROS(msg) {
    CartesianToleranceWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'CartesianTolerance'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get position_error() {
    return this._wrapperFields.position_error;
  }
  set position_error(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.position_error.copy(value);
    } else {
      this._wrapperFields.position_error.copy(new Vector3Wrapper(value));
    }
  }
  get orientation_error() {
    return this._wrapperFields.orientation_error;
  }
  set orientation_error(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.orientation_error.copy(value);
    } else {
      this._wrapperFields.orientation_error.copy(new Vector3Wrapper(value));
    }
  }
  get twist_error() {
    return this._wrapperFields.twist_error;
  }
  set twist_error(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.twist_error.copy(value);
    } else {
      this._wrapperFields.twist_error.copy(new TwistWrapper(value));
    }
  }
  get acceleration_error() {
    return this._wrapperFields.acceleration_error;
  }
  set acceleration_error(value) {
    if (value instanceof AccelWrapper) {
      this._wrapperFields.acceleration_error.copy(value);
    } else {
      this._wrapperFields.acceleration_error.copy(new AccelWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new CartesianToleranceRefStruct(refObject.toObject());
    this._wrapperFields.position_error.copyRefObject(this._refObject.position_error);
    this._wrapperFields.orientation_error.copyRefObject(this._refObject.orientation_error);
    this._wrapperFields.twist_error.copyRefObject(this._refObject.twist_error);
    this._wrapperFields.acceleration_error.copyRefObject(this._refObject.acceleration_error);
  }
  copy(other) {
    this._refObject = new CartesianToleranceRefStruct(other._refObject.toObject());
    this._wrapperFields.position_error.copy(other._wrapperFields.position_error);
    this._wrapperFields.orientation_error.copy(other._wrapperFields.orientation_error);
    this._wrapperFields.twist_error.copy(other._wrapperFields.twist_error);
    this._wrapperFields.acceleration_error.copy(other._wrapperFields.acceleration_error);
  }
  static get classType() {
    return CartesianToleranceWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"position_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"orientation_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twist_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"acceleration_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Accel","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"CartesianTolerance","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CartesianTolerance"};
  }
  hasMember(name) {
    let memberNames = ["position_error","orientation_error","twist_error","acceleration_error"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CartesianToleranceArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CartesianToleranceWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CartesianToleranceWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianToleranceArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CartesianToleranceArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CartesianToleranceRefArray(size);
    this._refObject = new CartesianToleranceRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CartesianToleranceWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CartesianToleranceArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CartesianToleranceArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CartesianToleranceWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CartesianToleranceWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CartesianToleranceArrayWrapper;
  }
}
module.exports = CartesianToleranceWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "position_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "orientation_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twist_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "acceleration_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Accel",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "CartesianTolerance",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CartesianTolerance"
}
*/
