// rclnodejs/generated/atos_interfaces/atos_interfaces__msg__ObjectErrorType.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ObjectErrorTypeRefStruct = StructType({
  abort_request: primitiveTypes.bool,
  outside_geofence: primitiveTypes.bool,
  bad_positioning_accuracy: primitiveTypes.bool,
  engine_fault: primitiveTypes.bool,
  battery_fault: primitiveTypes.bool,
  sync_point_ended: primitiveTypes.bool,
  unknown_error: primitiveTypes.bool,
});
const ObjectErrorTypeRefArray = ArrayType(ObjectErrorTypeRefStruct);
const ObjectErrorTypeRefStructArray = StructType({
  data: ObjectErrorTypeRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ObjectErrorTypeWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ObjectErrorTypeRefStruct(other._refObject.toObject());
      this._abort_requestIntialized = true;
      this._outside_geofenceIntialized = true;
      this._bad_positioning_accuracyIntialized = true;
      this._engine_faultIntialized = true;
      this._battery_faultIntialized = true;
      this._sync_point_endedIntialized = true;
      this._unknown_errorIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ObjectErrorTypeRefStruct();
    this._abort_requestIntialized = false;
    this._outside_geofenceIntialized = false;
    this._bad_positioning_accuracyIntialized = false;
    this._engine_faultIntialized = false;
    this._battery_faultIntialized = false;
    this._sync_point_endedIntialized = false;
    this._unknown_errorIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ObjectErrorTypeWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ObjectErrorTypeArrayWrapper;
  }
  static get ArrayType() {
    return ObjectErrorTypeArrayWrapper;
  }
  static get refObjectArrayType() {
    return ObjectErrorTypeRefStructArray
  }
  static get refObjectType() {
    return ObjectErrorTypeRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._abort_requestIntialized) {
        throw new TypeError('Invalid argument: abort_request in ObjectErrorType');
      }
      if (!this._outside_geofenceIntialized) {
        throw new TypeError('Invalid argument: outside_geofence in ObjectErrorType');
      }
      if (!this._bad_positioning_accuracyIntialized) {
        throw new TypeError('Invalid argument: bad_positioning_accuracy in ObjectErrorType');
      }
      if (!this._engine_faultIntialized) {
        throw new TypeError('Invalid argument: engine_fault in ObjectErrorType');
      }
      if (!this._battery_faultIntialized) {
        throw new TypeError('Invalid argument: battery_fault in ObjectErrorType');
      }
      if (!this._sync_point_endedIntialized) {
        throw new TypeError('Invalid argument: sync_point_ended in ObjectErrorType');
      }
      if (!this._unknown_errorIntialized) {
        throw new TypeError('Invalid argument: unknown_error in ObjectErrorType');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._abort_requestIntialized = true;
    this._refObject.abort_request = refObject.abort_request;
    this._outside_geofenceIntialized = true;
    this._refObject.outside_geofence = refObject.outside_geofence;
    this._bad_positioning_accuracyIntialized = true;
    this._refObject.bad_positioning_accuracy = refObject.bad_positioning_accuracy;
    this._engine_faultIntialized = true;
    this._refObject.engine_fault = refObject.engine_fault;
    this._battery_faultIntialized = true;
    this._refObject.battery_fault = refObject.battery_fault;
    this._sync_point_endedIntialized = true;
    this._refObject.sync_point_ended = refObject.sync_point_ended;
    this._unknown_errorIntialized = true;
    this._refObject.unknown_error = refObject.unknown_error;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    ObjectErrorTypeWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'atos_interfaces', subFolder: 'msg', interfaceName: 'ObjectErrorType'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get abort_request() {
    if (!this._abort_requestIntialized) {
      return undefined;
    }
    return this._refObject.abort_request;
  }
  set abort_request(value) {
    this._abort_requestIntialized = true;
    this._refObject.abort_request = value;
  }
  get outside_geofence() {
    if (!this._outside_geofenceIntialized) {
      return undefined;
    }
    return this._refObject.outside_geofence;
  }
  set outside_geofence(value) {
    this._outside_geofenceIntialized = true;
    this._refObject.outside_geofence = value;
  }
  get bad_positioning_accuracy() {
    if (!this._bad_positioning_accuracyIntialized) {
      return undefined;
    }
    return this._refObject.bad_positioning_accuracy;
  }
  set bad_positioning_accuracy(value) {
    this._bad_positioning_accuracyIntialized = true;
    this._refObject.bad_positioning_accuracy = value;
  }
  get engine_fault() {
    if (!this._engine_faultIntialized) {
      return undefined;
    }
    return this._refObject.engine_fault;
  }
  set engine_fault(value) {
    this._engine_faultIntialized = true;
    this._refObject.engine_fault = value;
  }
  get battery_fault() {
    if (!this._battery_faultIntialized) {
      return undefined;
    }
    return this._refObject.battery_fault;
  }
  set battery_fault(value) {
    this._battery_faultIntialized = true;
    this._refObject.battery_fault = value;
  }
  get sync_point_ended() {
    if (!this._sync_point_endedIntialized) {
      return undefined;
    }
    return this._refObject.sync_point_ended;
  }
  set sync_point_ended(value) {
    this._sync_point_endedIntialized = true;
    this._refObject.sync_point_ended = value;
  }
  get unknown_error() {
    if (!this._unknown_errorIntialized) {
      return undefined;
    }
    return this._refObject.unknown_error;
  }
  set unknown_error(value) {
    this._unknown_errorIntialized = true;
    this._refObject.unknown_error = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ObjectErrorTypeRefStruct(refObject.toObject());
    this._abort_requestIntialized = true;
    this._outside_geofenceIntialized = true;
    this._bad_positioning_accuracyIntialized = true;
    this._engine_faultIntialized = true;
    this._battery_faultIntialized = true;
    this._sync_point_endedIntialized = true;
    this._unknown_errorIntialized = true;
  }
  copy(other) {
    this._refObject = new ObjectErrorTypeRefStruct(other._refObject.toObject());
    this._abort_requestIntialized = true;
    this._outside_geofenceIntialized = true;
    this._bad_positioning_accuracyIntialized = true;
    this._engine_faultIntialized = true;
    this._battery_faultIntialized = true;
    this._sync_point_endedIntialized = true;
    this._unknown_errorIntialized = true;
  }
  static get classType() {
    return ObjectErrorTypeWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"abort_request","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"outside_geofence","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"bad_positioning_accuracy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"engine_fault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"battery_fault","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sync_point_ended","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"unknown_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"atos_interfaces","type":"ObjectErrorType","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ObjectErrorType"};
  }
  hasMember(name) {
    let memberNames = ["abort_request","outside_geofence","bad_positioning_accuracy","engine_fault","battery_fault","sync_point_ended","unknown_error"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ObjectErrorTypeArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ObjectErrorTypeWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ObjectErrorTypeWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectErrorTypeArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ObjectErrorTypeArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ObjectErrorTypeRefArray(size);
    this._refObject = new ObjectErrorTypeRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ObjectErrorTypeWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ObjectErrorTypeArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ObjectErrorTypeArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ObjectErrorTypeWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ObjectErrorTypeWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ObjectErrorTypeArrayWrapper;
  }
}
module.exports = ObjectErrorTypeWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "abort_request",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "outside_geofence",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "bad_positioning_accuracy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "engine_fault",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "battery_fault",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sync_point_ended",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "unknown_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "atos_interfaces",
    "type": "ObjectErrorType",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ObjectErrorType"
}
*/
