// rclnodejs/generated/geographic_msgs/geographic_msgs__srv__GetRoutePlan_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const GetRoutePlan_RequestRefStruct = StructType({
  network: UUIDWrapper.refObjectType,
  start: UUIDWrapper.refObjectType,
  goal: UUIDWrapper.refObjectType,
});
const GetRoutePlan_RequestRefArray = ArrayType(GetRoutePlan_RequestRefStruct);
const GetRoutePlan_RequestRefStructArray = StructType({
  data: GetRoutePlan_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetRoutePlan_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetRoutePlan_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.network =  new UUIDWrapper(other._wrapperFields.network);
      this._wrapperFields.start =  new UUIDWrapper(other._wrapperFields.start);
      this._wrapperFields.goal =  new UUIDWrapper(other._wrapperFields.goal);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetRoutePlan_RequestRefStruct();
    this._wrapperFields.network = new UUIDWrapper();
    this._wrapperFields.start = new UUIDWrapper();
    this._wrapperFields.goal = new UUIDWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetRoutePlan_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetRoutePlan_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetRoutePlan_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetRoutePlan_RequestRefStructArray
  }
  static get refObjectType() {
    return GetRoutePlan_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.network.freeze(own, checkConsistency);
    this._refObject.network = this._wrapperFields.network.refObject;
    this._wrapperFields.start.freeze(own, checkConsistency);
    this._refObject.start = this._wrapperFields.start.refObject;
    this._wrapperFields.goal.freeze(own, checkConsistency);
    this._refObject.goal = this._wrapperFields.goal.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.network.copyRefObject(refObject.network);
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._wrapperFields.goal.copyRefObject(refObject.goal);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    UUIDWrapper.freeStruct(refObject.network);
    UUIDWrapper.freeStruct(refObject.start);
    UUIDWrapper.freeStruct(refObject.goal);
  }
  static destoryRawROS(msg) {
    GetRoutePlan_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'srv', interfaceName: 'GetRoutePlan_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get network() {
    return this._wrapperFields.network;
  }
  set network(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.network.copy(value);
    } else {
      this._wrapperFields.network.copy(new UUIDWrapper(value));
    }
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new UUIDWrapper(value));
    }
  }
  get goal() {
    return this._wrapperFields.goal;
  }
  set goal(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.goal.copy(value);
    } else {
      this._wrapperFields.goal.copy(new UUIDWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetRoutePlan_RequestRefStruct(refObject.toObject());
    this._wrapperFields.network.copyRefObject(this._refObject.network);
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._wrapperFields.goal.copyRefObject(this._refObject.goal);
  }
  copy(other) {
    this._refObject = new GetRoutePlan_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.network.copy(other._wrapperFields.network);
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._wrapperFields.goal.copy(other._wrapperFields.goal);
  }
  static get classType() {
    return GetRoutePlan_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"network","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"GetRoutePlan_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetRoutePlan_Request"};
  }
  hasMember(name) {
    let memberNames = ["network","start","goal"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetRoutePlan_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetRoutePlan_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetRoutePlan_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRoutePlan_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetRoutePlan_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetRoutePlan_RequestRefArray(size);
    this._refObject = new GetRoutePlan_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetRoutePlan_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetRoutePlan_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetRoutePlan_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetRoutePlan_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetRoutePlan_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetRoutePlan_RequestArrayWrapper;
  }
}
module.exports = GetRoutePlan_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "network",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "GetRoutePlan_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetRoutePlan_Request"
}
*/
