// rclnodejs/generated/geographic_msgs/geographic_msgs__msg__GeographicMapChanges.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const GeographicMapWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeographicMap.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const GeographicMapChangesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  diffs: GeographicMapWrapper.refObjectType,
  deletes: UUIDWrapper.refObjectArrayType,
});
const GeographicMapChangesRefArray = ArrayType(GeographicMapChangesRefStruct);
const GeographicMapChangesRefStructArray = StructType({
  data: GeographicMapChangesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GeographicMapChangesWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GeographicMapChangesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.diffs =  new GeographicMapWrapper(other._wrapperFields.diffs);
      this._wrapperFields.deletes = UUIDWrapper.createArray();
      this._wrapperFields.deletes.copy(other._wrapperFields.deletes);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GeographicMapChangesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.diffs = new GeographicMapWrapper();
    this._wrapperFields.deletes = UUIDWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GeographicMapChangesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GeographicMapChangesArrayWrapper;
  }
  static get ArrayType() {
    return GeographicMapChangesArrayWrapper;
  }
  static get refObjectArrayType() {
    return GeographicMapChangesRefStructArray
  }
  static get refObjectType() {
    return GeographicMapChangesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.diffs.freeze(own, checkConsistency);
    this._refObject.diffs = this._wrapperFields.diffs.refObject;
    this._wrapperFields.deletes.freeze(own, checkConsistency);
    this._refObject.deletes = this._wrapperFields.deletes.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.diffs.copyRefObject(refObject.diffs);
    this._wrapperFields.deletes.copyRefObject(refObject.deletes);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    GeographicMapWrapper.freeStruct(refObject.diffs);
    if (refObject.deletes.size != 0) {
      UUIDWrapper.ArrayType.freeArray(refObject.deletes);
      if (UUIDWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.deletes, UUIDWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GeographicMapChangesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'msg', interfaceName: 'GeographicMapChanges'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get diffs() {
    return this._wrapperFields.diffs;
  }
  set diffs(value) {
    if (value instanceof GeographicMapWrapper) {
      this._wrapperFields.diffs.copy(value);
    } else {
      this._wrapperFields.diffs.copy(new GeographicMapWrapper(value));
    }
  }
  get deletes() {
    return this._wrapperFields.deletes;
  }
  set deletes(value) {
    this._wrapperFields.deletes.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GeographicMapChangesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.diffs.copyRefObject(this._refObject.diffs);
    this._wrapperFields.deletes.copyRefObject(this._refObject.deletes);
  }
  copy(other) {
    this._refObject = new GeographicMapChangesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.diffs.copy(other._wrapperFields.diffs);
    this._wrapperFields.deletes.copy(other._wrapperFields.deletes);
  }
  static get classType() {
    return GeographicMapChangesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"diffs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeographicMap","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"deletes","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"GeographicMapChanges","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GeographicMapChanges"};
  }
  hasMember(name) {
    let memberNames = ["header","diffs","deletes"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GeographicMapChangesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GeographicMapChangesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GeographicMapChangesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeographicMapChangesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeographicMapChangesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GeographicMapChangesRefArray(size);
    this._refObject = new GeographicMapChangesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GeographicMapChangesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GeographicMapChangesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GeographicMapChangesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GeographicMapChangesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GeographicMapChangesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GeographicMapChangesArrayWrapper;
  }
}
module.exports = GeographicMapChangesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "diffs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeographicMap",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "deletes",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "GeographicMapChanges",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GeographicMapChanges"
}
*/
