// rclnodejs/generated/geographic_msgs/geographic_msgs__msg__MapFeature.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const KeyValueWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__KeyValue.js');
const MapFeatureRefStruct = StructType({
  id: UUIDWrapper.refObjectType,
  components: UUIDWrapper.refObjectArrayType,
  props: KeyValueWrapper.refObjectArrayType,
});
const MapFeatureRefArray = ArrayType(MapFeatureRefStruct);
const MapFeatureRefStructArray = StructType({
  data: MapFeatureRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MapFeatureWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MapFeatureRefStruct(other._refObject.toObject());
      this._wrapperFields.id =  new UUIDWrapper(other._wrapperFields.id);
      this._wrapperFields.components = UUIDWrapper.createArray();
      this._wrapperFields.components.copy(other._wrapperFields.components);
      this._wrapperFields.props = KeyValueWrapper.createArray();
      this._wrapperFields.props.copy(other._wrapperFields.props);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MapFeatureRefStruct();
    this._wrapperFields.id = new UUIDWrapper();
    this._wrapperFields.components = UUIDWrapper.createArray();
    this._wrapperFields.props = KeyValueWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MapFeatureWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MapFeatureArrayWrapper;
  }
  static get ArrayType() {
    return MapFeatureArrayWrapper;
  }
  static get refObjectArrayType() {
    return MapFeatureRefStructArray
  }
  static get refObjectType() {
    return MapFeatureRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.id.freeze(own, checkConsistency);
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.components.freeze(own, checkConsistency);
    this._refObject.components = this._wrapperFields.components.refObject;
    this._wrapperFields.props.freeze(own, checkConsistency);
    this._refObject.props = this._wrapperFields.props.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.id.copyRefObject(refObject.id);
    this._wrapperFields.components.copyRefObject(refObject.components);
    this._wrapperFields.props.copyRefObject(refObject.props);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    UUIDWrapper.freeStruct(refObject.id);
    if (refObject.components.size != 0) {
      UUIDWrapper.ArrayType.freeArray(refObject.components);
      if (UUIDWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.components, UUIDWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.props.size != 0) {
      KeyValueWrapper.ArrayType.freeArray(refObject.props);
      if (KeyValueWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.props, KeyValueWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MapFeatureWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'msg', interfaceName: 'MapFeature'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get id() {
    return this._wrapperFields.id;
  }
  set id(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.id.copy(value);
    } else {
      this._wrapperFields.id.copy(new UUIDWrapper(value));
    }
  }
  get components() {
    return this._wrapperFields.components;
  }
  set components(value) {
    this._wrapperFields.components.fill(value);
  }
  get props() {
    return this._wrapperFields.props;
  }
  set props(value) {
    this._wrapperFields.props.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MapFeatureRefStruct(refObject.toObject());
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._wrapperFields.components.copyRefObject(this._refObject.components);
    this._wrapperFields.props.copyRefObject(this._refObject.props);
  }
  copy(other) {
    this._refObject = new MapFeatureRefStruct(other._refObject.toObject());
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._wrapperFields.components.copy(other._wrapperFields.components);
    this._wrapperFields.props.copy(other._wrapperFields.props);
  }
  static get classType() {
    return MapFeatureWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"components","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"props","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geographic_msgs","type":"KeyValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"MapFeature","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MapFeature"};
  }
  hasMember(name) {
    let memberNames = ["id","components","props"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MapFeatureArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MapFeatureWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MapFeatureWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapFeatureArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MapFeatureArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MapFeatureRefArray(size);
    this._refObject = new MapFeatureRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MapFeatureWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MapFeatureArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MapFeatureArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MapFeatureWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MapFeatureWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MapFeatureArrayWrapper;
  }
}
module.exports = MapFeatureWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "components",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "props",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geographic_msgs",
        "type": "KeyValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "MapFeature",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MapFeature"
}
*/
