// rclnodejs/generated/geographic_msgs/geographic_msgs__srv__GetGeoPath_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GeoPathWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPath.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const GetGeoPath_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  status: primitiveTypes.string,
  plan: GeoPathWrapper.refObjectType,
  network: UUIDWrapper.refObjectType,
  start_seg: UUIDWrapper.refObjectType,
  goal_seg: UUIDWrapper.refObjectType,
  distance: primitiveTypes.float64,
});
const GetGeoPath_ResponseRefArray = ArrayType(GetGeoPath_ResponseRefStruct);
const GetGeoPath_ResponseRefStructArray = StructType({
  data: GetGeoPath_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetGeoPath_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetGeoPath_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._statusIntialized = true;
      this._wrapperFields.status =  new StringWrapper(other._wrapperFields.status);
      this._wrapperFields.plan =  new GeoPathWrapper(other._wrapperFields.plan);
      this._wrapperFields.network =  new UUIDWrapper(other._wrapperFields.network);
      this._wrapperFields.start_seg =  new UUIDWrapper(other._wrapperFields.start_seg);
      this._wrapperFields.goal_seg =  new UUIDWrapper(other._wrapperFields.goal_seg);
      this._distanceIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetGeoPath_ResponseRefStruct();
    this._successIntialized = false;
    this._statusIntialized = false;
    this._wrapperFields.status = new StringWrapper();
    this._wrapperFields.plan = new GeoPathWrapper();
    this._wrapperFields.network = new UUIDWrapper();
    this._wrapperFields.start_seg = new UUIDWrapper();
    this._wrapperFields.goal_seg = new UUIDWrapper();
    this._distanceIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GetGeoPath_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetGeoPath_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetGeoPath_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetGeoPath_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetGeoPath_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetGeoPath_Response');
      }
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in GetGeoPath_Response');
      }
      if (!this._distanceIntialized) {
        throw new TypeError('Invalid argument: distance in GetGeoPath_Response');
      }
    }
    if (own) {
      this._wrapperFields.status.freeze(own, checkConsistency);
    }
    this._refObject.status = this._wrapperFields.status.refObject;
    this._wrapperFields.plan.freeze(own, checkConsistency);
    this._refObject.plan = this._wrapperFields.plan.refObject;
    this._wrapperFields.network.freeze(own, checkConsistency);
    this._refObject.network = this._wrapperFields.network.refObject;
    this._wrapperFields.start_seg.freeze(own, checkConsistency);
    this._refObject.start_seg = this._wrapperFields.start_seg.refObject;
    this._wrapperFields.goal_seg.freeze(own, checkConsistency);
    this._refObject.goal_seg = this._wrapperFields.goal_seg.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._statusIntialized = true;
    this._wrapperFields.status.data = refObject.status.data;
    this._wrapperFields.plan.copyRefObject(refObject.plan);
    this._wrapperFields.network.copyRefObject(refObject.network);
    this._wrapperFields.start_seg.copyRefObject(refObject.start_seg);
    this._wrapperFields.goal_seg.copyRefObject(refObject.goal_seg);
    this._distanceIntialized = true;
    this._refObject.distance = refObject.distance;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.status);
    GeoPathWrapper.freeStruct(refObject.plan);
    UUIDWrapper.freeStruct(refObject.network);
    UUIDWrapper.freeStruct(refObject.start_seg);
    UUIDWrapper.freeStruct(refObject.goal_seg);
  }
  static destoryRawROS(msg) {
    GetGeoPath_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'srv', interfaceName: 'GetGeoPath_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._wrapperFields.status.data;
  }
  set status(value) {
    this._statusIntialized = true;
    this._wrapperFields.status.data = value;
  }
  get plan() {
    return this._wrapperFields.plan;
  }
  set plan(value) {
    if (value instanceof GeoPathWrapper) {
      this._wrapperFields.plan.copy(value);
    } else {
      this._wrapperFields.plan.copy(new GeoPathWrapper(value));
    }
  }
  get network() {
    return this._wrapperFields.network;
  }
  set network(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.network.copy(value);
    } else {
      this._wrapperFields.network.copy(new UUIDWrapper(value));
    }
  }
  get start_seg() {
    return this._wrapperFields.start_seg;
  }
  set start_seg(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.start_seg.copy(value);
    } else {
      this._wrapperFields.start_seg.copy(new UUIDWrapper(value));
    }
  }
  get goal_seg() {
    return this._wrapperFields.goal_seg;
  }
  set goal_seg(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.goal_seg.copy(value);
    } else {
      this._wrapperFields.goal_seg.copy(new UUIDWrapper(value));
    }
  }
  get distance() {
    if (!this._distanceIntialized) {
      return undefined;
    }
    return this._refObject.distance;
  }
  set distance(value) {
    this._distanceIntialized = true;
    this._refObject.distance = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetGeoPath_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._statusIntialized = true;
    this._wrapperFields.status.copyRefObject(this._refObject.status);
    this._wrapperFields.plan.copyRefObject(this._refObject.plan);
    this._wrapperFields.network.copyRefObject(this._refObject.network);
    this._wrapperFields.start_seg.copyRefObject(this._refObject.start_seg);
    this._wrapperFields.goal_seg.copyRefObject(this._refObject.goal_seg);
    this._distanceIntialized = true;
  }
  copy(other) {
    this._refObject = new GetGeoPath_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._statusIntialized = true;
    this._wrapperFields.status.copy(other._wrapperFields.status);
    this._wrapperFields.plan.copy(other._wrapperFields.plan);
    this._wrapperFields.network.copy(other._wrapperFields.network);
    this._wrapperFields.start_seg.copy(other._wrapperFields.start_seg);
    this._wrapperFields.goal_seg.copy(other._wrapperFields.goal_seg);
    this._distanceIntialized = true;
  }
  static get classType() {
    return GetGeoPath_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"plan","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPath","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"network","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_seg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"goal_seg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"distance","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"GetGeoPath_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetGeoPath_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","status","plan","network","start_seg","goal_seg","distance"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetGeoPath_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetGeoPath_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetGeoPath_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGeoPath_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetGeoPath_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetGeoPath_ResponseRefArray(size);
    this._refObject = new GetGeoPath_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetGeoPath_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetGeoPath_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetGeoPath_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetGeoPath_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetGeoPath_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetGeoPath_ResponseArrayWrapper;
  }
}
module.exports = GetGeoPath_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "plan",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPath",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "network",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start_seg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "goal_seg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "distance",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "GetGeoPath_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetGeoPath_Response"
}
*/
