// rclnodejs/generated/geographic_msgs/geographic_msgs__msg__RoutePath.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const KeyValueWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__KeyValue.js');
const RoutePathRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  network: UUIDWrapper.refObjectType,
  segments: UUIDWrapper.refObjectArrayType,
  props: KeyValueWrapper.refObjectArrayType,
});
const RoutePathRefArray = ArrayType(RoutePathRefStruct);
const RoutePathRefStructArray = StructType({
  data: RoutePathRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RoutePathWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RoutePathRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.network =  new UUIDWrapper(other._wrapperFields.network);
      this._wrapperFields.segments = UUIDWrapper.createArray();
      this._wrapperFields.segments.copy(other._wrapperFields.segments);
      this._wrapperFields.props = KeyValueWrapper.createArray();
      this._wrapperFields.props.copy(other._wrapperFields.props);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RoutePathRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.network = new UUIDWrapper();
    this._wrapperFields.segments = UUIDWrapper.createArray();
    this._wrapperFields.props = KeyValueWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RoutePathWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RoutePathArrayWrapper;
  }
  static get ArrayType() {
    return RoutePathArrayWrapper;
  }
  static get refObjectArrayType() {
    return RoutePathRefStructArray
  }
  static get refObjectType() {
    return RoutePathRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.network.freeze(own, checkConsistency);
    this._refObject.network = this._wrapperFields.network.refObject;
    this._wrapperFields.segments.freeze(own, checkConsistency);
    this._refObject.segments = this._wrapperFields.segments.refObject;
    this._wrapperFields.props.freeze(own, checkConsistency);
    this._refObject.props = this._wrapperFields.props.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.network.copyRefObject(refObject.network);
    this._wrapperFields.segments.copyRefObject(refObject.segments);
    this._wrapperFields.props.copyRefObject(refObject.props);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    UUIDWrapper.freeStruct(refObject.network);
    if (refObject.segments.size != 0) {
      UUIDWrapper.ArrayType.freeArray(refObject.segments);
      if (UUIDWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.segments, UUIDWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.props.size != 0) {
      KeyValueWrapper.ArrayType.freeArray(refObject.props);
      if (KeyValueWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.props, KeyValueWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    RoutePathWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'msg', interfaceName: 'RoutePath'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get network() {
    return this._wrapperFields.network;
  }
  set network(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.network.copy(value);
    } else {
      this._wrapperFields.network.copy(new UUIDWrapper(value));
    }
  }
  get segments() {
    return this._wrapperFields.segments;
  }
  set segments(value) {
    this._wrapperFields.segments.fill(value);
  }
  get props() {
    return this._wrapperFields.props;
  }
  set props(value) {
    this._wrapperFields.props.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new RoutePathRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.network.copyRefObject(this._refObject.network);
    this._wrapperFields.segments.copyRefObject(this._refObject.segments);
    this._wrapperFields.props.copyRefObject(this._refObject.props);
  }
  copy(other) {
    this._refObject = new RoutePathRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.network.copy(other._wrapperFields.network);
    this._wrapperFields.segments.copy(other._wrapperFields.segments);
    this._wrapperFields.props.copy(other._wrapperFields.props);
  }
  static get classType() {
    return RoutePathWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"network","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"segments","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"props","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geographic_msgs","type":"KeyValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"RoutePath","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RoutePath"};
  }
  hasMember(name) {
    let memberNames = ["header","network","segments","props"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RoutePathArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RoutePathWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RoutePathWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RoutePathArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RoutePathArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RoutePathRefArray(size);
    this._refObject = new RoutePathRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RoutePathWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RoutePathArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RoutePathArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RoutePathWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RoutePathWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RoutePathArrayWrapper;
  }
}
module.exports = RoutePathWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "network",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "segments",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "props",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geographic_msgs",
        "type": "KeyValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "RoutePath",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RoutePath"
}
*/
