// rclnodejs/generated/geographic_msgs/geographic_msgs__msg__RouteNetwork.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const UUIDWrapper = require('../../generated/unique_identifier_msgs/unique_identifier_msgs__msg__UUID.js');
const BoundingBoxWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__BoundingBox.js');
const WayPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__WayPoint.js');
const RouteSegmentWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__RouteSegment.js');
const KeyValueWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__KeyValue.js');
const RouteNetworkRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id: UUIDWrapper.refObjectType,
  bounds: BoundingBoxWrapper.refObjectType,
  points: WayPointWrapper.refObjectArrayType,
  segments: RouteSegmentWrapper.refObjectArrayType,
  props: KeyValueWrapper.refObjectArrayType,
});
const RouteNetworkRefArray = ArrayType(RouteNetworkRefStruct);
const RouteNetworkRefStructArray = StructType({
  data: RouteNetworkRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RouteNetworkWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RouteNetworkRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.id =  new UUIDWrapper(other._wrapperFields.id);
      this._wrapperFields.bounds =  new BoundingBoxWrapper(other._wrapperFields.bounds);
      this._wrapperFields.points = WayPointWrapper.createArray();
      this._wrapperFields.points.copy(other._wrapperFields.points);
      this._wrapperFields.segments = RouteSegmentWrapper.createArray();
      this._wrapperFields.segments.copy(other._wrapperFields.segments);
      this._wrapperFields.props = KeyValueWrapper.createArray();
      this._wrapperFields.props.copy(other._wrapperFields.props);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RouteNetworkRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.id = new UUIDWrapper();
    this._wrapperFields.bounds = new BoundingBoxWrapper();
    this._wrapperFields.points = WayPointWrapper.createArray();
    this._wrapperFields.segments = RouteSegmentWrapper.createArray();
    this._wrapperFields.props = KeyValueWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new RouteNetworkWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RouteNetworkArrayWrapper;
  }
  static get ArrayType() {
    return RouteNetworkArrayWrapper;
  }
  static get refObjectArrayType() {
    return RouteNetworkRefStructArray
  }
  static get refObjectType() {
    return RouteNetworkRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.id.freeze(own, checkConsistency);
    this._refObject.id = this._wrapperFields.id.refObject;
    this._wrapperFields.bounds.freeze(own, checkConsistency);
    this._refObject.bounds = this._wrapperFields.bounds.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
    this._wrapperFields.segments.freeze(own, checkConsistency);
    this._refObject.segments = this._wrapperFields.segments.refObject;
    this._wrapperFields.props.freeze(own, checkConsistency);
    this._refObject.props = this._wrapperFields.props.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.id.copyRefObject(refObject.id);
    this._wrapperFields.bounds.copyRefObject(refObject.bounds);
    this._wrapperFields.points.copyRefObject(refObject.points);
    this._wrapperFields.segments.copyRefObject(refObject.segments);
    this._wrapperFields.props.copyRefObject(refObject.props);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    UUIDWrapper.freeStruct(refObject.id);
    BoundingBoxWrapper.freeStruct(refObject.bounds);
    if (refObject.points.size != 0) {
      WayPointWrapper.ArrayType.freeArray(refObject.points);
      if (WayPointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.points, WayPointWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.segments.size != 0) {
      RouteSegmentWrapper.ArrayType.freeArray(refObject.segments);
      if (RouteSegmentWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.segments, RouteSegmentWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.props.size != 0) {
      KeyValueWrapper.ArrayType.freeArray(refObject.props);
      if (KeyValueWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.props, KeyValueWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    RouteNetworkWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'msg', interfaceName: 'RouteNetwork'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id() {
    return this._wrapperFields.id;
  }
  set id(value) {
    if (value instanceof UUIDWrapper) {
      this._wrapperFields.id.copy(value);
    } else {
      this._wrapperFields.id.copy(new UUIDWrapper(value));
    }
  }
  get bounds() {
    return this._wrapperFields.bounds;
  }
  set bounds(value) {
    if (value instanceof BoundingBoxWrapper) {
      this._wrapperFields.bounds.copy(value);
    } else {
      this._wrapperFields.bounds.copy(new BoundingBoxWrapper(value));
    }
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    this._wrapperFields.points.fill(value);
  }
  get segments() {
    return this._wrapperFields.segments;
  }
  set segments(value) {
    this._wrapperFields.segments.fill(value);
  }
  get props() {
    return this._wrapperFields.props;
  }
  set props(value) {
    this._wrapperFields.props.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new RouteNetworkRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._wrapperFields.bounds.copyRefObject(this._refObject.bounds);
    this._wrapperFields.points.copyRefObject(this._refObject.points);
    this._wrapperFields.segments.copyRefObject(this._refObject.segments);
    this._wrapperFields.props.copyRefObject(this._refObject.props);
  }
  copy(other) {
    this._refObject = new RouteNetworkRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._wrapperFields.bounds.copy(other._wrapperFields.bounds);
    this._wrapperFields.points.copy(other._wrapperFields.points);
    this._wrapperFields.segments.copy(other._wrapperFields.segments);
    this._wrapperFields.props.copy(other._wrapperFields.props);
  }
  static get classType() {
    return RouteNetworkWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"unique_identifier_msgs","type":"UUID","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"bounds","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"BoundingBox","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geographic_msgs","type":"WayPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"segments","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geographic_msgs","type":"RouteSegment","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"props","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geographic_msgs","type":"KeyValue","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"RouteNetwork","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RouteNetwork"};
  }
  hasMember(name) {
    let memberNames = ["header","id","bounds","points","segments","props"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RouteNetworkArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RouteNetworkWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RouteNetworkWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteNetworkArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RouteNetworkArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RouteNetworkRefArray(size);
    this._refObject = new RouteNetworkRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RouteNetworkWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RouteNetworkArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RouteNetworkArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RouteNetworkWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RouteNetworkWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RouteNetworkArrayWrapper;
  }
}
module.exports = RouteNetworkWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "unique_identifier_msgs",
        "type": "UUID",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "bounds",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "BoundingBox",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geographic_msgs",
        "type": "WayPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "segments",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geographic_msgs",
        "type": "RouteSegment",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "props",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geographic_msgs",
        "type": "KeyValue",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "RouteNetwork",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RouteNetwork"
}
*/
