// rclnodejs/generated/geographic_msgs/geographic_msgs__srv__UpdateGeographicMap_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GeographicMapChangesWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeographicMapChanges.js');
const UpdateGeographicMap_RequestRefStruct = StructType({
  updates: GeographicMapChangesWrapper.refObjectType,
});
const UpdateGeographicMap_RequestRefArray = ArrayType(UpdateGeographicMap_RequestRefStruct);
const UpdateGeographicMap_RequestRefStructArray = StructType({
  data: UpdateGeographicMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UpdateGeographicMap_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UpdateGeographicMap_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.updates =  new GeographicMapChangesWrapper(other._wrapperFields.updates);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UpdateGeographicMap_RequestRefStruct();
    this._wrapperFields.updates = new GeographicMapChangesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new UpdateGeographicMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UpdateGeographicMap_RequestArrayWrapper;
  }
  static get ArrayType() {
    return UpdateGeographicMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return UpdateGeographicMap_RequestRefStructArray
  }
  static get refObjectType() {
    return UpdateGeographicMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.updates.freeze(own, checkConsistency);
    this._refObject.updates = this._wrapperFields.updates.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.updates.copyRefObject(refObject.updates);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GeographicMapChangesWrapper.freeStruct(refObject.updates);
  }
  static destoryRawROS(msg) {
    UpdateGeographicMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'srv', interfaceName: 'UpdateGeographicMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get updates() {
    return this._wrapperFields.updates;
  }
  set updates(value) {
    if (value instanceof GeographicMapChangesWrapper) {
      this._wrapperFields.updates.copy(value);
    } else {
      this._wrapperFields.updates.copy(new GeographicMapChangesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new UpdateGeographicMap_RequestRefStruct(refObject.toObject());
    this._wrapperFields.updates.copyRefObject(this._refObject.updates);
  }
  copy(other) {
    this._refObject = new UpdateGeographicMap_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.updates.copy(other._wrapperFields.updates);
  }
  static get classType() {
    return UpdateGeographicMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"updates","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeographicMapChanges","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"UpdateGeographicMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UpdateGeographicMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["updates"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UpdateGeographicMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UpdateGeographicMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UpdateGeographicMap_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdateGeographicMap_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdateGeographicMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UpdateGeographicMap_RequestRefArray(size);
    this._refObject = new UpdateGeographicMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UpdateGeographicMap_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UpdateGeographicMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UpdateGeographicMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UpdateGeographicMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UpdateGeographicMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UpdateGeographicMap_RequestArrayWrapper;
  }
}
module.exports = UpdateGeographicMap_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "updates",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeographicMapChanges",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "UpdateGeographicMap_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UpdateGeographicMap_Request"
}
*/
