// rclnodejs/generated/geographic_msgs/geographic_msgs__msg__BoundingBox.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GeoPointWrapper = require('../../generated/geographic_msgs/geographic_msgs__msg__GeoPoint.js');
const BoundingBoxRefStruct = StructType({
  min_pt: GeoPointWrapper.refObjectType,
  max_pt: GeoPointWrapper.refObjectType,
});
const BoundingBoxRefArray = ArrayType(BoundingBoxRefStruct);
const BoundingBoxRefStructArray = StructType({
  data: BoundingBoxRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BoundingBoxWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BoundingBoxRefStruct(other._refObject.toObject());
      this._wrapperFields.min_pt =  new GeoPointWrapper(other._wrapperFields.min_pt);
      this._wrapperFields.max_pt =  new GeoPointWrapper(other._wrapperFields.max_pt);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BoundingBoxRefStruct();
    this._wrapperFields.min_pt = new GeoPointWrapper();
    this._wrapperFields.max_pt = new GeoPointWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BoundingBoxWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BoundingBoxArrayWrapper;
  }
  static get ArrayType() {
    return BoundingBoxArrayWrapper;
  }
  static get refObjectArrayType() {
    return BoundingBoxRefStructArray
  }
  static get refObjectType() {
    return BoundingBoxRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.min_pt.freeze(own, checkConsistency);
    this._refObject.min_pt = this._wrapperFields.min_pt.refObject;
    this._wrapperFields.max_pt.freeze(own, checkConsistency);
    this._refObject.max_pt = this._wrapperFields.max_pt.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.min_pt.copyRefObject(refObject.min_pt);
    this._wrapperFields.max_pt.copyRefObject(refObject.max_pt);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    GeoPointWrapper.freeStruct(refObject.min_pt);
    GeoPointWrapper.freeStruct(refObject.max_pt);
  }
  static destoryRawROS(msg) {
    BoundingBoxWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'msg', interfaceName: 'BoundingBox'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get min_pt() {
    return this._wrapperFields.min_pt;
  }
  set min_pt(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.min_pt.copy(value);
    } else {
      this._wrapperFields.min_pt.copy(new GeoPointWrapper(value));
    }
  }
  get max_pt() {
    return this._wrapperFields.max_pt;
  }
  set max_pt(value) {
    if (value instanceof GeoPointWrapper) {
      this._wrapperFields.max_pt.copy(value);
    } else {
      this._wrapperFields.max_pt.copy(new GeoPointWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new BoundingBoxRefStruct(refObject.toObject());
    this._wrapperFields.min_pt.copyRefObject(this._refObject.min_pt);
    this._wrapperFields.max_pt.copyRefObject(this._refObject.max_pt);
  }
  copy(other) {
    this._refObject = new BoundingBoxRefStruct(other._refObject.toObject());
    this._wrapperFields.min_pt.copy(other._wrapperFields.min_pt);
    this._wrapperFields.max_pt.copy(other._wrapperFields.max_pt);
  }
  static get classType() {
    return BoundingBoxWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"min_pt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"max_pt","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geographic_msgs","type":"GeoPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"BoundingBox","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BoundingBox"};
  }
  hasMember(name) {
    let memberNames = ["min_pt","max_pt"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BoundingBoxArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BoundingBoxWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BoundingBoxWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBoxArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BoundingBoxArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BoundingBoxRefArray(size);
    this._refObject = new BoundingBoxRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BoundingBoxWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BoundingBoxArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BoundingBoxArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BoundingBoxWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BoundingBoxWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BoundingBoxArrayWrapper;
  }
}
module.exports = BoundingBoxWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "min_pt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "max_pt",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geographic_msgs",
        "type": "GeoPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "BoundingBox",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BoundingBox"
}
*/
