// rclnodejs/generated/geographic_msgs/geographic_msgs__srv__UpdateGeographicMap_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const UpdateGeographicMap_ResponseRefStruct = StructType({
  success: primitiveTypes.bool,
  status: primitiveTypes.string,
});
const UpdateGeographicMap_ResponseRefArray = ArrayType(UpdateGeographicMap_ResponseRefStruct);
const UpdateGeographicMap_ResponseRefStructArray = StructType({
  data: UpdateGeographicMap_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class UpdateGeographicMap_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new UpdateGeographicMap_ResponseRefStruct(other._refObject.toObject());
      this._successIntialized = true;
      this._statusIntialized = true;
      this._wrapperFields.status =  new StringWrapper(other._wrapperFields.status);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new UpdateGeographicMap_ResponseRefStruct();
    this._successIntialized = false;
    this._statusIntialized = false;
    this._wrapperFields.status = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new UpdateGeographicMap_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new UpdateGeographicMap_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return UpdateGeographicMap_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return UpdateGeographicMap_ResponseRefStructArray
  }
  static get refObjectType() {
    return UpdateGeographicMap_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in UpdateGeographicMap_Response');
      }
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in UpdateGeographicMap_Response');
      }
    }
    if (own) {
      this._wrapperFields.status.freeze(own, checkConsistency);
    }
    this._refObject.status = this._wrapperFields.status.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._statusIntialized = true;
    this._wrapperFields.status.data = refObject.status.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.status);
  }
  static destoryRawROS(msg) {
    UpdateGeographicMap_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'geographic_msgs', subFolder: 'srv', interfaceName: 'UpdateGeographicMap_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._wrapperFields.status.data;
  }
  set status(value) {
    this._statusIntialized = true;
    this._wrapperFields.status.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new UpdateGeographicMap_ResponseRefStruct(refObject.toObject());
    this._successIntialized = true;
    this._statusIntialized = true;
    this._wrapperFields.status.copyRefObject(this._refObject.status);
  }
  copy(other) {
    this._refObject = new UpdateGeographicMap_ResponseRefStruct(other._refObject.toObject());
    this._successIntialized = true;
    this._statusIntialized = true;
    this._wrapperFields.status.copy(other._wrapperFields.status);
  }
  static get classType() {
    return UpdateGeographicMap_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"geographic_msgs","type":"UpdateGeographicMap_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"UpdateGeographicMap_Response"};
  }
  hasMember(name) {
    let memberNames = ["success","status"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class UpdateGeographicMap_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof UpdateGeographicMap_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new UpdateGeographicMap_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdateGeographicMap_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to UpdateGeographicMap_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new UpdateGeographicMap_ResponseRefArray(size);
    this._refObject = new UpdateGeographicMap_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new UpdateGeographicMap_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof UpdateGeographicMap_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "UpdateGeographicMap_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      UpdateGeographicMap_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return UpdateGeographicMap_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return UpdateGeographicMap_ResponseArrayWrapper;
  }
}
module.exports = UpdateGeographicMap_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "geographic_msgs",
    "type": "UpdateGeographicMap_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "UpdateGeographicMap_Response"
}
*/
