// rclnodejs/generated/pcl_msgs/pcl_msgs__msg__PolygonMesh.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const VerticesWrapper = require('../../generated/pcl_msgs/pcl_msgs__msg__Vertices.js');
const PolygonMeshRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  cloud: PointCloud2Wrapper.refObjectType,
  polygons: VerticesWrapper.refObjectArrayType,
});
const PolygonMeshRefArray = ArrayType(PolygonMeshRefStruct);
const PolygonMeshRefStructArray = StructType({
  data: PolygonMeshRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PolygonMeshWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PolygonMeshRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.cloud =  new PointCloud2Wrapper(other._wrapperFields.cloud);
      this._wrapperFields.polygons = VerticesWrapper.createArray();
      this._wrapperFields.polygons.copy(other._wrapperFields.polygons);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PolygonMeshRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.cloud = new PointCloud2Wrapper();
    this._wrapperFields.polygons = VerticesWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PolygonMeshWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PolygonMeshArrayWrapper;
  }
  static get ArrayType() {
    return PolygonMeshArrayWrapper;
  }
  static get refObjectArrayType() {
    return PolygonMeshRefStructArray
  }
  static get refObjectType() {
    return PolygonMeshRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.cloud.freeze(own, checkConsistency);
    this._refObject.cloud = this._wrapperFields.cloud.refObject;
    this._wrapperFields.polygons.freeze(own, checkConsistency);
    this._refObject.polygons = this._wrapperFields.polygons.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.cloud.copyRefObject(refObject.cloud);
    this._wrapperFields.polygons.copyRefObject(refObject.polygons);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PointCloud2Wrapper.freeStruct(refObject.cloud);
    if (refObject.polygons.size != 0) {
      VerticesWrapper.ArrayType.freeArray(refObject.polygons);
      if (VerticesWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.polygons, VerticesWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PolygonMeshWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pcl_msgs', subFolder: 'msg', interfaceName: 'PolygonMesh'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get cloud() {
    return this._wrapperFields.cloud;
  }
  set cloud(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.cloud.copy(value);
    } else {
      this._wrapperFields.cloud.copy(new PointCloud2Wrapper(value));
    }
  }
  get polygons() {
    return this._wrapperFields.polygons;
  }
  set polygons(value) {
    this._wrapperFields.polygons.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new PolygonMeshRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.cloud.copyRefObject(this._refObject.cloud);
    this._wrapperFields.polygons.copyRefObject(this._refObject.polygons);
  }
  copy(other) {
    this._refObject = new PolygonMeshRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.cloud.copy(other._wrapperFields.cloud);
    this._wrapperFields.polygons.copy(other._wrapperFields.polygons);
  }
  static get classType() {
    return PolygonMeshWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"cloud","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"polygons","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"pcl_msgs","type":"Vertices","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"pcl_msgs","type":"PolygonMesh","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PolygonMesh"};
  }
  hasMember(name) {
    let memberNames = ["header","cloud","polygons"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PolygonMeshArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PolygonMeshWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PolygonMeshWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PolygonMeshArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PolygonMeshArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PolygonMeshRefArray(size);
    this._refObject = new PolygonMeshRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PolygonMeshWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PolygonMeshArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PolygonMeshArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PolygonMeshWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PolygonMeshWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PolygonMeshArrayWrapper;
  }
}
module.exports = PolygonMeshWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "cloud",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "PointCloud2",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "polygons",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "pcl_msgs",
        "type": "Vertices",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "pcl_msgs",
    "type": "PolygonMesh",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PolygonMesh"
}
*/
