// rclnodejs/generated/pcl_msgs/pcl_msgs__msg__Vertices.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const UInt32Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt32.js');
const VerticesRefStruct = StructType({
  vertices: UInt32Wrapper.refObjectArrayType,
});
const VerticesRefArray = ArrayType(VerticesRefStruct);
const VerticesRefStructArray = StructType({
  data: VerticesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class VerticesWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new VerticesRefStruct(other._refObject.toObject());
      this._wrapperFields.vertices = UInt32Wrapper.createArray();
      this._wrapperFields.vertices.copy(other._wrapperFields.vertices);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new VerticesRefStruct();
    this._wrapperFields.vertices = UInt32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new VerticesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new VerticesArrayWrapper;
  }
  static get ArrayType() {
    return VerticesArrayWrapper;
  }
  static get refObjectArrayType() {
    return VerticesRefStructArray
  }
  static get refObjectType() {
    return VerticesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.vertices.freeze(own, checkConsistency);
    this._refObject.vertices = this._wrapperFields.vertices.refObject;
    if (own) {
      this._wrapperFields.vertices.fill(Uint32Array.from([]));
      this._wrapperFields.vertices.freeze(own, checkConsistency);
      this._refObject.vertices = this._wrapperFields.vertices.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.vertices.copyRefObject(refObject.vertices);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.vertices.size != 0) {
      UInt32Wrapper.ArrayType.freeArray(refObject.vertices);
      if (UInt32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.vertices, UInt32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    VerticesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pcl_msgs', subFolder: 'msg', interfaceName: 'Vertices'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get vertices() {
    return this._wrapperFields['vertices'].data;
  }
  set vertices(value) {
    this._wrapperFields['vertices'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new VerticesRefStruct(refObject.toObject());
    this._wrapperFields.vertices.copyRefObject(this._refObject.vertices);
  }
  copy(other) {
    this._refObject = new VerticesRefStruct(other._refObject.toObject());
    this._wrapperFields.vertices.copy(other._wrapperFields.vertices);
  }
  static get classType() {
    return VerticesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"vertices","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pcl_msgs","type":"Vertices","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Vertices"};
  }
  hasMember(name) {
    let memberNames = ["vertices"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class VerticesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof VerticesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new VerticesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VerticesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to VerticesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new VerticesRefArray(size);
    this._refObject = new VerticesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new VerticesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof VerticesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "VerticesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      VerticesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return VerticesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return VerticesArrayWrapper;
  }
}
module.exports = VerticesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "vertices",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "pcl_msgs",
    "type": "Vertices",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Vertices"
}
*/
