// rclnodejs/generated/pcl_msgs/pcl_msgs__msg__PointIndices.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const PointIndicesRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  indices: Int32Wrapper.refObjectArrayType,
});
const PointIndicesRefArray = ArrayType(PointIndicesRefStruct);
const PointIndicesRefStructArray = StructType({
  data: PointIndicesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PointIndicesWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PointIndicesRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.indices = Int32Wrapper.createArray();
      this._wrapperFields.indices.copy(other._wrapperFields.indices);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PointIndicesRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.indices = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new PointIndicesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PointIndicesArrayWrapper;
  }
  static get ArrayType() {
    return PointIndicesArrayWrapper;
  }
  static get refObjectArrayType() {
    return PointIndicesRefStructArray
  }
  static get refObjectType() {
    return PointIndicesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.indices.freeze(own, checkConsistency);
    this._refObject.indices = this._wrapperFields.indices.refObject;
    if (own) {
      this._wrapperFields.indices.fill(Int32Array.from([]));
      this._wrapperFields.indices.freeze(own, checkConsistency);
      this._refObject.indices = this._wrapperFields.indices.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._wrapperFields.indices.copyRefObject(refObject.indices);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.indices.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.indices);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.indices, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    PointIndicesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'pcl_msgs', subFolder: 'msg', interfaceName: 'PointIndices'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get indices() {
    return this._wrapperFields['indices'].data;
  }
  set indices(value) {
    this._wrapperFields['indices'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new PointIndicesRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._wrapperFields.indices.copyRefObject(this._refObject.indices);
  }
  copy(other) {
    this._refObject = new PointIndicesRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.indices.copy(other._wrapperFields.indices);
  }
  static get classType() {
    return PointIndicesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"indices","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"pcl_msgs","type":"PointIndices","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PointIndices"};
  }
  hasMember(name) {
    let memberNames = ["header","indices"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PointIndicesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PointIndicesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PointIndicesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointIndicesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointIndicesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PointIndicesRefArray(size);
    this._refObject = new PointIndicesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PointIndicesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PointIndicesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PointIndicesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PointIndicesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PointIndicesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PointIndicesArrayWrapper;
  }
}
module.exports = PointIndicesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "indices",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "pcl_msgs",
    "type": "PointIndices",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PointIndices"
}
*/
