// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__GetLinkProperties_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetLinkProperties_ResponseRefStruct = StructType({
  com: PoseWrapper.refObjectType,
  gravity_mode: primitiveTypes.bool,
  mass: primitiveTypes.float64,
  ixx: primitiveTypes.float64,
  ixy: primitiveTypes.float64,
  ixz: primitiveTypes.float64,
  iyy: primitiveTypes.float64,
  iyz: primitiveTypes.float64,
  izz: primitiveTypes.float64,
  success: primitiveTypes.bool,
  status_message: primitiveTypes.string,
});
const GetLinkProperties_ResponseRefArray = ArrayType(GetLinkProperties_ResponseRefStruct);
const GetLinkProperties_ResponseRefStructArray = StructType({
  data: GetLinkProperties_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetLinkProperties_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetLinkProperties_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.com =  new PoseWrapper(other._wrapperFields.com);
      this._gravity_modeIntialized = true;
      this._massIntialized = true;
      this._ixxIntialized = true;
      this._ixyIntialized = true;
      this._ixzIntialized = true;
      this._iyyIntialized = true;
      this._iyzIntialized = true;
      this._izzIntialized = true;
      this._successIntialized = true;
      this._status_messageIntialized = true;
      this._wrapperFields.status_message =  new StringWrapper(other._wrapperFields.status_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetLinkProperties_ResponseRefStruct();
    this._wrapperFields.com = new PoseWrapper();
    this._gravity_modeIntialized = false;
    this._massIntialized = false;
    this._ixxIntialized = false;
    this._ixyIntialized = false;
    this._ixzIntialized = false;
    this._iyyIntialized = false;
    this._iyzIntialized = false;
    this._izzIntialized = false;
    this._successIntialized = false;
    this._status_messageIntialized = false;
    this._wrapperFields.status_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetLinkProperties_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetLinkProperties_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetLinkProperties_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetLinkProperties_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetLinkProperties_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._gravity_modeIntialized) {
        throw new TypeError('Invalid argument: gravity_mode in GetLinkProperties_Response');
      }
      if (!this._massIntialized) {
        throw new TypeError('Invalid argument: mass in GetLinkProperties_Response');
      }
      if (!this._ixxIntialized) {
        throw new TypeError('Invalid argument: ixx in GetLinkProperties_Response');
      }
      if (!this._ixyIntialized) {
        throw new TypeError('Invalid argument: ixy in GetLinkProperties_Response');
      }
      if (!this._ixzIntialized) {
        throw new TypeError('Invalid argument: ixz in GetLinkProperties_Response');
      }
      if (!this._iyyIntialized) {
        throw new TypeError('Invalid argument: iyy in GetLinkProperties_Response');
      }
      if (!this._iyzIntialized) {
        throw new TypeError('Invalid argument: iyz in GetLinkProperties_Response');
      }
      if (!this._izzIntialized) {
        throw new TypeError('Invalid argument: izz in GetLinkProperties_Response');
      }
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetLinkProperties_Response');
      }
      if (!this._status_messageIntialized) {
        throw new TypeError('Invalid argument: status_message in GetLinkProperties_Response');
      }
    }
    this._wrapperFields.com.freeze(own, checkConsistency);
    this._refObject.com = this._wrapperFields.com.refObject;
    if (own) {
      this._wrapperFields.status_message.freeze(own, checkConsistency);
    }
    this._refObject.status_message = this._wrapperFields.status_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.com.copyRefObject(refObject.com);
    this._gravity_modeIntialized = true;
    this._refObject.gravity_mode = refObject.gravity_mode;
    this._massIntialized = true;
    this._refObject.mass = refObject.mass;
    this._ixxIntialized = true;
    this._refObject.ixx = refObject.ixx;
    this._ixyIntialized = true;
    this._refObject.ixy = refObject.ixy;
    this._ixzIntialized = true;
    this._refObject.ixz = refObject.ixz;
    this._iyyIntialized = true;
    this._refObject.iyy = refObject.iyy;
    this._iyzIntialized = true;
    this._refObject.iyz = refObject.iyz;
    this._izzIntialized = true;
    this._refObject.izz = refObject.izz;
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = refObject.status_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseWrapper.freeStruct(refObject.com);
    StringWrapper.freeStruct(refObject.status_message);
  }
  static destoryRawROS(msg) {
    GetLinkProperties_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'GetLinkProperties_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get com() {
    return this._wrapperFields.com;
  }
  set com(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.com.copy(value);
    } else {
      this._wrapperFields.com.copy(new PoseWrapper(value));
    }
  }
  get gravity_mode() {
    if (!this._gravity_modeIntialized) {
      return undefined;
    }
    return this._refObject.gravity_mode;
  }
  set gravity_mode(value) {
    this._gravity_modeIntialized = true;
    this._refObject.gravity_mode = value;
  }
  get mass() {
    if (!this._massIntialized) {
      return undefined;
    }
    return this._refObject.mass;
  }
  set mass(value) {
    this._massIntialized = true;
    this._refObject.mass = value;
  }
  get ixx() {
    if (!this._ixxIntialized) {
      return undefined;
    }
    return this._refObject.ixx;
  }
  set ixx(value) {
    this._ixxIntialized = true;
    this._refObject.ixx = value;
  }
  get ixy() {
    if (!this._ixyIntialized) {
      return undefined;
    }
    return this._refObject.ixy;
  }
  set ixy(value) {
    this._ixyIntialized = true;
    this._refObject.ixy = value;
  }
  get ixz() {
    if (!this._ixzIntialized) {
      return undefined;
    }
    return this._refObject.ixz;
  }
  set ixz(value) {
    this._ixzIntialized = true;
    this._refObject.ixz = value;
  }
  get iyy() {
    if (!this._iyyIntialized) {
      return undefined;
    }
    return this._refObject.iyy;
  }
  set iyy(value) {
    this._iyyIntialized = true;
    this._refObject.iyy = value;
  }
  get iyz() {
    if (!this._iyzIntialized) {
      return undefined;
    }
    return this._refObject.iyz;
  }
  set iyz(value) {
    this._iyzIntialized = true;
    this._refObject.iyz = value;
  }
  get izz() {
    if (!this._izzIntialized) {
      return undefined;
    }
    return this._refObject.izz;
  }
  set izz(value) {
    this._izzIntialized = true;
    this._refObject.izz = value;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status_message() {
    if (!this._status_messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.status_message.data;
  }
  set status_message(value) {
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetLinkProperties_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.com.copyRefObject(this._refObject.com);
    this._gravity_modeIntialized = true;
    this._massIntialized = true;
    this._ixxIntialized = true;
    this._ixyIntialized = true;
    this._ixzIntialized = true;
    this._iyyIntialized = true;
    this._iyzIntialized = true;
    this._izzIntialized = true;
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copyRefObject(this._refObject.status_message);
  }
  copy(other) {
    this._refObject = new GetLinkProperties_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.com.copy(other._wrapperFields.com);
    this._gravity_modeIntialized = true;
    this._massIntialized = true;
    this._ixxIntialized = true;
    this._ixyIntialized = true;
    this._ixzIntialized = true;
    this._iyyIntialized = true;
    this._iyzIntialized = true;
    this._izzIntialized = true;
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copy(other._wrapperFields.status_message);
  }
  static get classType() {
    return GetLinkProperties_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"com","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"gravity_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mass","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"izz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"GetLinkProperties_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetLinkProperties_Response"};
  }
  hasMember(name) {
    let memberNames = ["com","gravity_mode","mass","ixx","ixy","ixz","iyy","iyz","izz","success","status_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetLinkProperties_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetLinkProperties_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetLinkProperties_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLinkProperties_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLinkProperties_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetLinkProperties_ResponseRefArray(size);
    this._refObject = new GetLinkProperties_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetLinkProperties_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetLinkProperties_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetLinkProperties_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetLinkProperties_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetLinkProperties_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetLinkProperties_ResponseArrayWrapper;
  }
}
module.exports = GetLinkProperties_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "com",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "gravity_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mass",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixx",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iyy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iyz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "izz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "GetLinkProperties_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetLinkProperties_Response"
}
*/
