// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__SetJointTrajectory_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const JointTrajectoryWrapper = require('../../generated/trajectory_msgs/trajectory_msgs__msg__JointTrajectory.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SetJointTrajectory_RequestRefStruct = StructType({
  model_name: primitiveTypes.string,
  joint_trajectory: JointTrajectoryWrapper.refObjectType,
  model_pose: PoseWrapper.refObjectType,
  set_model_pose: primitiveTypes.bool,
  disable_physics_updates: primitiveTypes.bool,
});
const SetJointTrajectory_RequestRefArray = ArrayType(SetJointTrajectory_RequestRefStruct);
const SetJointTrajectory_RequestRefStructArray = StructType({
  data: SetJointTrajectory_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetJointTrajectory_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetJointTrajectory_RequestRefStruct(other._refObject.toObject());
      this._model_nameIntialized = true;
      this._wrapperFields.model_name =  new StringWrapper(other._wrapperFields.model_name);
      this._wrapperFields.joint_trajectory =  new JointTrajectoryWrapper(other._wrapperFields.joint_trajectory);
      this._wrapperFields.model_pose =  new PoseWrapper(other._wrapperFields.model_pose);
      this._set_model_poseIntialized = true;
      this._disable_physics_updatesIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetJointTrajectory_RequestRefStruct();
    this._model_nameIntialized = false;
    this._wrapperFields.model_name = new StringWrapper();
    this._wrapperFields.joint_trajectory = new JointTrajectoryWrapper();
    this._wrapperFields.model_pose = new PoseWrapper();
    this._set_model_poseIntialized = false;
    this._disable_physics_updatesIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetJointTrajectory_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetJointTrajectory_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetJointTrajectory_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetJointTrajectory_RequestRefStructArray
  }
  static get refObjectType() {
    return SetJointTrajectory_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._model_nameIntialized) {
        throw new TypeError('Invalid argument: model_name in SetJointTrajectory_Request');
      }
      if (!this._set_model_poseIntialized) {
        throw new TypeError('Invalid argument: set_model_pose in SetJointTrajectory_Request');
      }
      if (!this._disable_physics_updatesIntialized) {
        throw new TypeError('Invalid argument: disable_physics_updates in SetJointTrajectory_Request');
      }
    }
    if (own) {
      this._wrapperFields.model_name.freeze(own, checkConsistency);
    }
    this._refObject.model_name = this._wrapperFields.model_name.refObject;
    this._wrapperFields.joint_trajectory.freeze(own, checkConsistency);
    this._refObject.joint_trajectory = this._wrapperFields.joint_trajectory.refObject;
    this._wrapperFields.model_pose.freeze(own, checkConsistency);
    this._refObject.model_pose = this._wrapperFields.model_pose.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = refObject.model_name.data;
    this._wrapperFields.joint_trajectory.copyRefObject(refObject.joint_trajectory);
    this._wrapperFields.model_pose.copyRefObject(refObject.model_pose);
    this._set_model_poseIntialized = true;
    this._refObject.set_model_pose = refObject.set_model_pose;
    this._disable_physics_updatesIntialized = true;
    this._refObject.disable_physics_updates = refObject.disable_physics_updates;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.model_name);
    JointTrajectoryWrapper.freeStruct(refObject.joint_trajectory);
    PoseWrapper.freeStruct(refObject.model_pose);
  }
  static destoryRawROS(msg) {
    SetJointTrajectory_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'SetJointTrajectory_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get model_name() {
    if (!this._model_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.model_name.data;
  }
  set model_name(value) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = value;
  }
  get joint_trajectory() {
    return this._wrapperFields.joint_trajectory;
  }
  set joint_trajectory(value) {
    if (value instanceof JointTrajectoryWrapper) {
      this._wrapperFields.joint_trajectory.copy(value);
    } else {
      this._wrapperFields.joint_trajectory.copy(new JointTrajectoryWrapper(value));
    }
  }
  get model_pose() {
    return this._wrapperFields.model_pose;
  }
  set model_pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.model_pose.copy(value);
    } else {
      this._wrapperFields.model_pose.copy(new PoseWrapper(value));
    }
  }
  get set_model_pose() {
    if (!this._set_model_poseIntialized) {
      return undefined;
    }
    return this._refObject.set_model_pose;
  }
  set set_model_pose(value) {
    this._set_model_poseIntialized = true;
    this._refObject.set_model_pose = value;
  }
  get disable_physics_updates() {
    if (!this._disable_physics_updatesIntialized) {
      return undefined;
    }
    return this._refObject.disable_physics_updates;
  }
  set disable_physics_updates(value) {
    this._disable_physics_updatesIntialized = true;
    this._refObject.disable_physics_updates = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetJointTrajectory_RequestRefStruct(refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copyRefObject(this._refObject.model_name);
    this._wrapperFields.joint_trajectory.copyRefObject(this._refObject.joint_trajectory);
    this._wrapperFields.model_pose.copyRefObject(this._refObject.model_pose);
    this._set_model_poseIntialized = true;
    this._disable_physics_updatesIntialized = true;
  }
  copy(other) {
    this._refObject = new SetJointTrajectory_RequestRefStruct(other._refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copy(other._wrapperFields.model_name);
    this._wrapperFields.joint_trajectory.copy(other._wrapperFields.joint_trajectory);
    this._wrapperFields.model_pose.copy(other._wrapperFields.model_pose);
    this._set_model_poseIntialized = true;
    this._disable_physics_updatesIntialized = true;
  }
  static get classType() {
    return SetJointTrajectory_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"model_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"joint_trajectory","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"trajectory_msgs","type":"JointTrajectory","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"model_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"set_model_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"disable_physics_updates","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SetJointTrajectory_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetJointTrajectory_Request"};
  }
  hasMember(name) {
    let memberNames = ["model_name","joint_trajectory","model_pose","set_model_pose","disable_physics_updates"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetJointTrajectory_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetJointTrajectory_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetJointTrajectory_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetJointTrajectory_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetJointTrajectory_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetJointTrajectory_RequestRefArray(size);
    this._refObject = new SetJointTrajectory_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetJointTrajectory_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetJointTrajectory_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetJointTrajectory_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetJointTrajectory_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetJointTrajectory_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetJointTrajectory_RequestArrayWrapper;
  }
}
module.exports = SetJointTrajectory_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "model_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "joint_trajectory",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "trajectory_msgs",
        "type": "JointTrajectory",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "model_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "set_model_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "disable_physics_updates",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SetJointTrajectory_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetJointTrajectory_Request"
}
*/
