// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__SetLightProperties_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ColorRGBAWrapper = require('../../generated/std_msgs/std_msgs__msg__ColorRGBA.js');
const SetLightProperties_RequestRefStruct = StructType({
  light_name: primitiveTypes.string,
  diffuse: ColorRGBAWrapper.refObjectType,
  attenuation_constant: primitiveTypes.float64,
  attenuation_linear: primitiveTypes.float64,
  attenuation_quadratic: primitiveTypes.float64,
});
const SetLightProperties_RequestRefArray = ArrayType(SetLightProperties_RequestRefStruct);
const SetLightProperties_RequestRefStructArray = StructType({
  data: SetLightProperties_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetLightProperties_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetLightProperties_RequestRefStruct(other._refObject.toObject());
      this._light_nameIntialized = true;
      this._wrapperFields.light_name =  new StringWrapper(other._wrapperFields.light_name);
      this._wrapperFields.diffuse =  new ColorRGBAWrapper(other._wrapperFields.diffuse);
      this._attenuation_constantIntialized = true;
      this._attenuation_linearIntialized = true;
      this._attenuation_quadraticIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetLightProperties_RequestRefStruct();
    this._light_nameIntialized = false;
    this._wrapperFields.light_name = new StringWrapper();
    this._wrapperFields.diffuse = new ColorRGBAWrapper();
    this._attenuation_constantIntialized = false;
    this._attenuation_linearIntialized = false;
    this._attenuation_quadraticIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetLightProperties_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetLightProperties_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetLightProperties_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetLightProperties_RequestRefStructArray
  }
  static get refObjectType() {
    return SetLightProperties_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._light_nameIntialized) {
        throw new TypeError('Invalid argument: light_name in SetLightProperties_Request');
      }
      if (!this._attenuation_constantIntialized) {
        throw new TypeError('Invalid argument: attenuation_constant in SetLightProperties_Request');
      }
      if (!this._attenuation_linearIntialized) {
        throw new TypeError('Invalid argument: attenuation_linear in SetLightProperties_Request');
      }
      if (!this._attenuation_quadraticIntialized) {
        throw new TypeError('Invalid argument: attenuation_quadratic in SetLightProperties_Request');
      }
    }
    if (own) {
      this._wrapperFields.light_name.freeze(own, checkConsistency);
    }
    this._refObject.light_name = this._wrapperFields.light_name.refObject;
    this._wrapperFields.diffuse.freeze(own, checkConsistency);
    this._refObject.diffuse = this._wrapperFields.diffuse.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._light_nameIntialized = true;
    this._wrapperFields.light_name.data = refObject.light_name.data;
    this._wrapperFields.diffuse.copyRefObject(refObject.diffuse);
    this._attenuation_constantIntialized = true;
    this._refObject.attenuation_constant = refObject.attenuation_constant;
    this._attenuation_linearIntialized = true;
    this._refObject.attenuation_linear = refObject.attenuation_linear;
    this._attenuation_quadraticIntialized = true;
    this._refObject.attenuation_quadratic = refObject.attenuation_quadratic;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.light_name);
    ColorRGBAWrapper.freeStruct(refObject.diffuse);
  }
  static destoryRawROS(msg) {
    SetLightProperties_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'SetLightProperties_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get light_name() {
    if (!this._light_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.light_name.data;
  }
  set light_name(value) {
    this._light_nameIntialized = true;
    this._wrapperFields.light_name.data = value;
  }
  get diffuse() {
    return this._wrapperFields.diffuse;
  }
  set diffuse(value) {
    if (value instanceof ColorRGBAWrapper) {
      this._wrapperFields.diffuse.copy(value);
    } else {
      this._wrapperFields.diffuse.copy(new ColorRGBAWrapper(value));
    }
  }
  get attenuation_constant() {
    if (!this._attenuation_constantIntialized) {
      return undefined;
    }
    return this._refObject.attenuation_constant;
  }
  set attenuation_constant(value) {
    this._attenuation_constantIntialized = true;
    this._refObject.attenuation_constant = value;
  }
  get attenuation_linear() {
    if (!this._attenuation_linearIntialized) {
      return undefined;
    }
    return this._refObject.attenuation_linear;
  }
  set attenuation_linear(value) {
    this._attenuation_linearIntialized = true;
    this._refObject.attenuation_linear = value;
  }
  get attenuation_quadratic() {
    if (!this._attenuation_quadraticIntialized) {
      return undefined;
    }
    return this._refObject.attenuation_quadratic;
  }
  set attenuation_quadratic(value) {
    this._attenuation_quadraticIntialized = true;
    this._refObject.attenuation_quadratic = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetLightProperties_RequestRefStruct(refObject.toObject());
    this._light_nameIntialized = true;
    this._wrapperFields.light_name.copyRefObject(this._refObject.light_name);
    this._wrapperFields.diffuse.copyRefObject(this._refObject.diffuse);
    this._attenuation_constantIntialized = true;
    this._attenuation_linearIntialized = true;
    this._attenuation_quadraticIntialized = true;
  }
  copy(other) {
    this._refObject = new SetLightProperties_RequestRefStruct(other._refObject.toObject());
    this._light_nameIntialized = true;
    this._wrapperFields.light_name.copy(other._wrapperFields.light_name);
    this._wrapperFields.diffuse.copy(other._wrapperFields.diffuse);
    this._attenuation_constantIntialized = true;
    this._attenuation_linearIntialized = true;
    this._attenuation_quadraticIntialized = true;
  }
  static get classType() {
    return SetLightProperties_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"light_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"diffuse","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"ColorRGBA","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"attenuation_constant","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"attenuation_linear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"attenuation_quadratic","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SetLightProperties_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetLightProperties_Request"};
  }
  hasMember(name) {
    let memberNames = ["light_name","diffuse","attenuation_constant","attenuation_linear","attenuation_quadratic"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetLightProperties_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetLightProperties_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetLightProperties_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLightProperties_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLightProperties_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetLightProperties_RequestRefArray(size);
    this._refObject = new SetLightProperties_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetLightProperties_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetLightProperties_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetLightProperties_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetLightProperties_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetLightProperties_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetLightProperties_RequestArrayWrapper;
  }
}
module.exports = SetLightProperties_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "light_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "diffuse",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "ColorRGBA",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "attenuation_constant",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "attenuation_linear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "attenuation_quadratic",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SetLightProperties_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetLightProperties_Request"
}
*/
