// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__GetPhysicsProperties_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const ODEPhysicsWrapper = require('../../generated/gazebo_msgs/gazebo_msgs__msg__ODEPhysics.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetPhysicsProperties_ResponseRefStruct = StructType({
  time_step: primitiveTypes.float64,
  pause: primitiveTypes.bool,
  max_update_rate: primitiveTypes.float64,
  gravity: Vector3Wrapper.refObjectType,
  ode_config: ODEPhysicsWrapper.refObjectType,
  success: primitiveTypes.bool,
  status_message: primitiveTypes.string,
});
const GetPhysicsProperties_ResponseRefArray = ArrayType(GetPhysicsProperties_ResponseRefStruct);
const GetPhysicsProperties_ResponseRefStructArray = StructType({
  data: GetPhysicsProperties_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPhysicsProperties_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPhysicsProperties_ResponseRefStruct(other._refObject.toObject());
      this._time_stepIntialized = true;
      this._pauseIntialized = true;
      this._max_update_rateIntialized = true;
      this._wrapperFields.gravity =  new Vector3Wrapper(other._wrapperFields.gravity);
      this._wrapperFields.ode_config =  new ODEPhysicsWrapper(other._wrapperFields.ode_config);
      this._successIntialized = true;
      this._status_messageIntialized = true;
      this._wrapperFields.status_message =  new StringWrapper(other._wrapperFields.status_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPhysicsProperties_ResponseRefStruct();
    this._time_stepIntialized = false;
    this._pauseIntialized = false;
    this._max_update_rateIntialized = false;
    this._wrapperFields.gravity = new Vector3Wrapper();
    this._wrapperFields.ode_config = new ODEPhysicsWrapper();
    this._successIntialized = false;
    this._status_messageIntialized = false;
    this._wrapperFields.status_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPhysicsProperties_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPhysicsProperties_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetPhysicsProperties_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPhysicsProperties_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetPhysicsProperties_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._time_stepIntialized) {
        throw new TypeError('Invalid argument: time_step in GetPhysicsProperties_Response');
      }
      if (!this._pauseIntialized) {
        throw new TypeError('Invalid argument: pause in GetPhysicsProperties_Response');
      }
      if (!this._max_update_rateIntialized) {
        throw new TypeError('Invalid argument: max_update_rate in GetPhysicsProperties_Response');
      }
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetPhysicsProperties_Response');
      }
      if (!this._status_messageIntialized) {
        throw new TypeError('Invalid argument: status_message in GetPhysicsProperties_Response');
      }
    }
    this._wrapperFields.gravity.freeze(own, checkConsistency);
    this._refObject.gravity = this._wrapperFields.gravity.refObject;
    this._wrapperFields.ode_config.freeze(own, checkConsistency);
    this._refObject.ode_config = this._wrapperFields.ode_config.refObject;
    if (own) {
      this._wrapperFields.status_message.freeze(own, checkConsistency);
    }
    this._refObject.status_message = this._wrapperFields.status_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._time_stepIntialized = true;
    this._refObject.time_step = refObject.time_step;
    this._pauseIntialized = true;
    this._refObject.pause = refObject.pause;
    this._max_update_rateIntialized = true;
    this._refObject.max_update_rate = refObject.max_update_rate;
    this._wrapperFields.gravity.copyRefObject(refObject.gravity);
    this._wrapperFields.ode_config.copyRefObject(refObject.ode_config);
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = refObject.status_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    Vector3Wrapper.freeStruct(refObject.gravity);
    ODEPhysicsWrapper.freeStruct(refObject.ode_config);
    StringWrapper.freeStruct(refObject.status_message);
  }
  static destoryRawROS(msg) {
    GetPhysicsProperties_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'GetPhysicsProperties_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get time_step() {
    if (!this._time_stepIntialized) {
      return undefined;
    }
    return this._refObject.time_step;
  }
  set time_step(value) {
    this._time_stepIntialized = true;
    this._refObject.time_step = value;
  }
  get pause() {
    if (!this._pauseIntialized) {
      return undefined;
    }
    return this._refObject.pause;
  }
  set pause(value) {
    this._pauseIntialized = true;
    this._refObject.pause = value;
  }
  get max_update_rate() {
    if (!this._max_update_rateIntialized) {
      return undefined;
    }
    return this._refObject.max_update_rate;
  }
  set max_update_rate(value) {
    this._max_update_rateIntialized = true;
    this._refObject.max_update_rate = value;
  }
  get gravity() {
    return this._wrapperFields.gravity;
  }
  set gravity(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.gravity.copy(value);
    } else {
      this._wrapperFields.gravity.copy(new Vector3Wrapper(value));
    }
  }
  get ode_config() {
    return this._wrapperFields.ode_config;
  }
  set ode_config(value) {
    if (value instanceof ODEPhysicsWrapper) {
      this._wrapperFields.ode_config.copy(value);
    } else {
      this._wrapperFields.ode_config.copy(new ODEPhysicsWrapper(value));
    }
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status_message() {
    if (!this._status_messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.status_message.data;
  }
  set status_message(value) {
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetPhysicsProperties_ResponseRefStruct(refObject.toObject());
    this._time_stepIntialized = true;
    this._pauseIntialized = true;
    this._max_update_rateIntialized = true;
    this._wrapperFields.gravity.copyRefObject(this._refObject.gravity);
    this._wrapperFields.ode_config.copyRefObject(this._refObject.ode_config);
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copyRefObject(this._refObject.status_message);
  }
  copy(other) {
    this._refObject = new GetPhysicsProperties_ResponseRefStruct(other._refObject.toObject());
    this._time_stepIntialized = true;
    this._pauseIntialized = true;
    this._max_update_rateIntialized = true;
    this._wrapperFields.gravity.copy(other._wrapperFields.gravity);
    this._wrapperFields.ode_config.copy(other._wrapperFields.ode_config);
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copy(other._wrapperFields.status_message);
  }
  static get classType() {
    return GetPhysicsProperties_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"time_step","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pause","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_update_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"gravity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ode_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"gazebo_msgs","type":"ODEPhysics","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"GetPhysicsProperties_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPhysicsProperties_Response"};
  }
  hasMember(name) {
    let memberNames = ["time_step","pause","max_update_rate","gravity","ode_config","success","status_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPhysicsProperties_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPhysicsProperties_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPhysicsProperties_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPhysicsProperties_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPhysicsProperties_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPhysicsProperties_ResponseRefArray(size);
    this._refObject = new GetPhysicsProperties_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPhysicsProperties_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPhysicsProperties_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPhysicsProperties_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPhysicsProperties_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPhysicsProperties_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPhysicsProperties_ResponseArrayWrapper;
  }
}
module.exports = GetPhysicsProperties_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "time_step",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pause",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_update_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "gravity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "ode_config",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "gazebo_msgs",
        "type": "ODEPhysics",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "GetPhysicsProperties_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPhysicsProperties_Response"
}
*/
