// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__GetLinkState_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const LinkStateWrapper = require('../../generated/gazebo_msgs/gazebo_msgs__msg__LinkState.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetLinkState_ResponseRefStruct = StructType({
  link_state: LinkStateWrapper.refObjectType,
  success: primitiveTypes.bool,
  status_message: primitiveTypes.string,
});
const GetLinkState_ResponseRefArray = ArrayType(GetLinkState_ResponseRefStruct);
const GetLinkState_ResponseRefStructArray = StructType({
  data: GetLinkState_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetLinkState_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetLinkState_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.link_state =  new LinkStateWrapper(other._wrapperFields.link_state);
      this._successIntialized = true;
      this._status_messageIntialized = true;
      this._wrapperFields.status_message =  new StringWrapper(other._wrapperFields.status_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetLinkState_ResponseRefStruct();
    this._wrapperFields.link_state = new LinkStateWrapper();
    this._successIntialized = false;
    this._status_messageIntialized = false;
    this._wrapperFields.status_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetLinkState_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetLinkState_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetLinkState_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetLinkState_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetLinkState_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetLinkState_Response');
      }
      if (!this._status_messageIntialized) {
        throw new TypeError('Invalid argument: status_message in GetLinkState_Response');
      }
    }
    this._wrapperFields.link_state.freeze(own, checkConsistency);
    this._refObject.link_state = this._wrapperFields.link_state.refObject;
    if (own) {
      this._wrapperFields.status_message.freeze(own, checkConsistency);
    }
    this._refObject.status_message = this._wrapperFields.status_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.link_state.copyRefObject(refObject.link_state);
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = refObject.status_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    LinkStateWrapper.freeStruct(refObject.link_state);
    StringWrapper.freeStruct(refObject.status_message);
  }
  static destoryRawROS(msg) {
    GetLinkState_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'GetLinkState_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link_state() {
    return this._wrapperFields.link_state;
  }
  set link_state(value) {
    if (value instanceof LinkStateWrapper) {
      this._wrapperFields.link_state.copy(value);
    } else {
      this._wrapperFields.link_state.copy(new LinkStateWrapper(value));
    }
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status_message() {
    if (!this._status_messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.status_message.data;
  }
  set status_message(value) {
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetLinkState_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.link_state.copyRefObject(this._refObject.link_state);
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copyRefObject(this._refObject.status_message);
  }
  copy(other) {
    this._refObject = new GetLinkState_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.link_state.copy(other._wrapperFields.link_state);
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copy(other._wrapperFields.status_message);
  }
  static get classType() {
    return GetLinkState_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link_state","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"gazebo_msgs","type":"LinkState","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"GetLinkState_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetLinkState_Response"};
  }
  hasMember(name) {
    let memberNames = ["link_state","success","status_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetLinkState_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetLinkState_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetLinkState_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLinkState_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetLinkState_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetLinkState_ResponseRefArray(size);
    this._refObject = new GetLinkState_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetLinkState_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetLinkState_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetLinkState_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetLinkState_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetLinkState_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetLinkState_ResponseArrayWrapper;
  }
}
module.exports = GetLinkState_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "link_state",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "gazebo_msgs",
        "type": "LinkState",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "GetLinkState_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetLinkState_Response"
}
*/
