// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__BodyRequest_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const BodyRequest_RequestRefStruct = StructType({
  body_name: primitiveTypes.string,
});
const BodyRequest_RequestRefArray = ArrayType(BodyRequest_RequestRefStruct);
const BodyRequest_RequestRefStructArray = StructType({
  data: BodyRequest_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class BodyRequest_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new BodyRequest_RequestRefStruct(other._refObject.toObject());
      this._body_nameIntialized = true;
      this._wrapperFields.body_name =  new StringWrapper(other._wrapperFields.body_name);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new BodyRequest_RequestRefStruct();
    this._body_nameIntialized = false;
    this._wrapperFields.body_name = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new BodyRequest_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new BodyRequest_RequestArrayWrapper;
  }
  static get ArrayType() {
    return BodyRequest_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return BodyRequest_RequestRefStructArray
  }
  static get refObjectType() {
    return BodyRequest_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._body_nameIntialized) {
        throw new TypeError('Invalid argument: body_name in BodyRequest_Request');
      }
    }
    if (own) {
      this._wrapperFields.body_name.freeze(own, checkConsistency);
    }
    this._refObject.body_name = this._wrapperFields.body_name.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._body_nameIntialized = true;
    this._wrapperFields.body_name.data = refObject.body_name.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.body_name);
  }
  static destoryRawROS(msg) {
    BodyRequest_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'BodyRequest_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get body_name() {
    if (!this._body_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.body_name.data;
  }
  set body_name(value) {
    this._body_nameIntialized = true;
    this._wrapperFields.body_name.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new BodyRequest_RequestRefStruct(refObject.toObject());
    this._body_nameIntialized = true;
    this._wrapperFields.body_name.copyRefObject(this._refObject.body_name);
  }
  copy(other) {
    this._refObject = new BodyRequest_RequestRefStruct(other._refObject.toObject());
    this._body_nameIntialized = true;
    this._wrapperFields.body_name.copy(other._wrapperFields.body_name);
  }
  static get classType() {
    return BodyRequest_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"body_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"BodyRequest_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"BodyRequest_Request"};
  }
  hasMember(name) {
    let memberNames = ["body_name"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class BodyRequest_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof BodyRequest_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new BodyRequest_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BodyRequest_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to BodyRequest_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new BodyRequest_RequestRefArray(size);
    this._refObject = new BodyRequest_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new BodyRequest_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof BodyRequest_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "BodyRequest_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      BodyRequest_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return BodyRequest_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return BodyRequest_RequestArrayWrapper;
  }
}
module.exports = BodyRequest_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "body_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "BodyRequest_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "BodyRequest_Request"
}
*/
