// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__SetJointProperties_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ODEJointPropertiesWrapper = require('../../generated/gazebo_msgs/gazebo_msgs__msg__ODEJointProperties.js');
const SetJointProperties_RequestRefStruct = StructType({
  joint_name: primitiveTypes.string,
  ode_joint_config: ODEJointPropertiesWrapper.refObjectType,
});
const SetJointProperties_RequestRefArray = ArrayType(SetJointProperties_RequestRefStruct);
const SetJointProperties_RequestRefStructArray = StructType({
  data: SetJointProperties_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetJointProperties_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetJointProperties_RequestRefStruct(other._refObject.toObject());
      this._joint_nameIntialized = true;
      this._wrapperFields.joint_name =  new StringWrapper(other._wrapperFields.joint_name);
      this._wrapperFields.ode_joint_config =  new ODEJointPropertiesWrapper(other._wrapperFields.ode_joint_config);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetJointProperties_RequestRefStruct();
    this._joint_nameIntialized = false;
    this._wrapperFields.joint_name = new StringWrapper();
    this._wrapperFields.ode_joint_config = new ODEJointPropertiesWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetJointProperties_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetJointProperties_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetJointProperties_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetJointProperties_RequestRefStructArray
  }
  static get refObjectType() {
    return SetJointProperties_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._joint_nameIntialized) {
        throw new TypeError('Invalid argument: joint_name in SetJointProperties_Request');
      }
    }
    if (own) {
      this._wrapperFields.joint_name.freeze(own, checkConsistency);
    }
    this._refObject.joint_name = this._wrapperFields.joint_name.refObject;
    this._wrapperFields.ode_joint_config.freeze(own, checkConsistency);
    this._refObject.ode_joint_config = this._wrapperFields.ode_joint_config.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._joint_nameIntialized = true;
    this._wrapperFields.joint_name.data = refObject.joint_name.data;
    this._wrapperFields.ode_joint_config.copyRefObject(refObject.ode_joint_config);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.joint_name);
    ODEJointPropertiesWrapper.freeStruct(refObject.ode_joint_config);
  }
  static destoryRawROS(msg) {
    SetJointProperties_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'SetJointProperties_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get joint_name() {
    if (!this._joint_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.joint_name.data;
  }
  set joint_name(value) {
    this._joint_nameIntialized = true;
    this._wrapperFields.joint_name.data = value;
  }
  get ode_joint_config() {
    return this._wrapperFields.ode_joint_config;
  }
  set ode_joint_config(value) {
    if (value instanceof ODEJointPropertiesWrapper) {
      this._wrapperFields.ode_joint_config.copy(value);
    } else {
      this._wrapperFields.ode_joint_config.copy(new ODEJointPropertiesWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new SetJointProperties_RequestRefStruct(refObject.toObject());
    this._joint_nameIntialized = true;
    this._wrapperFields.joint_name.copyRefObject(this._refObject.joint_name);
    this._wrapperFields.ode_joint_config.copyRefObject(this._refObject.ode_joint_config);
  }
  copy(other) {
    this._refObject = new SetJointProperties_RequestRefStruct(other._refObject.toObject());
    this._joint_nameIntialized = true;
    this._wrapperFields.joint_name.copy(other._wrapperFields.joint_name);
    this._wrapperFields.ode_joint_config.copy(other._wrapperFields.ode_joint_config);
  }
  static get classType() {
    return SetJointProperties_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"joint_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ode_joint_config","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"gazebo_msgs","type":"ODEJointProperties","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SetJointProperties_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetJointProperties_Request"};
  }
  hasMember(name) {
    let memberNames = ["joint_name","ode_joint_config"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetJointProperties_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetJointProperties_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetJointProperties_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetJointProperties_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetJointProperties_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetJointProperties_RequestRefArray(size);
    this._refObject = new SetJointProperties_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetJointProperties_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetJointProperties_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetJointProperties_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetJointProperties_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetJointProperties_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetJointProperties_RequestArrayWrapper;
  }
}
module.exports = SetJointProperties_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "joint_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ode_joint_config",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "gazebo_msgs",
        "type": "ODEJointProperties",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SetJointProperties_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetJointProperties_Request"
}
*/
