// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__SetLinkProperties_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SetLinkProperties_RequestRefStruct = StructType({
  link_name: primitiveTypes.string,
  com: PoseWrapper.refObjectType,
  gravity_mode: primitiveTypes.bool,
  mass: primitiveTypes.float64,
  ixx: primitiveTypes.float64,
  ixy: primitiveTypes.float64,
  ixz: primitiveTypes.float64,
  iyy: primitiveTypes.float64,
  iyz: primitiveTypes.float64,
  izz: primitiveTypes.float64,
});
const SetLinkProperties_RequestRefArray = ArrayType(SetLinkProperties_RequestRefStruct);
const SetLinkProperties_RequestRefStructArray = StructType({
  data: SetLinkProperties_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetLinkProperties_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetLinkProperties_RequestRefStruct(other._refObject.toObject());
      this._link_nameIntialized = true;
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
      this._wrapperFields.com =  new PoseWrapper(other._wrapperFields.com);
      this._gravity_modeIntialized = true;
      this._massIntialized = true;
      this._ixxIntialized = true;
      this._ixyIntialized = true;
      this._ixzIntialized = true;
      this._iyyIntialized = true;
      this._iyzIntialized = true;
      this._izzIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetLinkProperties_RequestRefStruct();
    this._link_nameIntialized = false;
    this._wrapperFields.link_name = new StringWrapper();
    this._wrapperFields.com = new PoseWrapper();
    this._gravity_modeIntialized = false;
    this._massIntialized = false;
    this._ixxIntialized = false;
    this._ixyIntialized = false;
    this._ixzIntialized = false;
    this._iyyIntialized = false;
    this._iyzIntialized = false;
    this._izzIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetLinkProperties_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetLinkProperties_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetLinkProperties_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetLinkProperties_RequestRefStructArray
  }
  static get refObjectType() {
    return SetLinkProperties_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._link_nameIntialized) {
        throw new TypeError('Invalid argument: link_name in SetLinkProperties_Request');
      }
      if (!this._gravity_modeIntialized) {
        throw new TypeError('Invalid argument: gravity_mode in SetLinkProperties_Request');
      }
      if (!this._massIntialized) {
        throw new TypeError('Invalid argument: mass in SetLinkProperties_Request');
      }
      if (!this._ixxIntialized) {
        throw new TypeError('Invalid argument: ixx in SetLinkProperties_Request');
      }
      if (!this._ixyIntialized) {
        throw new TypeError('Invalid argument: ixy in SetLinkProperties_Request');
      }
      if (!this._ixzIntialized) {
        throw new TypeError('Invalid argument: ixz in SetLinkProperties_Request');
      }
      if (!this._iyyIntialized) {
        throw new TypeError('Invalid argument: iyy in SetLinkProperties_Request');
      }
      if (!this._iyzIntialized) {
        throw new TypeError('Invalid argument: iyz in SetLinkProperties_Request');
      }
      if (!this._izzIntialized) {
        throw new TypeError('Invalid argument: izz in SetLinkProperties_Request');
      }
    }
    if (own) {
      this._wrapperFields.link_name.freeze(own, checkConsistency);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
    this._wrapperFields.com.freeze(own, checkConsistency);
    this._refObject.com = this._wrapperFields.com.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._wrapperFields.com.copyRefObject(refObject.com);
    this._gravity_modeIntialized = true;
    this._refObject.gravity_mode = refObject.gravity_mode;
    this._massIntialized = true;
    this._refObject.mass = refObject.mass;
    this._ixxIntialized = true;
    this._refObject.ixx = refObject.ixx;
    this._ixyIntialized = true;
    this._refObject.ixy = refObject.ixy;
    this._ixzIntialized = true;
    this._refObject.ixz = refObject.ixz;
    this._iyyIntialized = true;
    this._refObject.iyy = refObject.iyy;
    this._iyzIntialized = true;
    this._refObject.iyz = refObject.iyz;
    this._izzIntialized = true;
    this._refObject.izz = refObject.izz;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.link_name);
    PoseWrapper.freeStruct(refObject.com);
  }
  static destoryRawROS(msg) {
    SetLinkProperties_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'SetLinkProperties_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link_name() {
    if (!this._link_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = value;
  }
  get com() {
    return this._wrapperFields.com;
  }
  set com(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.com.copy(value);
    } else {
      this._wrapperFields.com.copy(new PoseWrapper(value));
    }
  }
  get gravity_mode() {
    if (!this._gravity_modeIntialized) {
      return undefined;
    }
    return this._refObject.gravity_mode;
  }
  set gravity_mode(value) {
    this._gravity_modeIntialized = true;
    this._refObject.gravity_mode = value;
  }
  get mass() {
    if (!this._massIntialized) {
      return undefined;
    }
    return this._refObject.mass;
  }
  set mass(value) {
    this._massIntialized = true;
    this._refObject.mass = value;
  }
  get ixx() {
    if (!this._ixxIntialized) {
      return undefined;
    }
    return this._refObject.ixx;
  }
  set ixx(value) {
    this._ixxIntialized = true;
    this._refObject.ixx = value;
  }
  get ixy() {
    if (!this._ixyIntialized) {
      return undefined;
    }
    return this._refObject.ixy;
  }
  set ixy(value) {
    this._ixyIntialized = true;
    this._refObject.ixy = value;
  }
  get ixz() {
    if (!this._ixzIntialized) {
      return undefined;
    }
    return this._refObject.ixz;
  }
  set ixz(value) {
    this._ixzIntialized = true;
    this._refObject.ixz = value;
  }
  get iyy() {
    if (!this._iyyIntialized) {
      return undefined;
    }
    return this._refObject.iyy;
  }
  set iyy(value) {
    this._iyyIntialized = true;
    this._refObject.iyy = value;
  }
  get iyz() {
    if (!this._iyzIntialized) {
      return undefined;
    }
    return this._refObject.iyz;
  }
  set iyz(value) {
    this._iyzIntialized = true;
    this._refObject.iyz = value;
  }
  get izz() {
    if (!this._izzIntialized) {
      return undefined;
    }
    return this._refObject.izz;
  }
  set izz(value) {
    this._izzIntialized = true;
    this._refObject.izz = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetLinkProperties_RequestRefStruct(refObject.toObject());
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
    this._wrapperFields.com.copyRefObject(this._refObject.com);
    this._gravity_modeIntialized = true;
    this._massIntialized = true;
    this._ixxIntialized = true;
    this._ixyIntialized = true;
    this._ixzIntialized = true;
    this._iyyIntialized = true;
    this._iyzIntialized = true;
    this._izzIntialized = true;
  }
  copy(other) {
    this._refObject = new SetLinkProperties_RequestRefStruct(other._refObject.toObject());
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
    this._wrapperFields.com.copy(other._wrapperFields.com);
    this._gravity_modeIntialized = true;
    this._massIntialized = true;
    this._ixxIntialized = true;
    this._ixyIntialized = true;
    this._ixzIntialized = true;
    this._iyyIntialized = true;
    this._iyzIntialized = true;
    this._izzIntialized = true;
  }
  static get classType() {
    return SetLinkProperties_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"com","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"gravity_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"mass","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixx","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ixz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyy","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"iyz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"izz","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SetLinkProperties_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetLinkProperties_Request"};
  }
  hasMember(name) {
    let memberNames = ["link_name","com","gravity_mode","mass","ixx","ixy","ixz","iyy","iyz","izz"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetLinkProperties_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetLinkProperties_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetLinkProperties_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLinkProperties_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetLinkProperties_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetLinkProperties_RequestRefArray(size);
    this._refObject = new SetLinkProperties_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetLinkProperties_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetLinkProperties_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetLinkProperties_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetLinkProperties_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetLinkProperties_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetLinkProperties_RequestArrayWrapper;
  }
}
module.exports = SetLinkProperties_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "link_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "com",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "gravity_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "mass",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixx",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ixz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iyy",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "iyz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "izz",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SetLinkProperties_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetLinkProperties_Request"
}
*/
