// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__LinkStates.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const LinkStatesRefStruct = StructType({
  name: StringWrapper.refObjectArrayType,
  pose: PoseWrapper.refObjectArrayType,
  twist: TwistWrapper.refObjectArrayType,
});
const LinkStatesRefArray = ArrayType(LinkStatesRefStruct);
const LinkStatesRefStructArray = StructType({
  data: LinkStatesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LinkStatesWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._nameArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LinkStatesRefStruct(other._refObject.toObject());
      this._wrapperFields.name = StringWrapper.createArray();
      this._wrapperFields.name.copy(other._wrapperFields.name);
      this.name = other.name;
      this._wrapperFields.pose = PoseWrapper.createArray();
      this._wrapperFields.pose.copy(other._wrapperFields.pose);
      this._wrapperFields.twist = TwistWrapper.createArray();
      this._wrapperFields.twist.copy(other._wrapperFields.twist);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LinkStatesRefStruct();
    this._wrapperFields.name = StringWrapper.createArray();
    this._wrapperFields.pose = PoseWrapper.createArray();
    this._wrapperFields.twist = TwistWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LinkStatesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LinkStatesArrayWrapper;
  }
  static get ArrayType() {
    return LinkStatesArrayWrapper;
  }
  static get refObjectArrayType() {
    return LinkStatesRefStructArray
  }
  static get refObjectType() {
    return LinkStatesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.name.fill(this._nameArray);
      this._wrapperFields.name.freeze(own, checkConsistency);
      this._refObject.name = this._wrapperFields.name.refObject;
    } else {
      this._wrapperFields.name.fill([]);
      this._wrapperFields.name.freeze(own, checkConsistency);
      this._refObject.name = this._wrapperFields.name.refObject;
    }
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.twist.freeze(own, checkConsistency);
    this._refObject.twist = this._wrapperFields.twist.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.twist.copyRefObject(refObject.twist);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.name.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.name);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.name, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.pose.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.pose);
      if (PoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.pose, PoseWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.twist.size != 0) {
      TwistWrapper.ArrayType.freeArray(refObject.twist);
      if (TwistWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.twist, TwistWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LinkStatesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'LinkStates'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    return this._nameArray;
  }
  set name(value) {
    this._nameArray = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    this._wrapperFields.pose.fill(value);
  }
  get twist() {
    return this._wrapperFields.twist;
  }
  set twist(value) {
    this._wrapperFields.twist.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LinkStatesRefStruct(refObject.toObject());
    refObject.name.data.length = refObject.name.size;
    for (let index = 0; index < refObject.name.size; index++) {
      this._nameArray[index] = refObject.name.data[index].data;
    }
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.twist.copyRefObject(this._refObject.twist);
  }
  copy(other) {
    this._refObject = new LinkStatesRefStruct(other._refObject.toObject());
    this._nameArray = other._nameArray.slice();
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.twist.copy(other._wrapperFields.twist);
  }
  static get classType() {
    return LinkStatesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twist","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"LinkStates","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LinkStates"};
  }
  hasMember(name) {
    let memberNames = ["name","pose","twist"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LinkStatesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LinkStatesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LinkStatesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkStatesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkStatesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LinkStatesRefArray(size);
    this._refObject = new LinkStatesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LinkStatesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LinkStatesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LinkStatesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LinkStatesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LinkStatesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LinkStatesArrayWrapper;
  }
}
module.exports = LinkStatesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twist",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "LinkStates",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LinkStates"
}
*/
