// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__ApplyLinkWrench_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PointWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Point.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const ApplyLinkWrench_RequestRefStruct = StructType({
  link_name: primitiveTypes.string,
  reference_frame: primitiveTypes.string,
  reference_point: PointWrapper.refObjectType,
  wrench: WrenchWrapper.refObjectType,
  start_time: TimeWrapper.refObjectType,
  duration: DurationWrapper.refObjectType,
});
const ApplyLinkWrench_RequestRefArray = ArrayType(ApplyLinkWrench_RequestRefStruct);
const ApplyLinkWrench_RequestRefStructArray = StructType({
  data: ApplyLinkWrench_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ApplyLinkWrench_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ApplyLinkWrench_RequestRefStruct(other._refObject.toObject());
      this._link_nameIntialized = true;
      this._wrapperFields.link_name =  new StringWrapper(other._wrapperFields.link_name);
      this._reference_frameIntialized = true;
      this._wrapperFields.reference_frame =  new StringWrapper(other._wrapperFields.reference_frame);
      this._wrapperFields.reference_point =  new PointWrapper(other._wrapperFields.reference_point);
      this._wrapperFields.wrench =  new WrenchWrapper(other._wrapperFields.wrench);
      this._wrapperFields.start_time =  new TimeWrapper(other._wrapperFields.start_time);
      this._wrapperFields.duration =  new DurationWrapper(other._wrapperFields.duration);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ApplyLinkWrench_RequestRefStruct();
    this._link_nameIntialized = false;
    this._wrapperFields.link_name = new StringWrapper();
    this._reference_frameIntialized = false;
    this._wrapperFields.reference_frame = new StringWrapper();
    this._wrapperFields.reference_point = new PointWrapper();
    this._wrapperFields.wrench = new WrenchWrapper();
    this._wrapperFields.start_time = new TimeWrapper();
    this._wrapperFields.duration = new DurationWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ApplyLinkWrench_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ApplyLinkWrench_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ApplyLinkWrench_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ApplyLinkWrench_RequestRefStructArray
  }
  static get refObjectType() {
    return ApplyLinkWrench_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._link_nameIntialized) {
        throw new TypeError('Invalid argument: link_name in ApplyLinkWrench_Request');
      }
      if (!this._reference_frameIntialized) {
        throw new TypeError('Invalid argument: reference_frame in ApplyLinkWrench_Request');
      }
    }
    if (own) {
      this._wrapperFields.link_name.freeze(own, checkConsistency);
    }
    this._refObject.link_name = this._wrapperFields.link_name.refObject;
    if (own) {
      this._wrapperFields.reference_frame.freeze(own, checkConsistency);
    }
    this._refObject.reference_frame = this._wrapperFields.reference_frame.refObject;
    this._wrapperFields.reference_point.freeze(own, checkConsistency);
    this._refObject.reference_point = this._wrapperFields.reference_point.refObject;
    this._wrapperFields.wrench.freeze(own, checkConsistency);
    this._refObject.wrench = this._wrapperFields.wrench.refObject;
    this._wrapperFields.start_time.freeze(own, checkConsistency);
    this._refObject.start_time = this._wrapperFields.start_time.refObject;
    this._wrapperFields.duration.freeze(own, checkConsistency);
    this._refObject.duration = this._wrapperFields.duration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = refObject.link_name.data;
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.data = refObject.reference_frame.data;
    this._wrapperFields.reference_point.copyRefObject(refObject.reference_point);
    this._wrapperFields.wrench.copyRefObject(refObject.wrench);
    this._wrapperFields.start_time.copyRefObject(refObject.start_time);
    this._wrapperFields.duration.copyRefObject(refObject.duration);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.link_name);
    StringWrapper.freeStruct(refObject.reference_frame);
    PointWrapper.freeStruct(refObject.reference_point);
    WrenchWrapper.freeStruct(refObject.wrench);
    TimeWrapper.freeStruct(refObject.start_time);
    DurationWrapper.freeStruct(refObject.duration);
  }
  static destoryRawROS(msg) {
    ApplyLinkWrench_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'ApplyLinkWrench_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get link_name() {
    if (!this._link_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.link_name.data;
  }
  set link_name(value) {
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.data = value;
  }
  get reference_frame() {
    if (!this._reference_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.reference_frame.data;
  }
  set reference_frame(value) {
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.data = value;
  }
  get reference_point() {
    return this._wrapperFields.reference_point;
  }
  set reference_point(value) {
    if (value instanceof PointWrapper) {
      this._wrapperFields.reference_point.copy(value);
    } else {
      this._wrapperFields.reference_point.copy(new PointWrapper(value));
    }
  }
  get wrench() {
    return this._wrapperFields.wrench;
  }
  set wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.wrench.copy(value);
    } else {
      this._wrapperFields.wrench.copy(new WrenchWrapper(value));
    }
  }
  get start_time() {
    return this._wrapperFields.start_time;
  }
  set start_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.start_time.copy(value);
    } else {
      this._wrapperFields.start_time.copy(new TimeWrapper(value));
    }
  }
  get duration() {
    return this._wrapperFields.duration;
  }
  set duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.duration.copy(value);
    } else {
      this._wrapperFields.duration.copy(new DurationWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ApplyLinkWrench_RequestRefStruct(refObject.toObject());
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copyRefObject(this._refObject.link_name);
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.copyRefObject(this._refObject.reference_frame);
    this._wrapperFields.reference_point.copyRefObject(this._refObject.reference_point);
    this._wrapperFields.wrench.copyRefObject(this._refObject.wrench);
    this._wrapperFields.start_time.copyRefObject(this._refObject.start_time);
    this._wrapperFields.duration.copyRefObject(this._refObject.duration);
  }
  copy(other) {
    this._refObject = new ApplyLinkWrench_RequestRefStruct(other._refObject.toObject());
    this._link_nameIntialized = true;
    this._wrapperFields.link_name.copy(other._wrapperFields.link_name);
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.copy(other._wrapperFields.reference_frame);
    this._wrapperFields.reference_point.copy(other._wrapperFields.reference_point);
    this._wrapperFields.wrench.copy(other._wrapperFields.wrench);
    this._wrapperFields.start_time.copy(other._wrapperFields.start_time);
    this._wrapperFields.duration.copy(other._wrapperFields.duration);
  }
  static get classType() {
    return ApplyLinkWrench_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"link_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"reference_point","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Point","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"ApplyLinkWrench_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ApplyLinkWrench_Request"};
  }
  hasMember(name) {
    let memberNames = ["link_name","reference_frame","reference_point","wrench","start_time","duration"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ApplyLinkWrench_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ApplyLinkWrench_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ApplyLinkWrench_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplyLinkWrench_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplyLinkWrench_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ApplyLinkWrench_RequestRefArray(size);
    this._refObject = new ApplyLinkWrench_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ApplyLinkWrench_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ApplyLinkWrench_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ApplyLinkWrench_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ApplyLinkWrench_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ApplyLinkWrench_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ApplyLinkWrench_RequestArrayWrapper;
  }
}
module.exports = ApplyLinkWrench_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "link_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reference_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "reference_point",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Point",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "wrench",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "ApplyLinkWrench_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ApplyLinkWrench_Request"
}
*/
