// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__SensorPerformanceMetric.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SensorPerformanceMetricRefStruct = StructType({
  name: primitiveTypes.string,
  sim_update_rate: primitiveTypes.float64,
  real_update_rate: primitiveTypes.float64,
  fps: primitiveTypes.float64,
});
const SensorPerformanceMetricRefArray = ArrayType(SensorPerformanceMetricRefStruct);
const SensorPerformanceMetricRefStructArray = StructType({
  data: SensorPerformanceMetricRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SensorPerformanceMetricWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SensorPerformanceMetricRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._sim_update_rateIntialized = true;
      this._real_update_rateIntialized = true;
      this._fpsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SensorPerformanceMetricRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._sim_update_rateIntialized = false;
    this._real_update_rateIntialized = false;
    this._fpsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SensorPerformanceMetricWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SensorPerformanceMetricArrayWrapper;
  }
  static get ArrayType() {
    return SensorPerformanceMetricArrayWrapper;
  }
  static get refObjectArrayType() {
    return SensorPerformanceMetricRefStructArray
  }
  static get refObjectType() {
    return SensorPerformanceMetricRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._nameIntialized) {
        throw new TypeError('Invalid argument: name in SensorPerformanceMetric');
      }
      if (!this._sim_update_rateIntialized) {
        throw new TypeError('Invalid argument: sim_update_rate in SensorPerformanceMetric');
      }
      if (!this._real_update_rateIntialized) {
        throw new TypeError('Invalid argument: real_update_rate in SensorPerformanceMetric');
      }
      if (!this._fpsIntialized) {
        throw new TypeError('Invalid argument: fps in SensorPerformanceMetric');
      }
    }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._sim_update_rateIntialized = true;
    this._refObject.sim_update_rate = refObject.sim_update_rate;
    this._real_update_rateIntialized = true;
    this._refObject.real_update_rate = refObject.real_update_rate;
    this._fpsIntialized = true;
    this._refObject.fps = refObject.fps;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destoryRawROS(msg) {
    SensorPerformanceMetricWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'SensorPerformanceMetric'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get sim_update_rate() {
    if (!this._sim_update_rateIntialized) {
      return undefined;
    }
    return this._refObject.sim_update_rate;
  }
  set sim_update_rate(value) {
    this._sim_update_rateIntialized = true;
    this._refObject.sim_update_rate = value;
  }
  get real_update_rate() {
    if (!this._real_update_rateIntialized) {
      return undefined;
    }
    return this._refObject.real_update_rate;
  }
  set real_update_rate(value) {
    this._real_update_rateIntialized = true;
    this._refObject.real_update_rate = value;
  }
  get fps() {
    if (!this._fpsIntialized) {
      return undefined;
    }
    return this._refObject.fps;
  }
  set fps(value) {
    this._fpsIntialized = true;
    this._refObject.fps = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SensorPerformanceMetricRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._sim_update_rateIntialized = true;
    this._real_update_rateIntialized = true;
    this._fpsIntialized = true;
  }
  copy(other) {
    this._refObject = new SensorPerformanceMetricRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._sim_update_rateIntialized = true;
    this._real_update_rateIntialized = true;
    this._fpsIntialized = true;
  }
  static get classType() {
    return SensorPerformanceMetricWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sim_update_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"real_update_rate","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fps","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SensorPerformanceMetric","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SensorPerformanceMetric"};
  }
  hasMember(name) {
    let memberNames = ["name","sim_update_rate","real_update_rate","fps"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SensorPerformanceMetricArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SensorPerformanceMetricWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SensorPerformanceMetricWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorPerformanceMetricArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SensorPerformanceMetricArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SensorPerformanceMetricRefArray(size);
    this._refObject = new SensorPerformanceMetricRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SensorPerformanceMetricWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SensorPerformanceMetricArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SensorPerformanceMetricArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SensorPerformanceMetricWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SensorPerformanceMetricWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SensorPerformanceMetricArrayWrapper;
  }
}
module.exports = SensorPerformanceMetricWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sim_update_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "real_update_rate",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fps",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SensorPerformanceMetric",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SensorPerformanceMetric"
}
*/
