// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__ModelState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const TwistWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Twist.js');
const ModelStateRefStruct = StructType({
  model_name: primitiveTypes.string,
  pose: PoseWrapper.refObjectType,
  twist: TwistWrapper.refObjectType,
  reference_frame: primitiveTypes.string,
});
const ModelStateRefArray = ArrayType(ModelStateRefStruct);
const ModelStateRefStructArray = StructType({
  data: ModelStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ModelStateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ModelStateRefStruct(other._refObject.toObject());
      this._model_nameIntialized = true;
      this._wrapperFields.model_name =  new StringWrapper(other._wrapperFields.model_name);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._wrapperFields.twist =  new TwistWrapper(other._wrapperFields.twist);
      this._reference_frameIntialized = true;
      this._wrapperFields.reference_frame =  new StringWrapper(other._wrapperFields.reference_frame);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ModelStateRefStruct();
    this._model_nameIntialized = false;
    this._wrapperFields.model_name = new StringWrapper();
    this._wrapperFields.pose = new PoseWrapper();
    this._wrapperFields.twist = new TwistWrapper();
    this._reference_frameIntialized = false;
    this._wrapperFields.reference_frame = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ModelStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ModelStateArrayWrapper;
  }
  static get ArrayType() {
    return ModelStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return ModelStateRefStructArray
  }
  static get refObjectType() {
    return ModelStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._model_nameIntialized) {
        throw new TypeError('Invalid argument: model_name in ModelState');
      }
      if (!this._reference_frameIntialized) {
        throw new TypeError('Invalid argument: reference_frame in ModelState');
      }
    }
    if (own) {
      this._wrapperFields.model_name.freeze(own, checkConsistency);
    }
    this._refObject.model_name = this._wrapperFields.model_name.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.twist.freeze(own, checkConsistency);
    this._refObject.twist = this._wrapperFields.twist.refObject;
    if (own) {
      this._wrapperFields.reference_frame.freeze(own, checkConsistency);
    }
    this._refObject.reference_frame = this._wrapperFields.reference_frame.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = refObject.model_name.data;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._wrapperFields.twist.copyRefObject(refObject.twist);
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.data = refObject.reference_frame.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.model_name);
    PoseWrapper.freeStruct(refObject.pose);
    TwistWrapper.freeStruct(refObject.twist);
    StringWrapper.freeStruct(refObject.reference_frame);
  }
  static destoryRawROS(msg) {
    ModelStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'ModelState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get model_name() {
    if (!this._model_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.model_name.data;
  }
  set model_name(value) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get twist() {
    return this._wrapperFields.twist;
  }
  set twist(value) {
    if (value instanceof TwistWrapper) {
      this._wrapperFields.twist.copy(value);
    } else {
      this._wrapperFields.twist.copy(new TwistWrapper(value));
    }
  }
  get reference_frame() {
    if (!this._reference_frameIntialized) {
      return undefined;
    }
    return this._wrapperFields.reference_frame.data;
  }
  set reference_frame(value) {
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ModelStateRefStruct(refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copyRefObject(this._refObject.model_name);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._wrapperFields.twist.copyRefObject(this._refObject.twist);
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.copyRefObject(this._refObject.reference_frame);
  }
  copy(other) {
    this._refObject = new ModelStateRefStruct(other._refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copy(other._wrapperFields.model_name);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._wrapperFields.twist.copy(other._wrapperFields.twist);
    this._reference_frameIntialized = true;
    this._wrapperFields.reference_frame.copy(other._wrapperFields.reference_frame);
  }
  static get classType() {
    return ModelStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"model_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"twist","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Twist","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"reference_frame","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"ModelState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ModelState"};
  }
  hasMember(name) {
    let memberNames = ["model_name","pose","twist","reference_frame"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ModelStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ModelStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ModelStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ModelStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ModelStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ModelStateRefArray(size);
    this._refObject = new ModelStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ModelStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ModelStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ModelStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ModelStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ModelStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ModelStateArrayWrapper;
  }
}
module.exports = ModelStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "model_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "twist",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Twist",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "reference_frame",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "ModelState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ModelState"
}
*/
