// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__ODEJointProperties.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const ODEJointPropertiesRefStruct = StructType({
  damping: Float64Wrapper.refObjectArrayType,
  hi_stop: Float64Wrapper.refObjectArrayType,
  lo_stop: Float64Wrapper.refObjectArrayType,
  erp: Float64Wrapper.refObjectArrayType,
  cfm: Float64Wrapper.refObjectArrayType,
  stop_erp: Float64Wrapper.refObjectArrayType,
  stop_cfm: Float64Wrapper.refObjectArrayType,
  fudge_factor: Float64Wrapper.refObjectArrayType,
  fmax: Float64Wrapper.refObjectArrayType,
  vel: Float64Wrapper.refObjectArrayType,
});
const ODEJointPropertiesRefArray = ArrayType(ODEJointPropertiesRefStruct);
const ODEJointPropertiesRefStructArray = StructType({
  data: ODEJointPropertiesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ODEJointPropertiesWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ODEJointPropertiesRefStruct(other._refObject.toObject());
      this._wrapperFields.damping = Float64Wrapper.createArray();
      this._wrapperFields.damping.copy(other._wrapperFields.damping);
      this._wrapperFields.hi_stop = Float64Wrapper.createArray();
      this._wrapperFields.hi_stop.copy(other._wrapperFields.hi_stop);
      this._wrapperFields.lo_stop = Float64Wrapper.createArray();
      this._wrapperFields.lo_stop.copy(other._wrapperFields.lo_stop);
      this._wrapperFields.erp = Float64Wrapper.createArray();
      this._wrapperFields.erp.copy(other._wrapperFields.erp);
      this._wrapperFields.cfm = Float64Wrapper.createArray();
      this._wrapperFields.cfm.copy(other._wrapperFields.cfm);
      this._wrapperFields.stop_erp = Float64Wrapper.createArray();
      this._wrapperFields.stop_erp.copy(other._wrapperFields.stop_erp);
      this._wrapperFields.stop_cfm = Float64Wrapper.createArray();
      this._wrapperFields.stop_cfm.copy(other._wrapperFields.stop_cfm);
      this._wrapperFields.fudge_factor = Float64Wrapper.createArray();
      this._wrapperFields.fudge_factor.copy(other._wrapperFields.fudge_factor);
      this._wrapperFields.fmax = Float64Wrapper.createArray();
      this._wrapperFields.fmax.copy(other._wrapperFields.fmax);
      this._wrapperFields.vel = Float64Wrapper.createArray();
      this._wrapperFields.vel.copy(other._wrapperFields.vel);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ODEJointPropertiesRefStruct();
    this._wrapperFields.damping = Float64Wrapper.createArray();
    this._wrapperFields.hi_stop = Float64Wrapper.createArray();
    this._wrapperFields.lo_stop = Float64Wrapper.createArray();
    this._wrapperFields.erp = Float64Wrapper.createArray();
    this._wrapperFields.cfm = Float64Wrapper.createArray();
    this._wrapperFields.stop_erp = Float64Wrapper.createArray();
    this._wrapperFields.stop_cfm = Float64Wrapper.createArray();
    this._wrapperFields.fudge_factor = Float64Wrapper.createArray();
    this._wrapperFields.fmax = Float64Wrapper.createArray();
    this._wrapperFields.vel = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ODEJointPropertiesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ODEJointPropertiesArrayWrapper;
  }
  static get ArrayType() {
    return ODEJointPropertiesArrayWrapper;
  }
  static get refObjectArrayType() {
    return ODEJointPropertiesRefStructArray
  }
  static get refObjectType() {
    return ODEJointPropertiesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.damping.freeze(own, checkConsistency);
    this._refObject.damping = this._wrapperFields.damping.refObject;
    if (own) {
      this._wrapperFields.damping.fill(Float64Array.from([]));
      this._wrapperFields.damping.freeze(own, checkConsistency);
      this._refObject.damping = this._wrapperFields.damping.refObject;
    }
    this._wrapperFields.hi_stop.freeze(own, checkConsistency);
    this._refObject.hi_stop = this._wrapperFields.hi_stop.refObject;
    if (own) {
      this._wrapperFields.hi_stop.fill(Float64Array.from([]));
      this._wrapperFields.hi_stop.freeze(own, checkConsistency);
      this._refObject.hi_stop = this._wrapperFields.hi_stop.refObject;
    }
    this._wrapperFields.lo_stop.freeze(own, checkConsistency);
    this._refObject.lo_stop = this._wrapperFields.lo_stop.refObject;
    if (own) {
      this._wrapperFields.lo_stop.fill(Float64Array.from([]));
      this._wrapperFields.lo_stop.freeze(own, checkConsistency);
      this._refObject.lo_stop = this._wrapperFields.lo_stop.refObject;
    }
    this._wrapperFields.erp.freeze(own, checkConsistency);
    this._refObject.erp = this._wrapperFields.erp.refObject;
    if (own) {
      this._wrapperFields.erp.fill(Float64Array.from([]));
      this._wrapperFields.erp.freeze(own, checkConsistency);
      this._refObject.erp = this._wrapperFields.erp.refObject;
    }
    this._wrapperFields.cfm.freeze(own, checkConsistency);
    this._refObject.cfm = this._wrapperFields.cfm.refObject;
    if (own) {
      this._wrapperFields.cfm.fill(Float64Array.from([]));
      this._wrapperFields.cfm.freeze(own, checkConsistency);
      this._refObject.cfm = this._wrapperFields.cfm.refObject;
    }
    this._wrapperFields.stop_erp.freeze(own, checkConsistency);
    this._refObject.stop_erp = this._wrapperFields.stop_erp.refObject;
    if (own) {
      this._wrapperFields.stop_erp.fill(Float64Array.from([]));
      this._wrapperFields.stop_erp.freeze(own, checkConsistency);
      this._refObject.stop_erp = this._wrapperFields.stop_erp.refObject;
    }
    this._wrapperFields.stop_cfm.freeze(own, checkConsistency);
    this._refObject.stop_cfm = this._wrapperFields.stop_cfm.refObject;
    if (own) {
      this._wrapperFields.stop_cfm.fill(Float64Array.from([]));
      this._wrapperFields.stop_cfm.freeze(own, checkConsistency);
      this._refObject.stop_cfm = this._wrapperFields.stop_cfm.refObject;
    }
    this._wrapperFields.fudge_factor.freeze(own, checkConsistency);
    this._refObject.fudge_factor = this._wrapperFields.fudge_factor.refObject;
    if (own) {
      this._wrapperFields.fudge_factor.fill(Float64Array.from([]));
      this._wrapperFields.fudge_factor.freeze(own, checkConsistency);
      this._refObject.fudge_factor = this._wrapperFields.fudge_factor.refObject;
    }
    this._wrapperFields.fmax.freeze(own, checkConsistency);
    this._refObject.fmax = this._wrapperFields.fmax.refObject;
    if (own) {
      this._wrapperFields.fmax.fill(Float64Array.from([]));
      this._wrapperFields.fmax.freeze(own, checkConsistency);
      this._refObject.fmax = this._wrapperFields.fmax.refObject;
    }
    this._wrapperFields.vel.freeze(own, checkConsistency);
    this._refObject.vel = this._wrapperFields.vel.refObject;
    if (own) {
      this._wrapperFields.vel.fill(Float64Array.from([]));
      this._wrapperFields.vel.freeze(own, checkConsistency);
      this._refObject.vel = this._wrapperFields.vel.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.damping.copyRefObject(refObject.damping);
    this._wrapperFields.hi_stop.copyRefObject(refObject.hi_stop);
    this._wrapperFields.lo_stop.copyRefObject(refObject.lo_stop);
    this._wrapperFields.erp.copyRefObject(refObject.erp);
    this._wrapperFields.cfm.copyRefObject(refObject.cfm);
    this._wrapperFields.stop_erp.copyRefObject(refObject.stop_erp);
    this._wrapperFields.stop_cfm.copyRefObject(refObject.stop_cfm);
    this._wrapperFields.fudge_factor.copyRefObject(refObject.fudge_factor);
    this._wrapperFields.fmax.copyRefObject(refObject.fmax);
    this._wrapperFields.vel.copyRefObject(refObject.vel);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.damping.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.damping);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.damping, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.hi_stop.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.hi_stop);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.hi_stop, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.lo_stop.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.lo_stop);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.lo_stop, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.erp.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.erp);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.erp, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.cfm.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.cfm);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.cfm, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.stop_erp.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.stop_erp);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.stop_erp, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.stop_cfm.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.stop_cfm);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.stop_cfm, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.fudge_factor.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.fudge_factor);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fudge_factor, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.fmax.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.fmax);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fmax, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.vel.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.vel);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.vel, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ODEJointPropertiesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'ODEJointProperties'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get damping() {
    return this._wrapperFields['damping'].data;
  }
  set damping(value) {
    this._wrapperFields['damping'].fill(value);
  }
  get hi_stop() {
    return this._wrapperFields['hi_stop'].data;
  }
  set hi_stop(value) {
    this._wrapperFields['hi_stop'].fill(value);
  }
  get lo_stop() {
    return this._wrapperFields['lo_stop'].data;
  }
  set lo_stop(value) {
    this._wrapperFields['lo_stop'].fill(value);
  }
  get erp() {
    return this._wrapperFields['erp'].data;
  }
  set erp(value) {
    this._wrapperFields['erp'].fill(value);
  }
  get cfm() {
    return this._wrapperFields['cfm'].data;
  }
  set cfm(value) {
    this._wrapperFields['cfm'].fill(value);
  }
  get stop_erp() {
    return this._wrapperFields['stop_erp'].data;
  }
  set stop_erp(value) {
    this._wrapperFields['stop_erp'].fill(value);
  }
  get stop_cfm() {
    return this._wrapperFields['stop_cfm'].data;
  }
  set stop_cfm(value) {
    this._wrapperFields['stop_cfm'].fill(value);
  }
  get fudge_factor() {
    return this._wrapperFields['fudge_factor'].data;
  }
  set fudge_factor(value) {
    this._wrapperFields['fudge_factor'].fill(value);
  }
  get fmax() {
    return this._wrapperFields['fmax'].data;
  }
  set fmax(value) {
    this._wrapperFields['fmax'].fill(value);
  }
  get vel() {
    return this._wrapperFields['vel'].data;
  }
  set vel(value) {
    this._wrapperFields['vel'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ODEJointPropertiesRefStruct(refObject.toObject());
    this._wrapperFields.damping.copyRefObject(this._refObject.damping);
    this._wrapperFields.hi_stop.copyRefObject(this._refObject.hi_stop);
    this._wrapperFields.lo_stop.copyRefObject(this._refObject.lo_stop);
    this._wrapperFields.erp.copyRefObject(this._refObject.erp);
    this._wrapperFields.cfm.copyRefObject(this._refObject.cfm);
    this._wrapperFields.stop_erp.copyRefObject(this._refObject.stop_erp);
    this._wrapperFields.stop_cfm.copyRefObject(this._refObject.stop_cfm);
    this._wrapperFields.fudge_factor.copyRefObject(this._refObject.fudge_factor);
    this._wrapperFields.fmax.copyRefObject(this._refObject.fmax);
    this._wrapperFields.vel.copyRefObject(this._refObject.vel);
  }
  copy(other) {
    this._refObject = new ODEJointPropertiesRefStruct(other._refObject.toObject());
    this._wrapperFields.damping.copy(other._wrapperFields.damping);
    this._wrapperFields.hi_stop.copy(other._wrapperFields.hi_stop);
    this._wrapperFields.lo_stop.copy(other._wrapperFields.lo_stop);
    this._wrapperFields.erp.copy(other._wrapperFields.erp);
    this._wrapperFields.cfm.copy(other._wrapperFields.cfm);
    this._wrapperFields.stop_erp.copy(other._wrapperFields.stop_erp);
    this._wrapperFields.stop_cfm.copy(other._wrapperFields.stop_cfm);
    this._wrapperFields.fudge_factor.copy(other._wrapperFields.fudge_factor);
    this._wrapperFields.fmax.copy(other._wrapperFields.fmax);
    this._wrapperFields.vel.copy(other._wrapperFields.vel);
  }
  static get classType() {
    return ODEJointPropertiesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"damping","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"hi_stop","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"lo_stop","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"erp","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cfm","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stop_erp","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"stop_cfm","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fudge_factor","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fmax","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"vel","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"ODEJointProperties","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ODEJointProperties"};
  }
  hasMember(name) {
    let memberNames = ["damping","hi_stop","lo_stop","erp","cfm","stop_erp","stop_cfm","fudge_factor","fmax","vel"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ODEJointPropertiesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ODEJointPropertiesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ODEJointPropertiesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ODEJointPropertiesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ODEJointPropertiesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ODEJointPropertiesRefArray(size);
    this._refObject = new ODEJointPropertiesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ODEJointPropertiesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ODEJointPropertiesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ODEJointPropertiesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ODEJointPropertiesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ODEJointPropertiesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ODEJointPropertiesArrayWrapper;
  }
}
module.exports = ODEJointPropertiesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "damping",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "hi_stop",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "lo_stop",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "erp",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cfm",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stop_erp",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "stop_cfm",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fudge_factor",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fmax",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "vel",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "ODEJointProperties",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ODEJointProperties"
}
*/
