// rclnodejs/generated/gazebo_msgs/gazebo_msgs__msg__ContactState.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const WrenchWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Wrench.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const ContactStateRefStruct = StructType({
  info: primitiveTypes.string,
  collision1_name: primitiveTypes.string,
  collision2_name: primitiveTypes.string,
  wrenches: WrenchWrapper.refObjectArrayType,
  total_wrench: WrenchWrapper.refObjectType,
  contact_positions: Vector3Wrapper.refObjectArrayType,
  contact_normals: Vector3Wrapper.refObjectArrayType,
  depths: Float64Wrapper.refObjectArrayType,
});
const ContactStateRefArray = ArrayType(ContactStateRefStruct);
const ContactStateRefStructArray = StructType({
  data: ContactStateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ContactStateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ContactStateRefStruct(other._refObject.toObject());
      this._infoIntialized = true;
      this._wrapperFields.info =  new StringWrapper(other._wrapperFields.info);
      this._collision1_nameIntialized = true;
      this._wrapperFields.collision1_name =  new StringWrapper(other._wrapperFields.collision1_name);
      this._collision2_nameIntialized = true;
      this._wrapperFields.collision2_name =  new StringWrapper(other._wrapperFields.collision2_name);
      this._wrapperFields.wrenches = WrenchWrapper.createArray();
      this._wrapperFields.wrenches.copy(other._wrapperFields.wrenches);
      this._wrapperFields.total_wrench =  new WrenchWrapper(other._wrapperFields.total_wrench);
      this._wrapperFields.contact_positions = Vector3Wrapper.createArray();
      this._wrapperFields.contact_positions.copy(other._wrapperFields.contact_positions);
      this._wrapperFields.contact_normals = Vector3Wrapper.createArray();
      this._wrapperFields.contact_normals.copy(other._wrapperFields.contact_normals);
      this._wrapperFields.depths = Float64Wrapper.createArray();
      this._wrapperFields.depths.copy(other._wrapperFields.depths);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ContactStateRefStruct();
    this._infoIntialized = false;
    this._wrapperFields.info = new StringWrapper();
    this._collision1_nameIntialized = false;
    this._wrapperFields.collision1_name = new StringWrapper();
    this._collision2_nameIntialized = false;
    this._wrapperFields.collision2_name = new StringWrapper();
    this._wrapperFields.wrenches = WrenchWrapper.createArray();
    this._wrapperFields.total_wrench = new WrenchWrapper();
    this._wrapperFields.contact_positions = Vector3Wrapper.createArray();
    this._wrapperFields.contact_normals = Vector3Wrapper.createArray();
    this._wrapperFields.depths = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ContactStateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ContactStateArrayWrapper;
  }
  static get ArrayType() {
    return ContactStateArrayWrapper;
  }
  static get refObjectArrayType() {
    return ContactStateRefStructArray
  }
  static get refObjectType() {
    return ContactStateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._infoIntialized) {
        throw new TypeError('Invalid argument: info in ContactState');
      }
      if (!this._collision1_nameIntialized) {
        throw new TypeError('Invalid argument: collision1_name in ContactState');
      }
      if (!this._collision2_nameIntialized) {
        throw new TypeError('Invalid argument: collision2_name in ContactState');
      }
    }
    if (own) {
      this._wrapperFields.info.freeze(own, checkConsistency);
    }
    this._refObject.info = this._wrapperFields.info.refObject;
    if (own) {
      this._wrapperFields.collision1_name.freeze(own, checkConsistency);
    }
    this._refObject.collision1_name = this._wrapperFields.collision1_name.refObject;
    if (own) {
      this._wrapperFields.collision2_name.freeze(own, checkConsistency);
    }
    this._refObject.collision2_name = this._wrapperFields.collision2_name.refObject;
    this._wrapperFields.wrenches.freeze(own, checkConsistency);
    this._refObject.wrenches = this._wrapperFields.wrenches.refObject;
    this._wrapperFields.total_wrench.freeze(own, checkConsistency);
    this._refObject.total_wrench = this._wrapperFields.total_wrench.refObject;
    this._wrapperFields.contact_positions.freeze(own, checkConsistency);
    this._refObject.contact_positions = this._wrapperFields.contact_positions.refObject;
    this._wrapperFields.contact_normals.freeze(own, checkConsistency);
    this._refObject.contact_normals = this._wrapperFields.contact_normals.refObject;
    this._wrapperFields.depths.freeze(own, checkConsistency);
    this._refObject.depths = this._wrapperFields.depths.refObject;
    if (own) {
      this._wrapperFields.depths.fill(Float64Array.from([]));
      this._wrapperFields.depths.freeze(own, checkConsistency);
      this._refObject.depths = this._wrapperFields.depths.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._infoIntialized = true;
    this._wrapperFields.info.data = refObject.info.data;
    this._collision1_nameIntialized = true;
    this._wrapperFields.collision1_name.data = refObject.collision1_name.data;
    this._collision2_nameIntialized = true;
    this._wrapperFields.collision2_name.data = refObject.collision2_name.data;
    this._wrapperFields.wrenches.copyRefObject(refObject.wrenches);
    this._wrapperFields.total_wrench.copyRefObject(refObject.total_wrench);
    this._wrapperFields.contact_positions.copyRefObject(refObject.contact_positions);
    this._wrapperFields.contact_normals.copyRefObject(refObject.contact_normals);
    this._wrapperFields.depths.copyRefObject(refObject.depths);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.info);
    StringWrapper.freeStruct(refObject.collision1_name);
    StringWrapper.freeStruct(refObject.collision2_name);
    if (refObject.wrenches.size != 0) {
      WrenchWrapper.ArrayType.freeArray(refObject.wrenches);
      if (WrenchWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.wrenches, WrenchWrapper.refObjectArrayType, 'data');
      }
    }
    WrenchWrapper.freeStruct(refObject.total_wrench);
    if (refObject.contact_positions.size != 0) {
      Vector3Wrapper.ArrayType.freeArray(refObject.contact_positions);
      if (Vector3Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.contact_positions, Vector3Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.contact_normals.size != 0) {
      Vector3Wrapper.ArrayType.freeArray(refObject.contact_normals);
      if (Vector3Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.contact_normals, Vector3Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.depths.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.depths);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.depths, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ContactStateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'msg', interfaceName: 'ContactState'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get info() {
    if (!this._infoIntialized) {
      return undefined;
    }
    return this._wrapperFields.info.data;
  }
  set info(value) {
    this._infoIntialized = true;
    this._wrapperFields.info.data = value;
  }
  get collision1_name() {
    if (!this._collision1_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.collision1_name.data;
  }
  set collision1_name(value) {
    this._collision1_nameIntialized = true;
    this._wrapperFields.collision1_name.data = value;
  }
  get collision2_name() {
    if (!this._collision2_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.collision2_name.data;
  }
  set collision2_name(value) {
    this._collision2_nameIntialized = true;
    this._wrapperFields.collision2_name.data = value;
  }
  get wrenches() {
    return this._wrapperFields.wrenches;
  }
  set wrenches(value) {
    this._wrapperFields.wrenches.fill(value);
  }
  get total_wrench() {
    return this._wrapperFields.total_wrench;
  }
  set total_wrench(value) {
    if (value instanceof WrenchWrapper) {
      this._wrapperFields.total_wrench.copy(value);
    } else {
      this._wrapperFields.total_wrench.copy(new WrenchWrapper(value));
    }
  }
  get contact_positions() {
    return this._wrapperFields.contact_positions;
  }
  set contact_positions(value) {
    this._wrapperFields.contact_positions.fill(value);
  }
  get contact_normals() {
    return this._wrapperFields.contact_normals;
  }
  set contact_normals(value) {
    this._wrapperFields.contact_normals.fill(value);
  }
  get depths() {
    return this._wrapperFields['depths'].data;
  }
  set depths(value) {
    this._wrapperFields['depths'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ContactStateRefStruct(refObject.toObject());
    this._infoIntialized = true;
    this._wrapperFields.info.copyRefObject(this._refObject.info);
    this._collision1_nameIntialized = true;
    this._wrapperFields.collision1_name.copyRefObject(this._refObject.collision1_name);
    this._collision2_nameIntialized = true;
    this._wrapperFields.collision2_name.copyRefObject(this._refObject.collision2_name);
    this._wrapperFields.wrenches.copyRefObject(this._refObject.wrenches);
    this._wrapperFields.total_wrench.copyRefObject(this._refObject.total_wrench);
    this._wrapperFields.contact_positions.copyRefObject(this._refObject.contact_positions);
    this._wrapperFields.contact_normals.copyRefObject(this._refObject.contact_normals);
    this._wrapperFields.depths.copyRefObject(this._refObject.depths);
  }
  copy(other) {
    this._refObject = new ContactStateRefStruct(other._refObject.toObject());
    this._infoIntialized = true;
    this._wrapperFields.info.copy(other._wrapperFields.info);
    this._collision1_nameIntialized = true;
    this._wrapperFields.collision1_name.copy(other._wrapperFields.collision1_name);
    this._collision2_nameIntialized = true;
    this._wrapperFields.collision2_name.copy(other._wrapperFields.collision2_name);
    this._wrapperFields.wrenches.copy(other._wrapperFields.wrenches);
    this._wrapperFields.total_wrench.copy(other._wrapperFields.total_wrench);
    this._wrapperFields.contact_positions.copy(other._wrapperFields.contact_positions);
    this._wrapperFields.contact_normals.copy(other._wrapperFields.contact_normals);
    this._wrapperFields.depths.copy(other._wrapperFields.depths);
  }
  static get classType() {
    return ContactStateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"info","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"collision1_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"collision2_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"wrenches","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"total_wrench","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Wrench","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"contact_positions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"contact_normals","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"depths","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"ContactState","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ContactState"};
  }
  hasMember(name) {
    let memberNames = ["info","collision1_name","collision2_name","wrenches","total_wrench","contact_positions","contact_normals","depths"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ContactStateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ContactStateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ContactStateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ContactStateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ContactStateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ContactStateRefArray(size);
    this._refObject = new ContactStateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ContactStateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ContactStateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ContactStateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ContactStateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ContactStateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ContactStateArrayWrapper;
  }
}
module.exports = ContactStateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "info",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "collision1_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "collision2_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "wrenches",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "total_wrench",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Wrench",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "contact_positions",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "contact_normals",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "depths",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "ContactState",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ContactState"
}
*/
