// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__GetWorldProperties_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetWorldProperties_ResponseRefStruct = StructType({
  sim_time: primitiveTypes.float64,
  model_names: StringWrapper.refObjectArrayType,
  rendering_enabled: primitiveTypes.bool,
  success: primitiveTypes.bool,
  status_message: primitiveTypes.string,
});
const GetWorldProperties_ResponseRefArray = ArrayType(GetWorldProperties_ResponseRefStruct);
const GetWorldProperties_ResponseRefStructArray = StructType({
  data: GetWorldProperties_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetWorldProperties_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._model_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetWorldProperties_ResponseRefStruct(other._refObject.toObject());
      this._sim_timeIntialized = true;
      this._wrapperFields.model_names = StringWrapper.createArray();
      this._wrapperFields.model_names.copy(other._wrapperFields.model_names);
      this.model_names = other.model_names;
      this._rendering_enabledIntialized = true;
      this._successIntialized = true;
      this._status_messageIntialized = true;
      this._wrapperFields.status_message =  new StringWrapper(other._wrapperFields.status_message);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetWorldProperties_ResponseRefStruct();
    this._sim_timeIntialized = false;
    this._wrapperFields.model_names = StringWrapper.createArray();
    this._rendering_enabledIntialized = false;
    this._successIntialized = false;
    this._status_messageIntialized = false;
    this._wrapperFields.status_message = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetWorldProperties_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetWorldProperties_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetWorldProperties_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetWorldProperties_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetWorldProperties_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._sim_timeIntialized) {
        throw new TypeError('Invalid argument: sim_time in GetWorldProperties_Response');
      }
      if (!this._rendering_enabledIntialized) {
        throw new TypeError('Invalid argument: rendering_enabled in GetWorldProperties_Response');
      }
      if (!this._successIntialized) {
        throw new TypeError('Invalid argument: success in GetWorldProperties_Response');
      }
      if (!this._status_messageIntialized) {
        throw new TypeError('Invalid argument: status_message in GetWorldProperties_Response');
      }
    }
    if (!own) {
      this._wrapperFields.model_names.fill(this._model_namesArray);
      this._wrapperFields.model_names.freeze(own, checkConsistency);
      this._refObject.model_names = this._wrapperFields.model_names.refObject;
    } else {
      this._wrapperFields.model_names.fill([]);
      this._wrapperFields.model_names.freeze(own, checkConsistency);
      this._refObject.model_names = this._wrapperFields.model_names.refObject;
    }
    if (own) {
      this._wrapperFields.status_message.freeze(own, checkConsistency);
    }
    this._refObject.status_message = this._wrapperFields.status_message.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._sim_timeIntialized = true;
    this._refObject.sim_time = refObject.sim_time;
    refObject.model_names.data.length = refObject.model_names.size;
    for (let index = 0; index < refObject.model_names.size; index++) {
      this._model_namesArray[index] = refObject.model_names.data[index].data;
    }
    this._rendering_enabledIntialized = true;
    this._refObject.rendering_enabled = refObject.rendering_enabled;
    this._successIntialized = true;
    this._refObject.success = refObject.success;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = refObject.status_message.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.model_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.model_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.model_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    StringWrapper.freeStruct(refObject.status_message);
  }
  static destoryRawROS(msg) {
    GetWorldProperties_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'GetWorldProperties_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sim_time() {
    if (!this._sim_timeIntialized) {
      return undefined;
    }
    return this._refObject.sim_time;
  }
  set sim_time(value) {
    this._sim_timeIntialized = true;
    this._refObject.sim_time = value;
  }
  get model_names() {
    return this._model_namesArray;
  }
  set model_names(value) {
    this._model_namesArray = value;
  }
  get rendering_enabled() {
    if (!this._rendering_enabledIntialized) {
      return undefined;
    }
    return this._refObject.rendering_enabled;
  }
  set rendering_enabled(value) {
    this._rendering_enabledIntialized = true;
    this._refObject.rendering_enabled = value;
  }
  get success() {
    if (!this._successIntialized) {
      return undefined;
    }
    return this._refObject.success;
  }
  set success(value) {
    this._successIntialized = true;
    this._refObject.success = value;
  }
  get status_message() {
    if (!this._status_messageIntialized) {
      return undefined;
    }
    return this._wrapperFields.status_message.data;
  }
  set status_message(value) {
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetWorldProperties_ResponseRefStruct(refObject.toObject());
    this._sim_timeIntialized = true;
    refObject.model_names.data.length = refObject.model_names.size;
    for (let index = 0; index < refObject.model_names.size; index++) {
      this._model_namesArray[index] = refObject.model_names.data[index].data;
    }
    this._rendering_enabledIntialized = true;
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copyRefObject(this._refObject.status_message);
  }
  copy(other) {
    this._refObject = new GetWorldProperties_ResponseRefStruct(other._refObject.toObject());
    this._sim_timeIntialized = true;
    this._model_namesArray = other._model_namesArray.slice();
    this._rendering_enabledIntialized = true;
    this._successIntialized = true;
    this._status_messageIntialized = true;
    this._wrapperFields.status_message.copy(other._wrapperFields.status_message);
  }
  static get classType() {
    return GetWorldProperties_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sim_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"model_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"rendering_enabled","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"success","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"status_message","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"GetWorldProperties_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetWorldProperties_Response"};
  }
  hasMember(name) {
    let memberNames = ["sim_time","model_names","rendering_enabled","success","status_message"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetWorldProperties_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetWorldProperties_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetWorldProperties_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetWorldProperties_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetWorldProperties_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetWorldProperties_ResponseRefArray(size);
    this._refObject = new GetWorldProperties_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetWorldProperties_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetWorldProperties_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetWorldProperties_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetWorldProperties_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetWorldProperties_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetWorldProperties_ResponseArrayWrapper;
  }
}
module.exports = GetWorldProperties_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "sim_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "model_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "rendering_enabled",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "success",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "status_message",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "GetWorldProperties_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetWorldProperties_Response"
}
*/
