// rclnodejs/generated/gazebo_msgs/gazebo_msgs__srv__SetModelConfiguration_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const SetModelConfiguration_RequestRefStruct = StructType({
  model_name: primitiveTypes.string,
  urdf_param_name: primitiveTypes.string,
  joint_names: StringWrapper.refObjectArrayType,
  joint_positions: Float64Wrapper.refObjectArrayType,
});
const SetModelConfiguration_RequestRefArray = ArrayType(SetModelConfiguration_RequestRefStruct);
const SetModelConfiguration_RequestRefStructArray = StructType({
  data: SetModelConfiguration_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetModelConfiguration_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._joint_namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetModelConfiguration_RequestRefStruct(other._refObject.toObject());
      this._model_nameIntialized = true;
      this._wrapperFields.model_name =  new StringWrapper(other._wrapperFields.model_name);
      this._urdf_param_nameIntialized = true;
      this._wrapperFields.urdf_param_name =  new StringWrapper(other._wrapperFields.urdf_param_name);
      this._wrapperFields.joint_names = StringWrapper.createArray();
      this._wrapperFields.joint_names.copy(other._wrapperFields.joint_names);
      this.joint_names = other.joint_names;
      this._wrapperFields.joint_positions = Float64Wrapper.createArray();
      this._wrapperFields.joint_positions.copy(other._wrapperFields.joint_positions);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetModelConfiguration_RequestRefStruct();
    this._model_nameIntialized = false;
    this._wrapperFields.model_name = new StringWrapper();
    this._urdf_param_nameIntialized = false;
    this._wrapperFields.urdf_param_name = new StringWrapper();
    this._wrapperFields.joint_names = StringWrapper.createArray();
    this._wrapperFields.joint_positions = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SetModelConfiguration_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetModelConfiguration_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetModelConfiguration_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetModelConfiguration_RequestRefStructArray
  }
  static get refObjectType() {
    return SetModelConfiguration_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._model_nameIntialized) {
        throw new TypeError('Invalid argument: model_name in SetModelConfiguration_Request');
      }
      if (!this._urdf_param_nameIntialized) {
        throw new TypeError('Invalid argument: urdf_param_name in SetModelConfiguration_Request');
      }
    }
    if (own) {
      this._wrapperFields.model_name.freeze(own, checkConsistency);
    }
    this._refObject.model_name = this._wrapperFields.model_name.refObject;
    if (own) {
      this._wrapperFields.urdf_param_name.freeze(own, checkConsistency);
    }
    this._refObject.urdf_param_name = this._wrapperFields.urdf_param_name.refObject;
    if (!own) {
      this._wrapperFields.joint_names.fill(this._joint_namesArray);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    } else {
      this._wrapperFields.joint_names.fill([]);
      this._wrapperFields.joint_names.freeze(own, checkConsistency);
      this._refObject.joint_names = this._wrapperFields.joint_names.refObject;
    }
    this._wrapperFields.joint_positions.freeze(own, checkConsistency);
    this._refObject.joint_positions = this._wrapperFields.joint_positions.refObject;
    if (own) {
      this._wrapperFields.joint_positions.fill(Float64Array.from([]));
      this._wrapperFields.joint_positions.freeze(own, checkConsistency);
      this._refObject.joint_positions = this._wrapperFields.joint_positions.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = refObject.model_name.data;
    this._urdf_param_nameIntialized = true;
    this._wrapperFields.urdf_param_name.data = refObject.urdf_param_name.data;
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.joint_positions.copyRefObject(refObject.joint_positions);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.model_name);
    StringWrapper.freeStruct(refObject.urdf_param_name);
    if (refObject.joint_names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.joint_names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_names, StringWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.joint_positions.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.joint_positions);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.joint_positions, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SetModelConfiguration_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'gazebo_msgs', subFolder: 'srv', interfaceName: 'SetModelConfiguration_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get model_name() {
    if (!this._model_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.model_name.data;
  }
  set model_name(value) {
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.data = value;
  }
  get urdf_param_name() {
    if (!this._urdf_param_nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.urdf_param_name.data;
  }
  set urdf_param_name(value) {
    this._urdf_param_nameIntialized = true;
    this._wrapperFields.urdf_param_name.data = value;
  }
  get joint_names() {
    return this._joint_namesArray;
  }
  set joint_names(value) {
    this._joint_namesArray = value;
  }
  get joint_positions() {
    return this._wrapperFields['joint_positions'].data;
  }
  set joint_positions(value) {
    this._wrapperFields['joint_positions'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SetModelConfiguration_RequestRefStruct(refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copyRefObject(this._refObject.model_name);
    this._urdf_param_nameIntialized = true;
    this._wrapperFields.urdf_param_name.copyRefObject(this._refObject.urdf_param_name);
    refObject.joint_names.data.length = refObject.joint_names.size;
    for (let index = 0; index < refObject.joint_names.size; index++) {
      this._joint_namesArray[index] = refObject.joint_names.data[index].data;
    }
    this._wrapperFields.joint_positions.copyRefObject(this._refObject.joint_positions);
  }
  copy(other) {
    this._refObject = new SetModelConfiguration_RequestRefStruct(other._refObject.toObject());
    this._model_nameIntialized = true;
    this._wrapperFields.model_name.copy(other._wrapperFields.model_name);
    this._urdf_param_nameIntialized = true;
    this._wrapperFields.urdf_param_name.copy(other._wrapperFields.urdf_param_name);
    this._joint_namesArray = other._joint_namesArray.slice();
    this._wrapperFields.joint_positions.copy(other._wrapperFields.joint_positions);
  }
  static get classType() {
    return SetModelConfiguration_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"model_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"urdf_param_name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"joint_names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"joint_positions","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"gazebo_msgs","type":"SetModelConfiguration_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetModelConfiguration_Request"};
  }
  hasMember(name) {
    let memberNames = ["model_name","urdf_param_name","joint_names","joint_positions"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetModelConfiguration_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetModelConfiguration_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetModelConfiguration_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetModelConfiguration_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetModelConfiguration_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetModelConfiguration_RequestRefArray(size);
    this._refObject = new SetModelConfiguration_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetModelConfiguration_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetModelConfiguration_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetModelConfiguration_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetModelConfiguration_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetModelConfiguration_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetModelConfiguration_RequestArrayWrapper;
  }
}
module.exports = SetModelConfiguration_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "model_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "urdf_param_name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "joint_names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "joint_positions",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "gazebo_msgs",
    "type": "SetModelConfiguration_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetModelConfiguration_Request"
}
*/
