// rclnodejs/generated/action_tutorials_interfaces/action_tutorials_interfaces__action__Fibonacci_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const Fibonacci_FeedbackRefStruct = StructType({
  partial_sequence: Int32Wrapper.refObjectArrayType,
});
const Fibonacci_FeedbackRefArray = ArrayType(Fibonacci_FeedbackRefStruct);
const Fibonacci_FeedbackRefStructArray = StructType({
  data: Fibonacci_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Fibonacci_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Fibonacci_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.partial_sequence = Int32Wrapper.createArray();
      this._wrapperFields.partial_sequence.copy(other._wrapperFields.partial_sequence);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Fibonacci_FeedbackRefStruct();
    this._wrapperFields.partial_sequence = Int32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Fibonacci_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Fibonacci_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return Fibonacci_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return Fibonacci_FeedbackRefStructArray
  }
  static get refObjectType() {
    return Fibonacci_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.partial_sequence.freeze(own, checkConsistency);
    this._refObject.partial_sequence = this._wrapperFields.partial_sequence.refObject;
    if (own) {
      this._wrapperFields.partial_sequence.fill(Int32Array.from([]));
      this._wrapperFields.partial_sequence.freeze(own, checkConsistency);
      this._refObject.partial_sequence = this._wrapperFields.partial_sequence.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.partial_sequence.copyRefObject(refObject.partial_sequence);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.partial_sequence.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.partial_sequence);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.partial_sequence, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Fibonacci_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_tutorials_interfaces', subFolder: 'action', interfaceName: 'Fibonacci_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get partial_sequence() {
    return this._wrapperFields['partial_sequence'].data;
  }
  set partial_sequence(value) {
    this._wrapperFields['partial_sequence'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new Fibonacci_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.partial_sequence.copyRefObject(this._refObject.partial_sequence);
  }
  copy(other) {
    this._refObject = new Fibonacci_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.partial_sequence.copy(other._wrapperFields.partial_sequence);
  }
  static get classType() {
    return Fibonacci_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"partial_sequence","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"action_tutorials_interfaces","type":"Fibonacci_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Fibonacci_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["partial_sequence"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Fibonacci_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Fibonacci_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Fibonacci_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Fibonacci_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Fibonacci_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Fibonacci_FeedbackRefArray(size);
    this._refObject = new Fibonacci_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Fibonacci_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Fibonacci_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Fibonacci_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Fibonacci_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Fibonacci_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Fibonacci_FeedbackArrayWrapper;
  }
}
module.exports = Fibonacci_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "partial_sequence",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_tutorials_interfaces",
    "type": "Fibonacci_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Fibonacci_Feedback"
}
*/
