// rclnodejs/generated/action_tutorials_interfaces/action_tutorials_interfaces__action__Fibonacci_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Fibonacci_GoalRefStruct = StructType({
  order: primitiveTypes.int32,
});
const Fibonacci_GoalRefArray = ArrayType(Fibonacci_GoalRefStruct);
const Fibonacci_GoalRefStructArray = StructType({
  data: Fibonacci_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Fibonacci_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Fibonacci_GoalRefStruct(other._refObject.toObject());
      this._orderIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Fibonacci_GoalRefStruct();
    this._orderIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new Fibonacci_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Fibonacci_GoalArrayWrapper;
  }
  static get ArrayType() {
    return Fibonacci_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return Fibonacci_GoalRefStructArray
  }
  static get refObjectType() {
    return Fibonacci_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._orderIntialized) {
        throw new TypeError('Invalid argument: order in Fibonacci_Goal');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._orderIntialized = true;
    this._refObject.order = refObject.order;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    Fibonacci_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'action_tutorials_interfaces', subFolder: 'action', interfaceName: 'Fibonacci_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get order() {
    if (!this._orderIntialized) {
      return undefined;
    }
    return this._refObject.order;
  }
  set order(value) {
    this._orderIntialized = true;
    this._refObject.order = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Fibonacci_GoalRefStruct(refObject.toObject());
    this._orderIntialized = true;
  }
  copy(other) {
    this._refObject = new Fibonacci_GoalRefStruct(other._refObject.toObject());
    this._orderIntialized = true;
  }
  static get classType() {
    return Fibonacci_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"order","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"action_tutorials_interfaces","type":"Fibonacci_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Fibonacci_Goal"};
  }
  hasMember(name) {
    let memberNames = ["order"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Fibonacci_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Fibonacci_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Fibonacci_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Fibonacci_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Fibonacci_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Fibonacci_GoalRefArray(size);
    this._refObject = new Fibonacci_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Fibonacci_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Fibonacci_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Fibonacci_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Fibonacci_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Fibonacci_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Fibonacci_GoalArrayWrapper;
  }
}
module.exports = Fibonacci_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "order",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "action_tutorials_interfaces",
    "type": "Fibonacci_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Fibonacci_Goal"
}
*/
