// rclnodejs/generated/map_msgs/map_msgs__msg__OccupancyGridUpdate.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const Int8Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int8.js');
const OccupancyGridUpdateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  x: primitiveTypes.int32,
  y: primitiveTypes.int32,
  width: primitiveTypes.uint32,
  height: primitiveTypes.uint32,
  data: Int8Wrapper.refObjectArrayType,
});
const OccupancyGridUpdateRefArray = ArrayType(OccupancyGridUpdateRefStruct);
const OccupancyGridUpdateRefStructArray = StructType({
  data: OccupancyGridUpdateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class OccupancyGridUpdateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new OccupancyGridUpdateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._xIntialized = true;
      this._yIntialized = true;
      this._widthIntialized = true;
      this._heightIntialized = true;
      this._wrapperFields.data = Int8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new OccupancyGridUpdateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._xIntialized = false;
    this._yIntialized = false;
    this._widthIntialized = false;
    this._heightIntialized = false;
    this._wrapperFields.data = Int8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new OccupancyGridUpdateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new OccupancyGridUpdateArrayWrapper;
  }
  static get ArrayType() {
    return OccupancyGridUpdateArrayWrapper;
  }
  static get refObjectArrayType() {
    return OccupancyGridUpdateRefStructArray
  }
  static get refObjectType() {
    return OccupancyGridUpdateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in OccupancyGridUpdate');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in OccupancyGridUpdate');
      }
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in OccupancyGridUpdate');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in OccupancyGridUpdate');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Int8Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    if (refObject.data.size != 0) {
      Int8Wrapper.ArrayType.freeArray(refObject.data);
      if (Int8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, Int8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    OccupancyGridUpdateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'msg', interfaceName: 'OccupancyGridUpdate'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new OccupancyGridUpdateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._xIntialized = true;
    this._yIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new OccupancyGridUpdateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._xIntialized = true;
    this._yIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return OccupancyGridUpdateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"OccupancyGridUpdate","stringUpperBound":null,"isPrimitiveType":false},"msgName":"OccupancyGridUpdate"};
  }
  hasMember(name) {
    let memberNames = ["header","x","y","width","height","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class OccupancyGridUpdateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof OccupancyGridUpdateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new OccupancyGridUpdateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OccupancyGridUpdateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to OccupancyGridUpdateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new OccupancyGridUpdateRefArray(size);
    this._refObject = new OccupancyGridUpdateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new OccupancyGridUpdateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof OccupancyGridUpdateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "OccupancyGridUpdateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      OccupancyGridUpdateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return OccupancyGridUpdateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return OccupancyGridUpdateArrayWrapper;
  }
}
module.exports = OccupancyGridUpdateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "OccupancyGridUpdate",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "OccupancyGridUpdate"
}
*/
