// rclnodejs/generated/map_msgs/map_msgs__srv__GetPointMapROI_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const GetPointMapROI_ResponseRefStruct = StructType({
  sub_map: PointCloud2Wrapper.refObjectType,
});
const GetPointMapROI_ResponseRefArray = ArrayType(GetPointMapROI_ResponseRefStruct);
const GetPointMapROI_ResponseRefStructArray = StructType({
  data: GetPointMapROI_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPointMapROI_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPointMapROI_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.sub_map =  new PointCloud2Wrapper(other._wrapperFields.sub_map);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPointMapROI_ResponseRefStruct();
    this._wrapperFields.sub_map = new PointCloud2Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetPointMapROI_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPointMapROI_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetPointMapROI_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPointMapROI_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetPointMapROI_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.sub_map.freeze(own, checkConsistency);
    this._refObject.sub_map = this._wrapperFields.sub_map.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.sub_map.copyRefObject(refObject.sub_map);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PointCloud2Wrapper.freeStruct(refObject.sub_map);
  }
  static destoryRawROS(msg) {
    GetPointMapROI_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'srv', interfaceName: 'GetPointMapROI_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get sub_map() {
    return this._wrapperFields.sub_map;
  }
  set sub_map(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.sub_map.copy(value);
    } else {
      this._wrapperFields.sub_map.copy(new PointCloud2Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetPointMapROI_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.sub_map.copyRefObject(this._refObject.sub_map);
  }
  copy(other) {
    this._refObject = new GetPointMapROI_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.sub_map.copy(other._wrapperFields.sub_map);
  }
  static get classType() {
    return GetPointMapROI_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"sub_map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"GetPointMapROI_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPointMapROI_Response"};
  }
  hasMember(name) {
    let memberNames = ["sub_map"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPointMapROI_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPointMapROI_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPointMapROI_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPointMapROI_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPointMapROI_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPointMapROI_ResponseRefArray(size);
    this._refObject = new GetPointMapROI_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPointMapROI_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPointMapROI_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPointMapROI_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPointMapROI_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPointMapROI_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPointMapROI_ResponseArrayWrapper;
  }
}
module.exports = GetPointMapROI_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "sub_map",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "PointCloud2",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "GetPointMapROI_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPointMapROI_Response"
}
*/
