// rclnodejs/generated/map_msgs/map_msgs__msg__PointCloud2Update.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const PointCloud2Wrapper = require('../../generated/sensor_msgs/sensor_msgs__msg__PointCloud2.js');
const PointCloud2UpdateRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type: primitiveTypes.uint32,
  points: PointCloud2Wrapper.refObjectType,
});
const PointCloud2UpdateRefArray = ArrayType(PointCloud2UpdateRefStruct);
const PointCloud2UpdateRefStructArray = StructType({
  data: PointCloud2UpdateRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PointCloud2UpdateWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PointCloud2UpdateRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._typeIntialized = true;
      this._wrapperFields.points =  new PointCloud2Wrapper(other._wrapperFields.points);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PointCloud2UpdateRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._typeIntialized = false;
    this._wrapperFields.points = new PointCloud2Wrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PointCloud2UpdateWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PointCloud2UpdateArrayWrapper;
  }
  static get ArrayType() {
    return PointCloud2UpdateArrayWrapper;
  }
  static get refObjectArrayType() {
    return PointCloud2UpdateRefStructArray
  }
  static get refObjectType() {
    return PointCloud2UpdateRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in PointCloud2Update');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._wrapperFields.points.copyRefObject(refObject.points);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    PointCloud2Wrapper.freeStruct(refObject.points);
  }
  static destoryRawROS(msg) {
    PointCloud2UpdateWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'msg', interfaceName: 'PointCloud2Update'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    if (value instanceof PointCloud2Wrapper) {
      this._wrapperFields.points.copy(value);
    } else {
      this._wrapperFields.points.copy(new PointCloud2Wrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new PointCloud2UpdateRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._typeIntialized = true;
    this._wrapperFields.points.copyRefObject(this._refObject.points);
  }
  copy(other) {
    this._refObject = new PointCloud2UpdateRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._typeIntialized = true;
    this._wrapperFields.points.copy(other._wrapperFields.points);
  }
  static get classType() {
    return PointCloud2UpdateWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint32","name":"ADD","value":0},{"type":"uint32","name":"DELETE","value":1}],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"points","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sensor_msgs","type":"PointCloud2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"PointCloud2Update","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PointCloud2Update"};
  }
  hasMember(name) {
    let memberNames = ["header","type","points"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PointCloud2UpdateArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PointCloud2UpdateWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PointCloud2UpdateWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointCloud2UpdateArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointCloud2UpdateArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PointCloud2UpdateRefArray(size);
    this._refObject = new PointCloud2UpdateRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PointCloud2UpdateWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PointCloud2UpdateArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PointCloud2UpdateArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PointCloud2UpdateWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PointCloud2UpdateWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PointCloud2UpdateArrayWrapper;
  }
}
// Define constants (2 in total)
Object.defineProperty(PointCloud2UpdateWrapper, "ADD", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PointCloud2UpdateWrapper, "DELETE", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = PointCloud2UpdateWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint32",
      "name": "ADD",
      "value": 0
    },
    {
      "type": "uint32",
      "name": "DELETE",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sensor_msgs",
        "type": "PointCloud2",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "PointCloud2Update",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PointCloud2Update"
}
*/
