// rclnodejs/generated/map_msgs/map_msgs__srv__SetMapProjections_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetMapProjections_RequestRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const SetMapProjections_RequestRefArray = ArrayType(SetMapProjections_RequestRefStruct);
const SetMapProjections_RequestRefStructArray = StructType({
  data: SetMapProjections_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetMapProjections_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMapProjections_RequestRefStruct(other._refObject.toObject());
      this.__dummyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetMapProjections_RequestRefStruct();
    this.__dummyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new SetMapProjections_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMapProjections_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetMapProjections_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMapProjections_RequestRefStructArray
  }
  static get refObjectType() {
    return SetMapProjections_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this.__dummyIntialized) {
        throw new TypeError('Invalid argument: _dummy in SetMapProjections_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this.__dummyIntialized = true;
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SetMapProjections_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'srv', interfaceName: 'SetMapProjections_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    if (!this.__dummyIntialized) {
      return undefined;
    }
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this.__dummyIntialized = true;
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetMapProjections_RequestRefStruct(refObject.toObject());
    this.__dummyIntialized = true;
  }
  copy(other) {
    this._refObject = new SetMapProjections_RequestRefStruct(other._refObject.toObject());
    this.__dummyIntialized = true;
  }
  static get classType() {
    return SetMapProjections_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[],"baseType":{"pkgName":"map_msgs","type":"SetMapProjections_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMapProjections_Request"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetMapProjections_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMapProjections_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMapProjections_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMapProjections_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMapProjections_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetMapProjections_RequestRefArray(size);
    this._refObject = new SetMapProjections_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMapProjections_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMapProjections_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMapProjections_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMapProjections_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMapProjections_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMapProjections_RequestArrayWrapper;
  }
}
module.exports = SetMapProjections_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "SetMapProjections_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetMapProjections_Request"
}
*/
