// rclnodejs/generated/map_msgs/map_msgs__srv__SetMapProjections_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ProjectedMapInfoWrapper = require('../../generated/map_msgs/map_msgs__msg__ProjectedMapInfo.js');
const SetMapProjections_ResponseRefStruct = StructType({
  projected_maps_info: ProjectedMapInfoWrapper.refObjectArrayType,
});
const SetMapProjections_ResponseRefArray = ArrayType(SetMapProjections_ResponseRefStruct);
const SetMapProjections_ResponseRefStructArray = StructType({
  data: SetMapProjections_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetMapProjections_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetMapProjections_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.projected_maps_info = ProjectedMapInfoWrapper.createArray();
      this._wrapperFields.projected_maps_info.copy(other._wrapperFields.projected_maps_info);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetMapProjections_ResponseRefStruct();
    this._wrapperFields.projected_maps_info = ProjectedMapInfoWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new SetMapProjections_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetMapProjections_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SetMapProjections_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetMapProjections_ResponseRefStructArray
  }
  static get refObjectType() {
    return SetMapProjections_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.projected_maps_info.freeze(own, checkConsistency);
    this._refObject.projected_maps_info = this._wrapperFields.projected_maps_info.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.projected_maps_info.copyRefObject(refObject.projected_maps_info);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.projected_maps_info.size != 0) {
      ProjectedMapInfoWrapper.ArrayType.freeArray(refObject.projected_maps_info);
      if (ProjectedMapInfoWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.projected_maps_info, ProjectedMapInfoWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SetMapProjections_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'srv', interfaceName: 'SetMapProjections_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get projected_maps_info() {
    return this._wrapperFields.projected_maps_info;
  }
  set projected_maps_info(value) {
    this._wrapperFields.projected_maps_info.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new SetMapProjections_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.projected_maps_info.copyRefObject(this._refObject.projected_maps_info);
  }
  copy(other) {
    this._refObject = new SetMapProjections_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.projected_maps_info.copy(other._wrapperFields.projected_maps_info);
  }
  static get classType() {
    return SetMapProjections_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"projected_maps_info","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"map_msgs","type":"ProjectedMapInfo","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"SetMapProjections_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetMapProjections_Response"};
  }
  hasMember(name) {
    let memberNames = ["projected_maps_info"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetMapProjections_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetMapProjections_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetMapProjections_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMapProjections_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetMapProjections_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetMapProjections_ResponseRefArray(size);
    this._refObject = new SetMapProjections_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetMapProjections_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetMapProjections_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetMapProjections_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetMapProjections_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetMapProjections_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetMapProjections_ResponseArrayWrapper;
  }
}
module.exports = SetMapProjections_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "projected_maps_info",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "map_msgs",
        "type": "ProjectedMapInfo",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "SetMapProjections_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetMapProjections_Response"
}
*/
