// rclnodejs/generated/map_msgs/map_msgs__msg__ProjectedMap.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const OccupancyGridWrapper = require('../../generated/nav_msgs/nav_msgs__msg__OccupancyGrid.js');
const ProjectedMapRefStruct = StructType({
  map: OccupancyGridWrapper.refObjectType,
  min_z: primitiveTypes.float64,
  max_z: primitiveTypes.float64,
});
const ProjectedMapRefArray = ArrayType(ProjectedMapRefStruct);
const ProjectedMapRefStructArray = StructType({
  data: ProjectedMapRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ProjectedMapWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ProjectedMapRefStruct(other._refObject.toObject());
      this._wrapperFields.map =  new OccupancyGridWrapper(other._wrapperFields.map);
      this._min_zIntialized = true;
      this._max_zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ProjectedMapRefStruct();
    this._wrapperFields.map = new OccupancyGridWrapper();
    this._min_zIntialized = false;
    this._max_zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ProjectedMapWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ProjectedMapArrayWrapper;
  }
  static get ArrayType() {
    return ProjectedMapArrayWrapper;
  }
  static get refObjectArrayType() {
    return ProjectedMapRefStructArray
  }
  static get refObjectType() {
    return ProjectedMapRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._min_zIntialized) {
        throw new TypeError('Invalid argument: min_z in ProjectedMap');
      }
      if (!this._max_zIntialized) {
        throw new TypeError('Invalid argument: max_z in ProjectedMap');
      }
    }
    this._wrapperFields.map.freeze(own, checkConsistency);
    this._refObject.map = this._wrapperFields.map.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.map.copyRefObject(refObject.map);
    this._min_zIntialized = true;
    this._refObject.min_z = refObject.min_z;
    this._max_zIntialized = true;
    this._refObject.max_z = refObject.max_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    OccupancyGridWrapper.freeStruct(refObject.map);
  }
  static destoryRawROS(msg) {
    ProjectedMapWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'msg', interfaceName: 'ProjectedMap'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get map() {
    return this._wrapperFields.map;
  }
  set map(value) {
    if (value instanceof OccupancyGridWrapper) {
      this._wrapperFields.map.copy(value);
    } else {
      this._wrapperFields.map.copy(new OccupancyGridWrapper(value));
    }
  }
  get min_z() {
    if (!this._min_zIntialized) {
      return undefined;
    }
    return this._refObject.min_z;
  }
  set min_z(value) {
    this._min_zIntialized = true;
    this._refObject.min_z = value;
  }
  get max_z() {
    if (!this._max_zIntialized) {
      return undefined;
    }
    return this._refObject.max_z;
  }
  set max_z(value) {
    this._max_zIntialized = true;
    this._refObject.max_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ProjectedMapRefStruct(refObject.toObject());
    this._wrapperFields.map.copyRefObject(this._refObject.map);
    this._min_zIntialized = true;
    this._max_zIntialized = true;
  }
  copy(other) {
    this._refObject = new ProjectedMapRefStruct(other._refObject.toObject());
    this._wrapperFields.map.copy(other._wrapperFields.map);
    this._min_zIntialized = true;
    this._max_zIntialized = true;
  }
  static get classType() {
    return ProjectedMapWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"OccupancyGrid","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"min_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"ProjectedMap","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ProjectedMap"};
  }
  hasMember(name) {
    let memberNames = ["map","min_z","max_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ProjectedMapArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ProjectedMapWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ProjectedMapWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ProjectedMapRefArray(size);
    this._refObject = new ProjectedMapRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ProjectedMapWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ProjectedMapArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ProjectedMapArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ProjectedMapWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ProjectedMapWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ProjectedMapArrayWrapper;
  }
}
module.exports = ProjectedMapWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "map",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "OccupancyGrid",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "min_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "ProjectedMap",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ProjectedMap"
}
*/
