// rclnodejs/generated/map_msgs/map_msgs__srv__GetPointMapROI_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetPointMapROI_RequestRefStruct = StructType({
  x: primitiveTypes.float64,
  y: primitiveTypes.float64,
  z: primitiveTypes.float64,
  r: primitiveTypes.float64,
  l_x: primitiveTypes.float64,
  l_y: primitiveTypes.float64,
  l_z: primitiveTypes.float64,
});
const GetPointMapROI_RequestRefArray = ArrayType(GetPointMapROI_RequestRefStruct);
const GetPointMapROI_RequestRefStructArray = StructType({
  data: GetPointMapROI_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetPointMapROI_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetPointMapROI_RequestRefStruct(other._refObject.toObject());
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._rIntialized = true;
      this._l_xIntialized = true;
      this._l_yIntialized = true;
      this._l_zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetPointMapROI_RequestRefStruct();
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._rIntialized = false;
    this._l_xIntialized = false;
    this._l_yIntialized = false;
    this._l_zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GetPointMapROI_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetPointMapROI_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetPointMapROI_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetPointMapROI_RequestRefStructArray
  }
  static get refObjectType() {
    return GetPointMapROI_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in GetPointMapROI_Request');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in GetPointMapROI_Request');
      }
      if (!this._zIntialized) {
        throw new TypeError('Invalid argument: z in GetPointMapROI_Request');
      }
      if (!this._rIntialized) {
        throw new TypeError('Invalid argument: r in GetPointMapROI_Request');
      }
      if (!this._l_xIntialized) {
        throw new TypeError('Invalid argument: l_x in GetPointMapROI_Request');
      }
      if (!this._l_yIntialized) {
        throw new TypeError('Invalid argument: l_y in GetPointMapROI_Request');
      }
      if (!this._l_zIntialized) {
        throw new TypeError('Invalid argument: l_z in GetPointMapROI_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._rIntialized = true;
    this._refObject.r = refObject.r;
    this._l_xIntialized = true;
    this._refObject.l_x = refObject.l_x;
    this._l_yIntialized = true;
    this._refObject.l_y = refObject.l_y;
    this._l_zIntialized = true;
    this._refObject.l_z = refObject.l_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GetPointMapROI_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'srv', interfaceName: 'GetPointMapROI_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get r() {
    if (!this._rIntialized) {
      return undefined;
    }
    return this._refObject.r;
  }
  set r(value) {
    this._rIntialized = true;
    this._refObject.r = value;
  }
  get l_x() {
    if (!this._l_xIntialized) {
      return undefined;
    }
    return this._refObject.l_x;
  }
  set l_x(value) {
    this._l_xIntialized = true;
    this._refObject.l_x = value;
  }
  get l_y() {
    if (!this._l_yIntialized) {
      return undefined;
    }
    return this._refObject.l_y;
  }
  set l_y(value) {
    this._l_yIntialized = true;
    this._refObject.l_y = value;
  }
  get l_z() {
    if (!this._l_zIntialized) {
      return undefined;
    }
    return this._refObject.l_z;
  }
  set l_z(value) {
    this._l_zIntialized = true;
    this._refObject.l_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetPointMapROI_RequestRefStruct(refObject.toObject());
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._rIntialized = true;
    this._l_xIntialized = true;
    this._l_yIntialized = true;
    this._l_zIntialized = true;
  }
  copy(other) {
    this._refObject = new GetPointMapROI_RequestRefStruct(other._refObject.toObject());
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._rIntialized = true;
    this._l_xIntialized = true;
    this._l_yIntialized = true;
    this._l_zIntialized = true;
  }
  static get classType() {
    return GetPointMapROI_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"r","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"l_x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"l_y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"l_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"GetPointMapROI_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetPointMapROI_Request"};
  }
  hasMember(name) {
    let memberNames = ["x","y","z","r","l_x","l_y","l_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetPointMapROI_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetPointMapROI_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetPointMapROI_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPointMapROI_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetPointMapROI_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetPointMapROI_RequestRefArray(size);
    this._refObject = new GetPointMapROI_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetPointMapROI_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetPointMapROI_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetPointMapROI_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetPointMapROI_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetPointMapROI_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetPointMapROI_RequestArrayWrapper;
  }
}
module.exports = GetPointMapROI_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "r",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "l_x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "l_y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "l_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "GetPointMapROI_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetPointMapROI_Request"
}
*/
