// rclnodejs/generated/map_msgs/map_msgs__msg__ProjectedMapInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ProjectedMapInfoRefStruct = StructType({
  frame_id: primitiveTypes.string,
  x: primitiveTypes.float64,
  y: primitiveTypes.float64,
  width: primitiveTypes.float64,
  height: primitiveTypes.float64,
  min_z: primitiveTypes.float64,
  max_z: primitiveTypes.float64,
});
const ProjectedMapInfoRefArray = ArrayType(ProjectedMapInfoRefStruct);
const ProjectedMapInfoRefStructArray = StructType({
  data: ProjectedMapInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ProjectedMapInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ProjectedMapInfoRefStruct(other._refObject.toObject());
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
      this._xIntialized = true;
      this._yIntialized = true;
      this._widthIntialized = true;
      this._heightIntialized = true;
      this._min_zIntialized = true;
      this._max_zIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ProjectedMapInfoRefStruct();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
    this._xIntialized = false;
    this._yIntialized = false;
    this._widthIntialized = false;
    this._heightIntialized = false;
    this._min_zIntialized = false;
    this._max_zIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ProjectedMapInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ProjectedMapInfoArrayWrapper;
  }
  static get ArrayType() {
    return ProjectedMapInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return ProjectedMapInfoRefStructArray
  }
  static get refObjectType() {
    return ProjectedMapInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in ProjectedMapInfo');
      }
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in ProjectedMapInfo');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in ProjectedMapInfo');
      }
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in ProjectedMapInfo');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in ProjectedMapInfo');
      }
      if (!this._min_zIntialized) {
        throw new TypeError('Invalid argument: min_z in ProjectedMapInfo');
      }
      if (!this._max_zIntialized) {
        throw new TypeError('Invalid argument: max_z in ProjectedMapInfo');
      }
    }
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._min_zIntialized = true;
    this._refObject.min_z = refObject.min_z;
    this._max_zIntialized = true;
    this._refObject.max_z = refObject.max_z;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destoryRawROS(msg) {
    ProjectedMapInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'map_msgs', subFolder: 'msg', interfaceName: 'ProjectedMapInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get min_z() {
    if (!this._min_zIntialized) {
      return undefined;
    }
    return this._refObject.min_z;
  }
  set min_z(value) {
    this._min_zIntialized = true;
    this._refObject.min_z = value;
  }
  get max_z() {
    if (!this._max_zIntialized) {
      return undefined;
    }
    return this._refObject.max_z;
  }
  set max_z(value) {
    this._max_zIntialized = true;
    this._refObject.max_z = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ProjectedMapInfoRefStruct(refObject.toObject());
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
    this._xIntialized = true;
    this._yIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._min_zIntialized = true;
    this._max_zIntialized = true;
  }
  copy(other) {
    this._refObject = new ProjectedMapInfoRefStruct(other._refObject.toObject());
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
    this._xIntialized = true;
    this._yIntialized = true;
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._min_zIntialized = true;
    this._max_zIntialized = true;
  }
  static get classType() {
    return ProjectedMapInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"min_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"map_msgs","type":"ProjectedMapInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ProjectedMapInfo"};
  }
  hasMember(name) {
    let memberNames = ["frame_id","x","y","width","height","min_z","max_z"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ProjectedMapInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ProjectedMapInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ProjectedMapInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ProjectedMapInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ProjectedMapInfoRefArray(size);
    this._refObject = new ProjectedMapInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ProjectedMapInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ProjectedMapInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ProjectedMapInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ProjectedMapInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ProjectedMapInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ProjectedMapInfoArrayWrapper;
  }
}
module.exports = ProjectedMapInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "min_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "map_msgs",
    "type": "ProjectedMapInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ProjectedMapInfo"
}
*/
