// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__LaserScan.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const LaserScanRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  frame_id: primitiveTypes.string,
  pose: PoseWrapper.refObjectType,
  start_angle: primitiveTypes.float64,
  end_angle: primitiveTypes.float64,
  ranges: Float64Wrapper.refObjectArrayType,
  intensities: Float64Wrapper.refObjectArrayType,
});
const LaserScanRefArray = ArrayType(LaserScanRefStruct);
const LaserScanRefStructArray = StructType({
  data: LaserScanRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LaserScanWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LaserScanRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._start_angleIntialized = true;
      this._end_angleIntialized = true;
      this._wrapperFields.ranges = Float64Wrapper.createArray();
      this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
      this._wrapperFields.intensities = Float64Wrapper.createArray();
      this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LaserScanRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
    this._wrapperFields.pose = new PoseWrapper();
    this._start_angleIntialized = false;
    this._end_angleIntialized = false;
    this._wrapperFields.ranges = Float64Wrapper.createArray();
    this._wrapperFields.intensities = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LaserScanWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LaserScanArrayWrapper;
  }
  static get ArrayType() {
    return LaserScanArrayWrapper;
  }
  static get refObjectArrayType() {
    return LaserScanRefStructArray
  }
  static get refObjectType() {
    return LaserScanRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in LaserScan');
      }
      if (!this._start_angleIntialized) {
        throw new TypeError('Invalid argument: start_angle in LaserScan');
      }
      if (!this._end_angleIntialized) {
        throw new TypeError('Invalid argument: end_angle in LaserScan');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.ranges.freeze(own, checkConsistency);
    this._refObject.ranges = this._wrapperFields.ranges.refObject;
    if (own) {
      this._wrapperFields.ranges.fill(Float64Array.from([]));
      this._wrapperFields.ranges.freeze(own, checkConsistency);
      this._refObject.ranges = this._wrapperFields.ranges.refObject;
    }
    this._wrapperFields.intensities.freeze(own, checkConsistency);
    this._refObject.intensities = this._wrapperFields.intensities.refObject;
    if (own) {
      this._wrapperFields.intensities.fill(Float64Array.from([]));
      this._wrapperFields.intensities.freeze(own, checkConsistency);
      this._refObject.intensities = this._wrapperFields.intensities.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._start_angleIntialized = true;
    this._refObject.start_angle = refObject.start_angle;
    this._end_angleIntialized = true;
    this._refObject.end_angle = refObject.end_angle;
    this._wrapperFields.ranges.copyRefObject(refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(refObject.intensities);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    StringWrapper.freeStruct(refObject.frame_id);
    PoseWrapper.freeStruct(refObject.pose);
    if (refObject.ranges.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.ranges);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ranges, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.intensities.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.intensities);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.intensities, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    LaserScanWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'LaserScan'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get start_angle() {
    if (!this._start_angleIntialized) {
      return undefined;
    }
    return this._refObject.start_angle;
  }
  set start_angle(value) {
    this._start_angleIntialized = true;
    this._refObject.start_angle = value;
  }
  get end_angle() {
    if (!this._end_angleIntialized) {
      return undefined;
    }
    return this._refObject.end_angle;
  }
  set end_angle(value) {
    this._end_angleIntialized = true;
    this._refObject.end_angle = value;
  }
  get ranges() {
    return this._wrapperFields['ranges'].data;
  }
  set ranges(value) {
    this._wrapperFields['ranges'].fill(value);
  }
  get intensities() {
    return this._wrapperFields['intensities'].data;
  }
  set intensities(value) {
    this._wrapperFields['intensities'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LaserScanRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._start_angleIntialized = true;
    this._end_angleIntialized = true;
    this._wrapperFields.ranges.copyRefObject(this._refObject.ranges);
    this._wrapperFields.intensities.copyRefObject(this._refObject.intensities);
  }
  copy(other) {
    this._refObject = new LaserScanRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._start_angleIntialized = true;
    this._end_angleIntialized = true;
    this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
    this._wrapperFields.intensities.copy(other._wrapperFields.intensities);
  }
  static get classType() {
    return LaserScanWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"end_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"ranges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"intensities","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"LaserScan","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LaserScan"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","frame_id","pose","start_angle","end_angle","ranges","intensities"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LaserScanArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LaserScanWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LaserScanWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserScanArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LaserScanArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LaserScanRefArray(size);
    this._refObject = new LaserScanRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LaserScanWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LaserScanArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LaserScanArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LaserScanWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LaserScanWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LaserScanArrayWrapper;
  }
}
module.exports = LaserScanWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "end_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "ranges",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "intensities",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "LaserScan",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LaserScan"
}
*/
