// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__Grid.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Vector2Wrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__Vector2.js');
const PackedElementFieldWrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__PackedElementField.js');
const UInt8Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt8.js');
const GridRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  frame_id: primitiveTypes.string,
  pose: PoseWrapper.refObjectType,
  column_count: primitiveTypes.uint32,
  cell_size: Vector2Wrapper.refObjectType,
  row_stride: primitiveTypes.uint32,
  cell_stride: primitiveTypes.uint32,
  fields: PackedElementFieldWrapper.refObjectArrayType,
  data: UInt8Wrapper.refObjectArrayType,
});
const GridRefArray = ArrayType(GridRefStruct);
const GridRefStructArray = StructType({
  data: GridRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GridWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GridRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
      this._wrapperFields.pose =  new PoseWrapper(other._wrapperFields.pose);
      this._column_countIntialized = true;
      this._wrapperFields.cell_size =  new Vector2Wrapper(other._wrapperFields.cell_size);
      this._row_strideIntialized = true;
      this._cell_strideIntialized = true;
      this._wrapperFields.fields = PackedElementFieldWrapper.createArray();
      this._wrapperFields.fields.copy(other._wrapperFields.fields);
      this._wrapperFields.data = UInt8Wrapper.createArray();
      this._wrapperFields.data.copy(other._wrapperFields.data);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GridRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
    this._wrapperFields.pose = new PoseWrapper();
    this._column_countIntialized = false;
    this._wrapperFields.cell_size = new Vector2Wrapper();
    this._row_strideIntialized = false;
    this._cell_strideIntialized = false;
    this._wrapperFields.fields = PackedElementFieldWrapper.createArray();
    this._wrapperFields.data = UInt8Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GridWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GridArrayWrapper;
  }
  static get ArrayType() {
    return GridArrayWrapper;
  }
  static get refObjectArrayType() {
    return GridRefStructArray
  }
  static get refObjectType() {
    return GridRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in Grid');
      }
      if (!this._column_countIntialized) {
        throw new TypeError('Invalid argument: column_count in Grid');
      }
      if (!this._row_strideIntialized) {
        throw new TypeError('Invalid argument: row_stride in Grid');
      }
      if (!this._cell_strideIntialized) {
        throw new TypeError('Invalid argument: cell_stride in Grid');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
    this._wrapperFields.pose.freeze(own, checkConsistency);
    this._refObject.pose = this._wrapperFields.pose.refObject;
    this._wrapperFields.cell_size.freeze(own, checkConsistency);
    this._refObject.cell_size = this._wrapperFields.cell_size.refObject;
    this._wrapperFields.fields.freeze(own, checkConsistency);
    this._refObject.fields = this._wrapperFields.fields.refObject;
    this._wrapperFields.data.freeze(own, checkConsistency);
    this._refObject.data = this._wrapperFields.data.refObject;
    if (own) {
      this._wrapperFields.data.fill(Uint8Array.from([]));
      this._wrapperFields.data.freeze(own, checkConsistency);
      this._refObject.data = this._wrapperFields.data.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
    this._wrapperFields.pose.copyRefObject(refObject.pose);
    this._column_countIntialized = true;
    this._refObject.column_count = refObject.column_count;
    this._wrapperFields.cell_size.copyRefObject(refObject.cell_size);
    this._row_strideIntialized = true;
    this._refObject.row_stride = refObject.row_stride;
    this._cell_strideIntialized = true;
    this._refObject.cell_stride = refObject.cell_stride;
    this._wrapperFields.fields.copyRefObject(refObject.fields);
    this._wrapperFields.data.copyRefObject(refObject.data);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    StringWrapper.freeStruct(refObject.frame_id);
    PoseWrapper.freeStruct(refObject.pose);
    Vector2Wrapper.freeStruct(refObject.cell_size);
    if (refObject.fields.size != 0) {
      PackedElementFieldWrapper.ArrayType.freeArray(refObject.fields);
      if (PackedElementFieldWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fields, PackedElementFieldWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.data.size != 0) {
      UInt8Wrapper.ArrayType.freeArray(refObject.data);
      if (UInt8Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.data, UInt8Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GridWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'Grid'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  get pose() {
    return this._wrapperFields.pose;
  }
  set pose(value) {
    if (value instanceof PoseWrapper) {
      this._wrapperFields.pose.copy(value);
    } else {
      this._wrapperFields.pose.copy(new PoseWrapper(value));
    }
  }
  get column_count() {
    if (!this._column_countIntialized) {
      return undefined;
    }
    return this._refObject.column_count;
  }
  set column_count(value) {
    this._column_countIntialized = true;
    this._refObject.column_count = value;
  }
  get cell_size() {
    return this._wrapperFields.cell_size;
  }
  set cell_size(value) {
    if (value instanceof Vector2Wrapper) {
      this._wrapperFields.cell_size.copy(value);
    } else {
      this._wrapperFields.cell_size.copy(new Vector2Wrapper(value));
    }
  }
  get row_stride() {
    if (!this._row_strideIntialized) {
      return undefined;
    }
    return this._refObject.row_stride;
  }
  set row_stride(value) {
    this._row_strideIntialized = true;
    this._refObject.row_stride = value;
  }
  get cell_stride() {
    if (!this._cell_strideIntialized) {
      return undefined;
    }
    return this._refObject.cell_stride;
  }
  set cell_stride(value) {
    this._cell_strideIntialized = true;
    this._refObject.cell_stride = value;
  }
  get fields() {
    return this._wrapperFields.fields;
  }
  set fields(value) {
    this._wrapperFields.fields.fill(value);
  }
  get data() {
    return this._wrapperFields['data'].data;
  }
  set data(value) {
    this._wrapperFields['data'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GridRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
    this._wrapperFields.pose.copyRefObject(this._refObject.pose);
    this._column_countIntialized = true;
    this._wrapperFields.cell_size.copyRefObject(this._refObject.cell_size);
    this._row_strideIntialized = true;
    this._cell_strideIntialized = true;
    this._wrapperFields.fields.copyRefObject(this._refObject.fields);
    this._wrapperFields.data.copyRefObject(this._refObject.data);
  }
  copy(other) {
    this._refObject = new GridRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
    this._wrapperFields.pose.copy(other._wrapperFields.pose);
    this._column_countIntialized = true;
    this._wrapperFields.cell_size.copy(other._wrapperFields.cell_size);
    this._row_strideIntialized = true;
    this._cell_strideIntialized = true;
    this._wrapperFields.fields.copy(other._wrapperFields.fields);
    this._wrapperFields.data.copy(other._wrapperFields.data);
  }
  static get classType() {
    return GridWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"column_count","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cell_size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"foxglove_msgs","type":"Vector2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"row_stride","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"cell_stride","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fields","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"foxglove_msgs","type":"PackedElementField","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"data","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"Grid","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Grid"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","frame_id","pose","column_count","cell_size","row_stride","cell_stride","fields","data"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GridArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GridWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GridWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GridArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GridRefArray(size);
    this._refObject = new GridRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GridWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GridArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GridArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GridWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GridWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GridArrayWrapper;
  }
}
module.exports = GridWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "column_count",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cell_size",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "foxglove_msgs",
        "type": "Vector2",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "row_stride",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "cell_stride",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fields",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "foxglove_msgs",
        "type": "PackedElementField",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "data",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "Grid",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Grid"
}
*/
