// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__PackedElementField.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const PackedElementFieldRefStruct = StructType({
  name: primitiveTypes.string,
  offset: primitiveTypes.uint32,
  type: primitiveTypes.uint8,
});
const PackedElementFieldRefArray = ArrayType(PackedElementFieldRefStruct);
const PackedElementFieldRefStructArray = StructType({
  data: PackedElementFieldRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PackedElementFieldWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PackedElementFieldRefStruct(other._refObject.toObject());
      this._nameIntialized = true;
      this._wrapperFields.name =  new StringWrapper(other._wrapperFields.name);
      this._offsetIntialized = true;
      this._typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PackedElementFieldRefStruct();
    this._nameIntialized = false;
    this._wrapperFields.name = new StringWrapper();
    this._offsetIntialized = false;
    this._typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PackedElementFieldWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PackedElementFieldArrayWrapper;
  }
  static get ArrayType() {
    return PackedElementFieldArrayWrapper;
  }
  static get refObjectArrayType() {
    return PackedElementFieldRefStructArray
  }
  static get refObjectType() {
    return PackedElementFieldRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._nameIntialized) {
        throw new TypeError('Invalid argument: name in PackedElementField');
      }
      if (!this._offsetIntialized) {
        throw new TypeError('Invalid argument: offset in PackedElementField');
      }
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in PackedElementField');
      }
    }
    if (own) {
      this._wrapperFields.name.freeze(own, checkConsistency);
    }
    this._refObject.name = this._wrapperFields.name.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = refObject.name.data;
    this._offsetIntialized = true;
    this._refObject.offset = refObject.offset;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.name);
  }
  static destoryRawROS(msg) {
    PackedElementFieldWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'PackedElementField'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get name() {
    if (!this._nameIntialized) {
      return undefined;
    }
    return this._wrapperFields.name.data;
  }
  set name(value) {
    this._nameIntialized = true;
    this._wrapperFields.name.data = value;
  }
  get offset() {
    if (!this._offsetIntialized) {
      return undefined;
    }
    return this._refObject.offset;
  }
  set offset(value) {
    this._offsetIntialized = true;
    this._refObject.offset = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PackedElementFieldRefStruct(refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copyRefObject(this._refObject.name);
    this._offsetIntialized = true;
    this._typeIntialized = true;
  }
  copy(other) {
    this._refObject = new PackedElementFieldRefStruct(other._refObject.toObject());
    this._nameIntialized = true;
    this._wrapperFields.name.copy(other._wrapperFields.name);
    this._offsetIntialized = true;
    this._typeIntialized = true;
  }
  static get classType() {
    return PackedElementFieldWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"UNKNOWN","value":0},{"type":"uint8","name":"UINT8","value":1},{"type":"uint8","name":"INT8","value":2},{"type":"uint8","name":"UINT16","value":3},{"type":"uint8","name":"INT16","value":4},{"type":"uint8","name":"UINT32","value":5},{"type":"uint8","name":"INT32","value":6},{"type":"uint8","name":"FLOAT32","value":7},{"type":"uint8","name":"FLOAT64","value":8}],"fields":[{"name":"name","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"offset","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"PackedElementField","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PackedElementField"};
  }
  hasMember(name) {
    let memberNames = ["name","offset","type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PackedElementFieldArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PackedElementFieldWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PackedElementFieldWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PackedElementFieldArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PackedElementFieldArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PackedElementFieldRefArray(size);
    this._refObject = new PackedElementFieldRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PackedElementFieldWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PackedElementFieldArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PackedElementFieldArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PackedElementFieldWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PackedElementFieldWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PackedElementFieldArrayWrapper;
  }
}
// Define constants (9 in total)
Object.defineProperty(PackedElementFieldWrapper, "UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "UINT8", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "INT8", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "UINT16", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "INT16", {value: 4, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "UINT32", {value: 5, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "INT32", {value: 6, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "FLOAT32", {value: 7, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PackedElementFieldWrapper, "FLOAT64", {value: 8, writable: false, enumerable: true, configurable: true});
module.exports = PackedElementFieldWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "UINT8",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "INT8",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "UINT16",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "INT16",
      "value": 4
    },
    {
      "type": "uint8",
      "name": "UINT32",
      "value": 5
    },
    {
      "type": "uint8",
      "name": "INT32",
      "value": 6
    },
    {
      "type": "uint8",
      "name": "FLOAT32",
      "value": 7
    },
    {
      "type": "uint8",
      "name": "FLOAT64",
      "value": 8
    }
  ],
  "fields": [
    {
      "name": "name",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "offset",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "PackedElementField",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PackedElementField"
}
*/
