// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__FrameTransform.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Vector3Wrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Vector3.js');
const QuaternionWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Quaternion.js');
const FrameTransformRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  parent_frame_id: primitiveTypes.string,
  child_frame_id: primitiveTypes.string,
  translation: Vector3Wrapper.refObjectType,
  rotation: QuaternionWrapper.refObjectType,
});
const FrameTransformRefArray = ArrayType(FrameTransformRefStruct);
const FrameTransformRefStructArray = StructType({
  data: FrameTransformRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FrameTransformWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FrameTransformRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._parent_frame_idIntialized = true;
      this._wrapperFields.parent_frame_id =  new StringWrapper(other._wrapperFields.parent_frame_id);
      this._child_frame_idIntialized = true;
      this._wrapperFields.child_frame_id =  new StringWrapper(other._wrapperFields.child_frame_id);
      this._wrapperFields.translation =  new Vector3Wrapper(other._wrapperFields.translation);
      this._wrapperFields.rotation =  new QuaternionWrapper(other._wrapperFields.rotation);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FrameTransformRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._parent_frame_idIntialized = false;
    this._wrapperFields.parent_frame_id = new StringWrapper();
    this._child_frame_idIntialized = false;
    this._wrapperFields.child_frame_id = new StringWrapper();
    this._wrapperFields.translation = new Vector3Wrapper();
    this._wrapperFields.rotation = new QuaternionWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new FrameTransformWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FrameTransformArrayWrapper;
  }
  static get ArrayType() {
    return FrameTransformArrayWrapper;
  }
  static get refObjectArrayType() {
    return FrameTransformRefStructArray
  }
  static get refObjectType() {
    return FrameTransformRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._parent_frame_idIntialized) {
        throw new TypeError('Invalid argument: parent_frame_id in FrameTransform');
      }
      if (!this._child_frame_idIntialized) {
        throw new TypeError('Invalid argument: child_frame_id in FrameTransform');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    if (own) {
      this._wrapperFields.parent_frame_id.freeze(own, checkConsistency);
    }
    this._refObject.parent_frame_id = this._wrapperFields.parent_frame_id.refObject;
    if (own) {
      this._wrapperFields.child_frame_id.freeze(own, checkConsistency);
    }
    this._refObject.child_frame_id = this._wrapperFields.child_frame_id.refObject;
    this._wrapperFields.translation.freeze(own, checkConsistency);
    this._refObject.translation = this._wrapperFields.translation.refObject;
    this._wrapperFields.rotation.freeze(own, checkConsistency);
    this._refObject.rotation = this._wrapperFields.rotation.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._parent_frame_idIntialized = true;
    this._wrapperFields.parent_frame_id.data = refObject.parent_frame_id.data;
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.data = refObject.child_frame_id.data;
    this._wrapperFields.translation.copyRefObject(refObject.translation);
    this._wrapperFields.rotation.copyRefObject(refObject.rotation);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    StringWrapper.freeStruct(refObject.parent_frame_id);
    StringWrapper.freeStruct(refObject.child_frame_id);
    Vector3Wrapper.freeStruct(refObject.translation);
    QuaternionWrapper.freeStruct(refObject.rotation);
  }
  static destoryRawROS(msg) {
    FrameTransformWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'FrameTransform'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get parent_frame_id() {
    if (!this._parent_frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.parent_frame_id.data;
  }
  set parent_frame_id(value) {
    this._parent_frame_idIntialized = true;
    this._wrapperFields.parent_frame_id.data = value;
  }
  get child_frame_id() {
    if (!this._child_frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.child_frame_id.data;
  }
  set child_frame_id(value) {
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.data = value;
  }
  get translation() {
    return this._wrapperFields.translation;
  }
  set translation(value) {
    if (value instanceof Vector3Wrapper) {
      this._wrapperFields.translation.copy(value);
    } else {
      this._wrapperFields.translation.copy(new Vector3Wrapper(value));
    }
  }
  get rotation() {
    return this._wrapperFields.rotation;
  }
  set rotation(value) {
    if (value instanceof QuaternionWrapper) {
      this._wrapperFields.rotation.copy(value);
    } else {
      this._wrapperFields.rotation.copy(new QuaternionWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new FrameTransformRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._parent_frame_idIntialized = true;
    this._wrapperFields.parent_frame_id.copyRefObject(this._refObject.parent_frame_id);
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.copyRefObject(this._refObject.child_frame_id);
    this._wrapperFields.translation.copyRefObject(this._refObject.translation);
    this._wrapperFields.rotation.copyRefObject(this._refObject.rotation);
  }
  copy(other) {
    this._refObject = new FrameTransformRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._parent_frame_idIntialized = true;
    this._wrapperFields.parent_frame_id.copy(other._wrapperFields.parent_frame_id);
    this._child_frame_idIntialized = true;
    this._wrapperFields.child_frame_id.copy(other._wrapperFields.child_frame_id);
    this._wrapperFields.translation.copy(other._wrapperFields.translation);
    this._wrapperFields.rotation.copy(other._wrapperFields.rotation);
  }
  static get classType() {
    return FrameTransformWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"parent_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"child_frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"translation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Vector3","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"rotation","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Quaternion","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"FrameTransform","stringUpperBound":null,"isPrimitiveType":false},"msgName":"FrameTransform"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","parent_frame_id","child_frame_id","translation","rotation"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FrameTransformArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FrameTransformWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FrameTransformWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameTransformArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FrameTransformArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FrameTransformRefArray(size);
    this._refObject = new FrameTransformRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FrameTransformWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FrameTransformArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FrameTransformArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FrameTransformWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FrameTransformWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FrameTransformArrayWrapper;
  }
}
module.exports = FrameTransformWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "parent_frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "child_frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "translation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Vector3",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "rotation",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Quaternion",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "FrameTransform",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "FrameTransform"
}
*/
