// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__PointsAnnotation.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const Point2Wrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__Point2.js');
const ColorWrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__Color.js');
const PointsAnnotationRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  type: primitiveTypes.uint8,
  points: Point2Wrapper.refObjectArrayType,
  outline_colors: ColorWrapper.refObjectArrayType,
  fill_color: ColorWrapper.refObjectType,
});
const PointsAnnotationRefArray = ArrayType(PointsAnnotationRefStruct);
const PointsAnnotationRefStructArray = StructType({
  data: PointsAnnotationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PointsAnnotationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PointsAnnotationRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._typeIntialized = true;
      this._wrapperFields.points = Point2Wrapper.createArray();
      this._wrapperFields.points.copy(other._wrapperFields.points);
      this._wrapperFields.outline_colors = ColorWrapper.createArray();
      this._wrapperFields.outline_colors.copy(other._wrapperFields.outline_colors);
      this._wrapperFields.fill_color =  new ColorWrapper(other._wrapperFields.fill_color);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PointsAnnotationRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._typeIntialized = false;
    this._wrapperFields.points = Point2Wrapper.createArray();
    this._wrapperFields.outline_colors = ColorWrapper.createArray();
    this._wrapperFields.fill_color = new ColorWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new PointsAnnotationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PointsAnnotationArrayWrapper;
  }
  static get ArrayType() {
    return PointsAnnotationArrayWrapper;
  }
  static get refObjectArrayType() {
    return PointsAnnotationRefStructArray
  }
  static get refObjectType() {
    return PointsAnnotationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in PointsAnnotation');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
    this._wrapperFields.outline_colors.freeze(own, checkConsistency);
    this._refObject.outline_colors = this._wrapperFields.outline_colors.refObject;
    this._wrapperFields.fill_color.freeze(own, checkConsistency);
    this._refObject.fill_color = this._wrapperFields.fill_color.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._wrapperFields.points.copyRefObject(refObject.points);
    this._wrapperFields.outline_colors.copyRefObject(refObject.outline_colors);
    this._wrapperFields.fill_color.copyRefObject(refObject.fill_color);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    if (refObject.points.size != 0) {
      Point2Wrapper.ArrayType.freeArray(refObject.points);
      if (Point2Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.points, Point2Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.outline_colors.size != 0) {
      ColorWrapper.ArrayType.freeArray(refObject.outline_colors);
      if (ColorWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.outline_colors, ColorWrapper.refObjectArrayType, 'data');
      }
    }
    ColorWrapper.freeStruct(refObject.fill_color);
  }
  static destoryRawROS(msg) {
    PointsAnnotationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'PointsAnnotation'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    this._wrapperFields.points.fill(value);
  }
  get outline_colors() {
    return this._wrapperFields.outline_colors;
  }
  set outline_colors(value) {
    this._wrapperFields.outline_colors.fill(value);
  }
  get fill_color() {
    return this._wrapperFields.fill_color;
  }
  set fill_color(value) {
    if (value instanceof ColorWrapper) {
      this._wrapperFields.fill_color.copy(value);
    } else {
      this._wrapperFields.fill_color.copy(new ColorWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new PointsAnnotationRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._typeIntialized = true;
    this._wrapperFields.points.copyRefObject(this._refObject.points);
    this._wrapperFields.outline_colors.copyRefObject(this._refObject.outline_colors);
    this._wrapperFields.fill_color.copyRefObject(this._refObject.fill_color);
  }
  copy(other) {
    this._refObject = new PointsAnnotationRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._typeIntialized = true;
    this._wrapperFields.points.copy(other._wrapperFields.points);
    this._wrapperFields.outline_colors.copy(other._wrapperFields.outline_colors);
    this._wrapperFields.fill_color.copy(other._wrapperFields.fill_color);
  }
  static get classType() {
    return PointsAnnotationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"UNKNOWN","value":0},{"type":"uint8","name":"POINTS","value":1},{"type":"uint8","name":"LINE_LOOP","value":2},{"type":"uint8","name":"LINE_STRIP","value":3},{"type":"uint8","name":"LINE_LIST","value":4}],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"foxglove_msgs","type":"Point2","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"outline_colors","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"foxglove_msgs","type":"Color","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"fill_color","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"foxglove_msgs","type":"Color","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"PointsAnnotation","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PointsAnnotation"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","type","points","outline_colors","fill_color"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PointsAnnotationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PointsAnnotationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PointsAnnotationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointsAnnotationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PointsAnnotationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PointsAnnotationRefArray(size);
    this._refObject = new PointsAnnotationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PointsAnnotationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PointsAnnotationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PointsAnnotationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PointsAnnotationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PointsAnnotationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PointsAnnotationArrayWrapper;
  }
}
// Define constants (5 in total)
Object.defineProperty(PointsAnnotationWrapper, "UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PointsAnnotationWrapper, "POINTS", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PointsAnnotationWrapper, "LINE_LOOP", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PointsAnnotationWrapper, "LINE_STRIP", {value: 3, writable: false, enumerable: true, configurable: true});
Object.defineProperty(PointsAnnotationWrapper, "LINE_LIST", {value: 4, writable: false, enumerable: true, configurable: true});
module.exports = PointsAnnotationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "POINTS",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "LINE_LOOP",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "LINE_STRIP",
      "value": 3
    },
    {
      "type": "uint8",
      "name": "LINE_LIST",
      "value": 4
    }
  ],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "foxglove_msgs",
        "type": "Point2",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "outline_colors",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "foxglove_msgs",
        "type": "Color",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "fill_color",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "foxglove_msgs",
        "type": "Color",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "PointsAnnotation",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PointsAnnotation"
}
*/
