// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__ImageAnnotations.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CircleAnnotationWrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__CircleAnnotation.js');
const PointsAnnotationWrapper = require('../../generated/foxglove_msgs/foxglove_msgs__msg__PointsAnnotation.js');
const ImageAnnotationsRefStruct = StructType({
  circles: CircleAnnotationWrapper.refObjectArrayType,
  points: PointsAnnotationWrapper.refObjectArrayType,
});
const ImageAnnotationsRefArray = ArrayType(ImageAnnotationsRefStruct);
const ImageAnnotationsRefStructArray = StructType({
  data: ImageAnnotationsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ImageAnnotationsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ImageAnnotationsRefStruct(other._refObject.toObject());
      this._wrapperFields.circles = CircleAnnotationWrapper.createArray();
      this._wrapperFields.circles.copy(other._wrapperFields.circles);
      this._wrapperFields.points = PointsAnnotationWrapper.createArray();
      this._wrapperFields.points.copy(other._wrapperFields.points);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ImageAnnotationsRefStruct();
    this._wrapperFields.circles = CircleAnnotationWrapper.createArray();
    this._wrapperFields.points = PointsAnnotationWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ImageAnnotationsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ImageAnnotationsArrayWrapper;
  }
  static get ArrayType() {
    return ImageAnnotationsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ImageAnnotationsRefStructArray
  }
  static get refObjectType() {
    return ImageAnnotationsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.circles.freeze(own, checkConsistency);
    this._refObject.circles = this._wrapperFields.circles.refObject;
    this._wrapperFields.points.freeze(own, checkConsistency);
    this._refObject.points = this._wrapperFields.points.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.circles.copyRefObject(refObject.circles);
    this._wrapperFields.points.copyRefObject(refObject.points);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.circles.size != 0) {
      CircleAnnotationWrapper.ArrayType.freeArray(refObject.circles);
      if (CircleAnnotationWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.circles, CircleAnnotationWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.points.size != 0) {
      PointsAnnotationWrapper.ArrayType.freeArray(refObject.points);
      if (PointsAnnotationWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.points, PointsAnnotationWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ImageAnnotationsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'ImageAnnotations'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get circles() {
    return this._wrapperFields.circles;
  }
  set circles(value) {
    this._wrapperFields.circles.fill(value);
  }
  get points() {
    return this._wrapperFields.points;
  }
  set points(value) {
    this._wrapperFields.points.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ImageAnnotationsRefStruct(refObject.toObject());
    this._wrapperFields.circles.copyRefObject(this._refObject.circles);
    this._wrapperFields.points.copyRefObject(this._refObject.points);
  }
  copy(other) {
    this._refObject = new ImageAnnotationsRefStruct(other._refObject.toObject());
    this._wrapperFields.circles.copy(other._wrapperFields.circles);
    this._wrapperFields.points.copy(other._wrapperFields.points);
  }
  static get classType() {
    return ImageAnnotationsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"circles","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"foxglove_msgs","type":"CircleAnnotation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"foxglove_msgs","type":"PointsAnnotation","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"ImageAnnotations","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ImageAnnotations"};
  }
  hasMember(name) {
    let memberNames = ["circles","points"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ImageAnnotationsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ImageAnnotationsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ImageAnnotationsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ImageAnnotationsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ImageAnnotationsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ImageAnnotationsRefArray(size);
    this._refObject = new ImageAnnotationsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ImageAnnotationsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ImageAnnotationsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ImageAnnotationsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ImageAnnotationsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ImageAnnotationsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ImageAnnotationsArrayWrapper;
  }
}
module.exports = ImageAnnotationsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "circles",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "foxglove_msgs",
        "type": "CircleAnnotation",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "foxglove_msgs",
        "type": "PointsAnnotation",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "ImageAnnotations",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ImageAnnotations"
}
*/
