// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__LocationFix.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const LocationFixRefStruct = StructType({
  latitude: primitiveTypes.float64,
  longitude: primitiveTypes.float64,
  altitude: primitiveTypes.float64,
  position_covariance: ArrayType(primitiveTypes.float64, 9),
  position_covariance_type: primitiveTypes.uint8,
});
const LocationFixRefArray = ArrayType(LocationFixRefStruct);
const LocationFixRefStructArray = StructType({
  data: LocationFixRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LocationFixWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LocationFixRefStruct(other._refObject.toObject());
      this._latitudeIntialized = true;
      this._longitudeIntialized = true;
      this._altitudeIntialized = true;
      this._wrapperFields.position_covariance = Float64Wrapper.createArray();
      this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
      this._position_covariance_typeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LocationFixRefStruct();
    this._latitudeIntialized = false;
    this._longitudeIntialized = false;
    this._altitudeIntialized = false;
    this._wrapperFields.position_covariance = Float64Wrapper.createArray();
    this._position_covariance_typeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LocationFixWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LocationFixArrayWrapper;
  }
  static get ArrayType() {
    return LocationFixArrayWrapper;
  }
  static get refObjectArrayType() {
    return LocationFixRefStructArray
  }
  static get refObjectType() {
    return LocationFixRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._latitudeIntialized) {
        throw new TypeError('Invalid argument: latitude in LocationFix');
      }
      if (!this._longitudeIntialized) {
        throw new TypeError('Invalid argument: longitude in LocationFix');
      }
      if (!this._altitudeIntialized) {
        throw new TypeError('Invalid argument: altitude in LocationFix');
      }
      if (!this._position_covariance_typeIntialized) {
        throw new TypeError('Invalid argument: position_covariance_type in LocationFix');
      }
    }
    for (let i = 0; i < this._wrapperFields.position_covariance.data.length; i++) {
      this._refObject.position_covariance[i] = this._wrapperFields.position_covariance.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._latitudeIntialized = true;
    this._refObject.latitude = refObject.latitude;
    this._longitudeIntialized = true;
    this._refObject.longitude = refObject.longitude;
    this._altitudeIntialized = true;
    this._refObject.altitude = refObject.altitude;
    this._refObject.position_covariance = refObject.position_covariance;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = refObject.position_covariance_type;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    LocationFixWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'LocationFix'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get latitude() {
    if (!this._latitudeIntialized) {
      return undefined;
    }
    return this._refObject.latitude;
  }
  set latitude(value) {
    this._latitudeIntialized = true;
    this._refObject.latitude = value;
  }
  get longitude() {
    if (!this._longitudeIntialized) {
      return undefined;
    }
    return this._refObject.longitude;
  }
  set longitude(value) {
    this._longitudeIntialized = true;
    this._refObject.longitude = value;
  }
  get altitude() {
    if (!this._altitudeIntialized) {
      return undefined;
    }
    return this._refObject.altitude;
  }
  set altitude(value) {
    this._altitudeIntialized = true;
    this._refObject.altitude = value;
  }
  get position_covariance() {
    return this._wrapperFields['position_covariance'].data;
  }
  set position_covariance(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['position_covariance'].fill(value);
  }
  get position_covariance_type() {
    if (!this._position_covariance_typeIntialized) {
      return undefined;
    }
    return this._refObject.position_covariance_type;
  }
  set position_covariance_type(value) {
    this._position_covariance_typeIntialized = true;
    this._refObject.position_covariance_type = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LocationFixRefStruct(refObject.toObject());
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.position_covariance.fill(refObject.position_covariance.toArray());
    this._position_covariance_typeIntialized = true;
  }
  copy(other) {
    this._refObject = new LocationFixRefStruct(other._refObject.toObject());
    this._latitudeIntialized = true;
    this._longitudeIntialized = true;
    this._altitudeIntialized = true;
    this._wrapperFields.position_covariance.copy(other._wrapperFields.position_covariance);
    this._position_covariance_typeIntialized = true;
  }
  static get classType() {
    return LocationFixWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"uint8","name":"UNKNOWN","value":0},{"type":"uint8","name":"APPROXIMATED","value":1},{"type":"uint8","name":"DIAGONAL_KNOWN","value":2},{"type":"uint8","name":"KNOWN","value":3}],"fields":[{"name":"latitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"longitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"altitude","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"position_covariance_type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"LocationFix","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LocationFix"};
  }
  hasMember(name) {
    let memberNames = ["latitude","longitude","altitude","position_covariance","position_covariance_type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LocationFixArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LocationFixWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LocationFixWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocationFixArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LocationFixArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LocationFixRefArray(size);
    this._refObject = new LocationFixRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LocationFixWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LocationFixArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LocationFixArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LocationFixWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LocationFixWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LocationFixArrayWrapper;
  }
}
// Define constants (4 in total)
Object.defineProperty(LocationFixWrapper, "UNKNOWN", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LocationFixWrapper, "APPROXIMATED", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LocationFixWrapper, "DIAGONAL_KNOWN", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(LocationFixWrapper, "KNOWN", {value: 3, writable: false, enumerable: true, configurable: true});
module.exports = LocationFixWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "uint8",
      "name": "UNKNOWN",
      "value": 0
    },
    {
      "type": "uint8",
      "name": "APPROXIMATED",
      "value": 1
    },
    {
      "type": "uint8",
      "name": "DIAGONAL_KNOWN",
      "value": 2
    },
    {
      "type": "uint8",
      "name": "KNOWN",
      "value": 3
    }
  ],
  "fields": [
    {
      "name": "latitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "longitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "altitude",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "position_covariance_type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "LocationFix",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LocationFix"
}
*/
