// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__GeoJSON.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GeoJSONRefStruct = StructType({
  geojson: primitiveTypes.string,
});
const GeoJSONRefArray = ArrayType(GeoJSONRefStruct);
const GeoJSONRefStructArray = StructType({
  data: GeoJSONRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GeoJSONWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GeoJSONRefStruct(other._refObject.toObject());
      this._geojsonIntialized = true;
      this._wrapperFields.geojson =  new StringWrapper(other._wrapperFields.geojson);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GeoJSONRefStruct();
    this._geojsonIntialized = false;
    this._wrapperFields.geojson = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GeoJSONWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GeoJSONArrayWrapper;
  }
  static get ArrayType() {
    return GeoJSONArrayWrapper;
  }
  static get refObjectArrayType() {
    return GeoJSONRefStructArray
  }
  static get refObjectType() {
    return GeoJSONRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._geojsonIntialized) {
        throw new TypeError('Invalid argument: geojson in GeoJSON');
      }
    }
    if (own) {
      this._wrapperFields.geojson.freeze(own, checkConsistency);
    }
    this._refObject.geojson = this._wrapperFields.geojson.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._geojsonIntialized = true;
    this._wrapperFields.geojson.data = refObject.geojson.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.geojson);
  }
  static destoryRawROS(msg) {
    GeoJSONWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'GeoJSON'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get geojson() {
    if (!this._geojsonIntialized) {
      return undefined;
    }
    return this._wrapperFields.geojson.data;
  }
  set geojson(value) {
    this._geojsonIntialized = true;
    this._wrapperFields.geojson.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GeoJSONRefStruct(refObject.toObject());
    this._geojsonIntialized = true;
    this._wrapperFields.geojson.copyRefObject(this._refObject.geojson);
  }
  copy(other) {
    this._refObject = new GeoJSONRefStruct(other._refObject.toObject());
    this._geojsonIntialized = true;
    this._wrapperFields.geojson.copy(other._wrapperFields.geojson);
  }
  static get classType() {
    return GeoJSONWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"geojson","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"GeoJSON","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GeoJSON"};
  }
  hasMember(name) {
    let memberNames = ["geojson"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GeoJSONArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GeoJSONWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GeoJSONWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeoJSONArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GeoJSONArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GeoJSONRefArray(size);
    this._refObject = new GeoJSONRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GeoJSONWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GeoJSONArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GeoJSONArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GeoJSONWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GeoJSONWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GeoJSONArrayWrapper;
  }
}
module.exports = GeoJSONWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "geojson",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "GeoJSON",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GeoJSON"
}
*/
