// rclnodejs/generated/foxglove_msgs/foxglove_msgs__msg__CameraCalibration.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const CameraCalibrationRefStruct = StructType({
  timestamp: TimeWrapper.refObjectType,
  width: primitiveTypes.uint32,
  height: primitiveTypes.uint32,
  distortion_model: primitiveTypes.string,
  d: Float64Wrapper.refObjectArrayType,
  k: ArrayType(primitiveTypes.float64, 9),
  r: ArrayType(primitiveTypes.float64, 9),
  p: ArrayType(primitiveTypes.float64, 12),
});
const CameraCalibrationRefArray = ArrayType(CameraCalibrationRefStruct);
const CameraCalibrationRefStructArray = StructType({
  data: CameraCalibrationRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CameraCalibrationWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CameraCalibrationRefStruct(other._refObject.toObject());
      this._wrapperFields.timestamp =  new TimeWrapper(other._wrapperFields.timestamp);
      this._widthIntialized = true;
      this._heightIntialized = true;
      this._distortion_modelIntialized = true;
      this._wrapperFields.distortion_model =  new StringWrapper(other._wrapperFields.distortion_model);
      this._wrapperFields.d = Float64Wrapper.createArray();
      this._wrapperFields.d.copy(other._wrapperFields.d);
      this._wrapperFields.k = Float64Wrapper.createArray();
      this._wrapperFields.k.copy(other._wrapperFields.k);
      this._wrapperFields.r = Float64Wrapper.createArray();
      this._wrapperFields.r.copy(other._wrapperFields.r);
      this._wrapperFields.p = Float64Wrapper.createArray();
      this._wrapperFields.p.copy(other._wrapperFields.p);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CameraCalibrationRefStruct();
    this._wrapperFields.timestamp = new TimeWrapper();
    this._widthIntialized = false;
    this._heightIntialized = false;
    this._distortion_modelIntialized = false;
    this._wrapperFields.distortion_model = new StringWrapper();
    this._wrapperFields.d = Float64Wrapper.createArray();
    this._wrapperFields.k = Float64Wrapper.createArray();
    this._wrapperFields.r = Float64Wrapper.createArray();
    this._wrapperFields.p = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new CameraCalibrationWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CameraCalibrationArrayWrapper;
  }
  static get ArrayType() {
    return CameraCalibrationArrayWrapper;
  }
  static get refObjectArrayType() {
    return CameraCalibrationRefStructArray
  }
  static get refObjectType() {
    return CameraCalibrationRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in CameraCalibration');
      }
      if (!this._heightIntialized) {
        throw new TypeError('Invalid argument: height in CameraCalibration');
      }
      if (!this._distortion_modelIntialized) {
        throw new TypeError('Invalid argument: distortion_model in CameraCalibration');
      }
    }
    this._wrapperFields.timestamp.freeze(own, checkConsistency);
    this._refObject.timestamp = this._wrapperFields.timestamp.refObject;
    if (own) {
      this._wrapperFields.distortion_model.freeze(own, checkConsistency);
    }
    this._refObject.distortion_model = this._wrapperFields.distortion_model.refObject;
    this._wrapperFields.d.freeze(own, checkConsistency);
    this._refObject.d = this._wrapperFields.d.refObject;
    if (own) {
      this._wrapperFields.d.fill(Float64Array.from([]));
      this._wrapperFields.d.freeze(own, checkConsistency);
      this._refObject.d = this._wrapperFields.d.refObject;
    }
    for (let i = 0; i < this._wrapperFields.k.data.length; i++) {
      this._refObject.k[i] = this._wrapperFields.k.data[i];
    }
    for (let i = 0; i < this._wrapperFields.r.data.length; i++) {
      this._refObject.r[i] = this._wrapperFields.r.data[i];
    }
    for (let i = 0; i < this._wrapperFields.p.data.length; i++) {
      this._refObject.p[i] = this._wrapperFields.p.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.timestamp.copyRefObject(refObject.timestamp);
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._heightIntialized = true;
    this._refObject.height = refObject.height;
    this._distortion_modelIntialized = true;
    this._wrapperFields.distortion_model.data = refObject.distortion_model.data;
    this._wrapperFields.d.copyRefObject(refObject.d);
    this._refObject.k = refObject.k;
    this._wrapperFields.k.fill(refObject.k.toArray());
    this._refObject.r = refObject.r;
    this._wrapperFields.r.fill(refObject.r.toArray());
    this._refObject.p = refObject.p;
    this._wrapperFields.p.fill(refObject.p.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TimeWrapper.freeStruct(refObject.timestamp);
    StringWrapper.freeStruct(refObject.distortion_model);
    if (refObject.d.size != 0) {
      Float64Wrapper.ArrayType.freeArray(refObject.d);
      if (Float64Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.d, Float64Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    CameraCalibrationWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'foxglove_msgs', subFolder: 'msg', interfaceName: 'CameraCalibration'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get timestamp() {
    return this._wrapperFields.timestamp;
  }
  set timestamp(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.timestamp.copy(value);
    } else {
      this._wrapperFields.timestamp.copy(new TimeWrapper(value));
    }
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get height() {
    if (!this._heightIntialized) {
      return undefined;
    }
    return this._refObject.height;
  }
  set height(value) {
    this._heightIntialized = true;
    this._refObject.height = value;
  }
  get distortion_model() {
    if (!this._distortion_modelIntialized) {
      return undefined;
    }
    return this._wrapperFields.distortion_model.data;
  }
  set distortion_model(value) {
    this._distortion_modelIntialized = true;
    this._wrapperFields.distortion_model.data = value;
  }
  get d() {
    return this._wrapperFields['d'].data;
  }
  set d(value) {
    this._wrapperFields['d'].fill(value);
  }
  get k() {
    return this._wrapperFields['k'].data;
  }
  set k(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['k'].fill(value);
  }
  get r() {
    return this._wrapperFields['r'].data;
  }
  set r(value) {
    if (value.length !== 9) {
      throw new RangeError('The length of the array must be 9.');
    }
    this._wrapperFields['r'].fill(value);
  }
  get p() {
    return this._wrapperFields['p'].data;
  }
  set p(value) {
    if (value.length !== 12) {
      throw new RangeError('The length of the array must be 12.');
    }
    this._wrapperFields['p'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new CameraCalibrationRefStruct(refObject.toObject());
    this._wrapperFields.timestamp.copyRefObject(this._refObject.timestamp);
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._distortion_modelIntialized = true;
    this._wrapperFields.distortion_model.copyRefObject(this._refObject.distortion_model);
    this._wrapperFields.d.copyRefObject(this._refObject.d);
    this._wrapperFields.k.fill(refObject.k.toArray());
    this._wrapperFields.r.fill(refObject.r.toArray());
    this._wrapperFields.p.fill(refObject.p.toArray());
  }
  copy(other) {
    this._refObject = new CameraCalibrationRefStruct(other._refObject.toObject());
    this._wrapperFields.timestamp.copy(other._wrapperFields.timestamp);
    this._widthIntialized = true;
    this._heightIntialized = true;
    this._distortion_modelIntialized = true;
    this._wrapperFields.distortion_model.copy(other._wrapperFields.distortion_model);
    this._wrapperFields.d.copy(other._wrapperFields.d);
    this._wrapperFields.k.copy(other._wrapperFields.k);
    this._wrapperFields.r.copy(other._wrapperFields.r);
    this._wrapperFields.p.copy(other._wrapperFields.p);
  }
  static get classType() {
    return CameraCalibrationWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"timestamp","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"height","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distortion_model","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"d","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"k","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"r","type":{"isArray":true,"arraySize":9,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"p","type":{"isArray":true,"arraySize":12,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"foxglove_msgs","type":"CameraCalibration","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CameraCalibration"};
  }
  hasMember(name) {
    let memberNames = ["timestamp","width","height","distortion_model","d","k","r","p"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CameraCalibrationArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CameraCalibrationWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CameraCalibrationWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraCalibrationArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CameraCalibrationArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CameraCalibrationRefArray(size);
    this._refObject = new CameraCalibrationRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CameraCalibrationWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CameraCalibrationArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CameraCalibrationArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CameraCalibrationWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CameraCalibrationWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CameraCalibrationArrayWrapper;
  }
}
module.exports = CameraCalibrationWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "timestamp",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "height",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distortion_model",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "d",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "k",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "r",
      "type": {
        "isArray": true,
        "arraySize": 9,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "p",
      "type": {
        "isArray": true,
        "arraySize": 12,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "foxglove_msgs",
    "type": "CameraCalibration",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CameraCalibration"
}
*/
