// rclnodejs/generated/rosbridge_msgs/rosbridge_msgs__msg__ConnectedClient.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const TimeWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Time.js');
const ConnectedClientRefStruct = StructType({
  ip_address: primitiveTypes.string,
  connection_time: TimeWrapper.refObjectType,
});
const ConnectedClientRefArray = ArrayType(ConnectedClientRefStruct);
const ConnectedClientRefStructArray = StructType({
  data: ConnectedClientRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ConnectedClientWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ConnectedClientRefStruct(other._refObject.toObject());
      this._ip_addressIntialized = true;
      this._wrapperFields.ip_address =  new StringWrapper(other._wrapperFields.ip_address);
      this._wrapperFields.connection_time =  new TimeWrapper(other._wrapperFields.connection_time);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ConnectedClientRefStruct();
    this._ip_addressIntialized = false;
    this._wrapperFields.ip_address = new StringWrapper();
    this._wrapperFields.connection_time = new TimeWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ConnectedClientWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ConnectedClientArrayWrapper;
  }
  static get ArrayType() {
    return ConnectedClientArrayWrapper;
  }
  static get refObjectArrayType() {
    return ConnectedClientRefStructArray
  }
  static get refObjectType() {
    return ConnectedClientRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._ip_addressIntialized) {
        throw new TypeError('Invalid argument: ip_address in ConnectedClient');
      }
    }
    if (own) {
      this._wrapperFields.ip_address.freeze(own, checkConsistency);
    }
    this._refObject.ip_address = this._wrapperFields.ip_address.refObject;
    this._wrapperFields.connection_time.freeze(own, checkConsistency);
    this._refObject.connection_time = this._wrapperFields.connection_time.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.data = refObject.ip_address.data;
    this._wrapperFields.connection_time.copyRefObject(refObject.connection_time);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.ip_address);
    TimeWrapper.freeStruct(refObject.connection_time);
  }
  static destoryRawROS(msg) {
    ConnectedClientWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbridge_msgs', subFolder: 'msg', interfaceName: 'ConnectedClient'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ip_address() {
    if (!this._ip_addressIntialized) {
      return undefined;
    }
    return this._wrapperFields.ip_address.data;
  }
  set ip_address(value) {
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.data = value;
  }
  get connection_time() {
    return this._wrapperFields.connection_time;
  }
  set connection_time(value) {
    if (value instanceof TimeWrapper) {
      this._wrapperFields.connection_time.copy(value);
    } else {
      this._wrapperFields.connection_time.copy(new TimeWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ConnectedClientRefStruct(refObject.toObject());
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.copyRefObject(this._refObject.ip_address);
    this._wrapperFields.connection_time.copyRefObject(this._refObject.connection_time);
  }
  copy(other) {
    this._refObject = new ConnectedClientRefStruct(other._refObject.toObject());
    this._ip_addressIntialized = true;
    this._wrapperFields.ip_address.copy(other._wrapperFields.ip_address);
    this._wrapperFields.connection_time.copy(other._wrapperFields.connection_time);
  }
  static get classType() {
    return ConnectedClientWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ip_address","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"connection_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Time","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosbridge_msgs","type":"ConnectedClient","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ConnectedClient"};
  }
  hasMember(name) {
    let memberNames = ["ip_address","connection_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ConnectedClientArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ConnectedClientWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ConnectedClientWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConnectedClientArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConnectedClientArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ConnectedClientRefArray(size);
    this._refObject = new ConnectedClientRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ConnectedClientWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ConnectedClientArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ConnectedClientArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ConnectedClientWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ConnectedClientWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ConnectedClientArrayWrapper;
  }
}
module.exports = ConnectedClientWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ip_address",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "connection_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Time",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosbridge_msgs",
    "type": "ConnectedClient",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ConnectedClient"
}
*/
