// rclnodejs/generated/rosbridge_msgs/rosbridge_msgs__msg__ConnectedClients.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ConnectedClientWrapper = require('../../generated/rosbridge_msgs/rosbridge_msgs__msg__ConnectedClient.js');
const ConnectedClientsRefStruct = StructType({
  clients: ConnectedClientWrapper.refObjectArrayType,
});
const ConnectedClientsRefArray = ArrayType(ConnectedClientsRefStruct);
const ConnectedClientsRefStructArray = StructType({
  data: ConnectedClientsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ConnectedClientsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ConnectedClientsRefStruct(other._refObject.toObject());
      this._wrapperFields.clients = ConnectedClientWrapper.createArray();
      this._wrapperFields.clients.copy(other._wrapperFields.clients);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ConnectedClientsRefStruct();
    this._wrapperFields.clients = ConnectedClientWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ConnectedClientsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ConnectedClientsArrayWrapper;
  }
  static get ArrayType() {
    return ConnectedClientsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ConnectedClientsRefStructArray
  }
  static get refObjectType() {
    return ConnectedClientsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.clients.freeze(own, checkConsistency);
    this._refObject.clients = this._wrapperFields.clients.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.clients.copyRefObject(refObject.clients);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.clients.size != 0) {
      ConnectedClientWrapper.ArrayType.freeArray(refObject.clients);
      if (ConnectedClientWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.clients, ConnectedClientWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ConnectedClientsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbridge_msgs', subFolder: 'msg', interfaceName: 'ConnectedClients'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get clients() {
    return this._wrapperFields.clients;
  }
  set clients(value) {
    this._wrapperFields.clients.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new ConnectedClientsRefStruct(refObject.toObject());
    this._wrapperFields.clients.copyRefObject(this._refObject.clients);
  }
  copy(other) {
    this._refObject = new ConnectedClientsRefStruct(other._refObject.toObject());
    this._wrapperFields.clients.copy(other._wrapperFields.clients);
  }
  static get classType() {
    return ConnectedClientsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"clients","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"rosbridge_msgs","type":"ConnectedClient","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"rosbridge_msgs","type":"ConnectedClients","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ConnectedClients"};
  }
  hasMember(name) {
    let memberNames = ["clients"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ConnectedClientsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ConnectedClientsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ConnectedClientsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConnectedClientsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ConnectedClientsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ConnectedClientsRefArray(size);
    this._refObject = new ConnectedClientsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ConnectedClientsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ConnectedClientsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ConnectedClientsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ConnectedClientsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ConnectedClientsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ConnectedClientsArrayWrapper;
  }
}
module.exports = ConnectedClientsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "clients",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "rosbridge_msgs",
        "type": "ConnectedClient",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosbridge_msgs",
    "type": "ConnectedClients",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ConnectedClients"
}
*/
