// rclnodejs/generated/turtlesim/turtlesim__msg__Pose.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseRefStruct = StructType({
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  theta: primitiveTypes.float32,
  linear_velocity: primitiveTypes.float32,
  angular_velocity: primitiveTypes.float32,
});
const PoseRefArray = ArrayType(PoseRefStruct);
const PoseRefStructArray = StructType({
  data: PoseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PoseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PoseRefStruct(other._refObject.toObject());
      this._xIntialized = true;
      this._yIntialized = true;
      this._thetaIntialized = true;
      this._linear_velocityIntialized = true;
      this._angular_velocityIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PoseRefStruct();
    this._xIntialized = false;
    this._yIntialized = false;
    this._thetaIntialized = false;
    this._linear_velocityIntialized = false;
    this._angular_velocityIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PoseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PoseArrayWrapper;
  }
  static get ArrayType() {
    return PoseArrayWrapper;
  }
  static get refObjectArrayType() {
    return PoseRefStructArray
  }
  static get refObjectType() {
    return PoseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._xIntialized) {
        throw new TypeError('Invalid argument: x in Pose');
      }
      if (!this._yIntialized) {
        throw new TypeError('Invalid argument: y in Pose');
      }
      if (!this._thetaIntialized) {
        throw new TypeError('Invalid argument: theta in Pose');
      }
      if (!this._linear_velocityIntialized) {
        throw new TypeError('Invalid argument: linear_velocity in Pose');
      }
      if (!this._angular_velocityIntialized) {
        throw new TypeError('Invalid argument: angular_velocity in Pose');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._thetaIntialized = true;
    this._refObject.theta = refObject.theta;
    this._linear_velocityIntialized = true;
    this._refObject.linear_velocity = refObject.linear_velocity;
    this._angular_velocityIntialized = true;
    this._refObject.angular_velocity = refObject.angular_velocity;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PoseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'msg', interfaceName: 'Pose'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get theta() {
    if (!this._thetaIntialized) {
      return undefined;
    }
    return this._refObject.theta;
  }
  set theta(value) {
    this._thetaIntialized = true;
    this._refObject.theta = value;
  }
  get linear_velocity() {
    if (!this._linear_velocityIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity;
  }
  set linear_velocity(value) {
    this._linear_velocityIntialized = true;
    this._refObject.linear_velocity = value;
  }
  get angular_velocity() {
    if (!this._angular_velocityIntialized) {
      return undefined;
    }
    return this._refObject.angular_velocity;
  }
  set angular_velocity(value) {
    this._angular_velocityIntialized = true;
    this._refObject.angular_velocity = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PoseRefStruct(refObject.toObject());
    this._xIntialized = true;
    this._yIntialized = true;
    this._thetaIntialized = true;
    this._linear_velocityIntialized = true;
    this._angular_velocityIntialized = true;
  }
  copy(other) {
    this._refObject = new PoseRefStruct(other._refObject.toObject());
    this._xIntialized = true;
    this._yIntialized = true;
    this._thetaIntialized = true;
    this._linear_velocityIntialized = true;
    this._angular_velocityIntialized = true;
  }
  static get classType() {
    return PoseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"theta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_velocity","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Pose"};
  }
  hasMember(name) {
    let memberNames = ["x","y","theta","linear_velocity","angular_velocity"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PoseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PoseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PoseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PoseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PoseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PoseRefArray(size);
    this._refObject = new PoseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PoseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PoseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PoseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PoseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PoseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PoseArrayWrapper;
  }
}
module.exports = PoseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "theta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_velocity",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "Pose",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Pose"
}
*/
