// rclnodejs/generated/turtlesim/turtlesim__action__RotateAbsolute_GetResult_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RotateAbsolute_ResultWrapper = require('../../generated/turtlesim/turtlesim__action__RotateAbsolute_Result.js');
const RotateAbsolute_GetResult_ResponseRefStruct = StructType({
  status: primitiveTypes.int8,
  result: RotateAbsolute_ResultWrapper.refObjectType,
});
const RotateAbsolute_GetResult_ResponseRefArray = ArrayType(RotateAbsolute_GetResult_ResponseRefStruct);
const RotateAbsolute_GetResult_ResponseRefStructArray = StructType({
  data: RotateAbsolute_GetResult_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RotateAbsolute_GetResult_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RotateAbsolute_GetResult_ResponseRefStruct(other._refObject.toObject());
      this._statusIntialized = true;
      this._wrapperFields.result =  new RotateAbsolute_ResultWrapper(other._wrapperFields.result);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RotateAbsolute_GetResult_ResponseRefStruct();
    this._statusIntialized = false;
    this._wrapperFields.result = new RotateAbsolute_ResultWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RotateAbsolute_GetResult_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RotateAbsolute_GetResult_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return RotateAbsolute_GetResult_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return RotateAbsolute_GetResult_ResponseRefStructArray
  }
  static get refObjectType() {
    return RotateAbsolute_GetResult_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._statusIntialized) {
        throw new TypeError('Invalid argument: status in RotateAbsolute_GetResult_Response');
      }
    }
    this._wrapperFields.result.freeze(own, checkConsistency);
    this._refObject.result = this._wrapperFields.result.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._statusIntialized = true;
    this._refObject.status = refObject.status;
    this._wrapperFields.result.copyRefObject(refObject.result);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    RotateAbsolute_ResultWrapper.freeStruct(refObject.result);
  }
  static destoryRawROS(msg) {
    RotateAbsolute_GetResult_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'action', interfaceName: 'RotateAbsolute_GetResult_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get status() {
    if (!this._statusIntialized) {
      return undefined;
    }
    return this._refObject.status;
  }
  set status(value) {
    this._statusIntialized = true;
    this._refObject.status = value;
  }
  get result() {
    return this._wrapperFields.result;
  }
  set result(value) {
    if (value instanceof RotateAbsolute_ResultWrapper) {
      this._wrapperFields.result.copy(value);
    } else {
      this._wrapperFields.result.copy(new RotateAbsolute_ResultWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new RotateAbsolute_GetResult_ResponseRefStruct(refObject.toObject());
    this._statusIntialized = true;
    this._wrapperFields.result.copyRefObject(this._refObject.result);
  }
  copy(other) {
    this._refObject = new RotateAbsolute_GetResult_ResponseRefStruct(other._refObject.toObject());
    this._statusIntialized = true;
    this._wrapperFields.result.copy(other._wrapperFields.result);
  }
  static get classType() {
    return RotateAbsolute_GetResult_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"status","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":null,"pkgName":"turtlesim","type":"RotateAbsolute_Result","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"RotateAbsolute_GetResult_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RotateAbsolute_GetResult_Response"};
  }
  hasMember(name) {
    let memberNames = ["status","result"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RotateAbsolute_GetResult_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RotateAbsolute_GetResult_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RotateAbsolute_GetResult_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_GetResult_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_GetResult_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RotateAbsolute_GetResult_ResponseRefArray(size);
    this._refObject = new RotateAbsolute_GetResult_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RotateAbsolute_GetResult_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RotateAbsolute_GetResult_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RotateAbsolute_GetResult_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RotateAbsolute_GetResult_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RotateAbsolute_GetResult_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RotateAbsolute_GetResult_ResponseArrayWrapper;
  }
}
module.exports = RotateAbsolute_GetResult_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "status",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": null,
        "pkgName": "turtlesim",
        "type": "RotateAbsolute_Result",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "RotateAbsolute_GetResult_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RotateAbsolute_GetResult_Response"
}
*/
