// rclnodejs/generated/turtlesim/turtlesim__srv__TeleportAbsolute_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TeleportAbsolute_ResponseRefStruct = StructType({
  _dummy: primitiveTypes.bool,
});
const TeleportAbsolute_ResponseRefArray = ArrayType(TeleportAbsolute_ResponseRefStruct);
const TeleportAbsolute_ResponseRefStructArray = StructType({
  data: TeleportAbsolute_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TeleportAbsolute_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TeleportAbsolute_ResponseRefStruct(other._refObject.toObject());
      this.__dummyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TeleportAbsolute_ResponseRefStruct();
    this.__dummyIntialized = true;
  }
  static createFromRefObject(refObject) {
    let self = new TeleportAbsolute_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TeleportAbsolute_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return TeleportAbsolute_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return TeleportAbsolute_ResponseRefStructArray
  }
  static get refObjectType() {
    return TeleportAbsolute_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this.__dummyIntialized) {
        throw new TypeError('Invalid argument: _dummy in TeleportAbsolute_Response');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this.__dummyIntialized = true;
    this._refObject._dummy = refObject._dummy;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TeleportAbsolute_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'TeleportAbsolute_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get _dummy() {
    if (!this.__dummyIntialized) {
      return undefined;
    }
    return this._refObject._dummy;
  }
  set _dummy(value) {
    this.__dummyIntialized = true;
    this._refObject._dummy = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TeleportAbsolute_ResponseRefStruct(refObject.toObject());
    this.__dummyIntialized = true;
  }
  copy(other) {
    this._refObject = new TeleportAbsolute_ResponseRefStruct(other._refObject.toObject());
    this.__dummyIntialized = true;
  }
  static get classType() {
    return TeleportAbsolute_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[],"baseType":{"pkgName":"turtlesim","type":"TeleportAbsolute_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TeleportAbsolute_Response"};
  }
  hasMember(name) {
    let memberNames = ["_dummy"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TeleportAbsolute_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TeleportAbsolute_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TeleportAbsolute_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportAbsolute_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportAbsolute_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TeleportAbsolute_ResponseRefArray(size);
    this._refObject = new TeleportAbsolute_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TeleportAbsolute_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TeleportAbsolute_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TeleportAbsolute_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TeleportAbsolute_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TeleportAbsolute_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TeleportAbsolute_ResponseArrayWrapper;
  }
}
module.exports = TeleportAbsolute_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "TeleportAbsolute_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TeleportAbsolute_Response"
}
*/
