// rclnodejs/generated/turtlesim/turtlesim__action__RotateAbsolute.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class turtlesim__action__RotateAbsolute {
  static get Goal() {
    return require('./turtlesim__action__RotateAbsolute_Goal.js');
  }
  static get Result() {
    return require('./turtlesim__action__RotateAbsolute_Result.js');
  }
  static get Feedback() {
    return require('./turtlesim__action__RotateAbsolute_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./turtlesim__action__RotateAbsolute_SendGoal.js');
      },
      get GetResultService() {
        return require('./turtlesim__action__RotateAbsolute_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./turtlesim__action__RotateAbsolute_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'action', interfaceName: 'RotateAbsolute'};
  }
}
module.exports = turtlesim__action__RotateAbsolute;
