// rclnodejs/generated/turtlesim/turtlesim__srv__SetPen_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const SetPen_RequestRefStruct = StructType({
  r: primitiveTypes.uint8,
  g: primitiveTypes.uint8,
  b: primitiveTypes.uint8,
  width: primitiveTypes.uint8,
  off: primitiveTypes.uint8,
});
const SetPen_RequestRefArray = ArrayType(SetPen_RequestRefStruct);
const SetPen_RequestRefStructArray = StructType({
  data: SetPen_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetPen_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetPen_RequestRefStruct(other._refObject.toObject());
      this._rIntialized = true;
      this._gIntialized = true;
      this._bIntialized = true;
      this._widthIntialized = true;
      this._offIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetPen_RequestRefStruct();
    this._rIntialized = false;
    this._gIntialized = false;
    this._bIntialized = false;
    this._widthIntialized = false;
    this._offIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetPen_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetPen_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetPen_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetPen_RequestRefStructArray
  }
  static get refObjectType() {
    return SetPen_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._rIntialized) {
        throw new TypeError('Invalid argument: r in SetPen_Request');
      }
      if (!this._gIntialized) {
        throw new TypeError('Invalid argument: g in SetPen_Request');
      }
      if (!this._bIntialized) {
        throw new TypeError('Invalid argument: b in SetPen_Request');
      }
      if (!this._widthIntialized) {
        throw new TypeError('Invalid argument: width in SetPen_Request');
      }
      if (!this._offIntialized) {
        throw new TypeError('Invalid argument: off in SetPen_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._rIntialized = true;
    this._refObject.r = refObject.r;
    this._gIntialized = true;
    this._refObject.g = refObject.g;
    this._bIntialized = true;
    this._refObject.b = refObject.b;
    this._widthIntialized = true;
    this._refObject.width = refObject.width;
    this._offIntialized = true;
    this._refObject.off = refObject.off;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    SetPen_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'SetPen_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get r() {
    if (!this._rIntialized) {
      return undefined;
    }
    return this._refObject.r;
  }
  set r(value) {
    this._rIntialized = true;
    this._refObject.r = value;
  }
  get g() {
    if (!this._gIntialized) {
      return undefined;
    }
    return this._refObject.g;
  }
  set g(value) {
    this._gIntialized = true;
    this._refObject.g = value;
  }
  get b() {
    if (!this._bIntialized) {
      return undefined;
    }
    return this._refObject.b;
  }
  set b(value) {
    this._bIntialized = true;
    this._refObject.b = value;
  }
  get width() {
    if (!this._widthIntialized) {
      return undefined;
    }
    return this._refObject.width;
  }
  set width(value) {
    this._widthIntialized = true;
    this._refObject.width = value;
  }
  get off() {
    if (!this._offIntialized) {
      return undefined;
    }
    return this._refObject.off;
  }
  set off(value) {
    this._offIntialized = true;
    this._refObject.off = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetPen_RequestRefStruct(refObject.toObject());
    this._rIntialized = true;
    this._gIntialized = true;
    this._bIntialized = true;
    this._widthIntialized = true;
    this._offIntialized = true;
  }
  copy(other) {
    this._refObject = new SetPen_RequestRefStruct(other._refObject.toObject());
    this._rIntialized = true;
    this._gIntialized = true;
    this._bIntialized = true;
    this._widthIntialized = true;
    this._offIntialized = true;
  }
  static get classType() {
    return SetPen_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"r","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"g","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"b","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"width","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"off","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"SetPen_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetPen_Request"};
  }
  hasMember(name) {
    let memberNames = ["r","g","b","width","off"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetPen_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetPen_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetPen_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPen_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetPen_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetPen_RequestRefArray(size);
    this._refObject = new SetPen_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetPen_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetPen_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetPen_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetPen_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetPen_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetPen_RequestArrayWrapper;
  }
}
module.exports = SetPen_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "r",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "g",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "b",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "width",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "off",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "SetPen_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetPen_Request"
}
*/
