// rclnodejs/generated/turtlesim/turtlesim__srv__TeleportRelative_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TeleportRelative_RequestRefStruct = StructType({
  linear: primitiveTypes.float32,
  angular: primitiveTypes.float32,
});
const TeleportRelative_RequestRefArray = ArrayType(TeleportRelative_RequestRefStruct);
const TeleportRelative_RequestRefStructArray = StructType({
  data: TeleportRelative_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class TeleportRelative_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new TeleportRelative_RequestRefStruct(other._refObject.toObject());
      this._linearIntialized = true;
      this._angularIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new TeleportRelative_RequestRefStruct();
    this._linearIntialized = false;
    this._angularIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new TeleportRelative_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new TeleportRelative_RequestArrayWrapper;
  }
  static get ArrayType() {
    return TeleportRelative_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return TeleportRelative_RequestRefStructArray
  }
  static get refObjectType() {
    return TeleportRelative_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._linearIntialized) {
        throw new TypeError('Invalid argument: linear in TeleportRelative_Request');
      }
      if (!this._angularIntialized) {
        throw new TypeError('Invalid argument: angular in TeleportRelative_Request');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._linearIntialized = true;
    this._refObject.linear = refObject.linear;
    this._angularIntialized = true;
    this._refObject.angular = refObject.angular;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    TeleportRelative_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'srv', interfaceName: 'TeleportRelative_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get linear() {
    if (!this._linearIntialized) {
      return undefined;
    }
    return this._refObject.linear;
  }
  set linear(value) {
    this._linearIntialized = true;
    this._refObject.linear = value;
  }
  get angular() {
    if (!this._angularIntialized) {
      return undefined;
    }
    return this._refObject.angular;
  }
  set angular(value) {
    this._angularIntialized = true;
    this._refObject.angular = value;
  }
  copyRefObject(refObject) {
    this._refObject = new TeleportRelative_RequestRefStruct(refObject.toObject());
    this._linearIntialized = true;
    this._angularIntialized = true;
  }
  copy(other) {
    this._refObject = new TeleportRelative_RequestRefStruct(other._refObject.toObject());
    this._linearIntialized = true;
    this._angularIntialized = true;
  }
  static get classType() {
    return TeleportRelative_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"linear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"TeleportRelative_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"TeleportRelative_Request"};
  }
  hasMember(name) {
    let memberNames = ["linear","angular"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class TeleportRelative_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof TeleportRelative_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new TeleportRelative_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportRelative_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to TeleportRelative_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new TeleportRelative_RequestRefArray(size);
    this._refObject = new TeleportRelative_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new TeleportRelative_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof TeleportRelative_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "TeleportRelative_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      TeleportRelative_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return TeleportRelative_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return TeleportRelative_RequestArrayWrapper;
  }
}
module.exports = TeleportRelative_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "linear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "TeleportRelative_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "TeleportRelative_Request"
}
*/
