// rclnodejs/generated/turtlesim/turtlesim__action__RotateAbsolute_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const RotateAbsolute_ResultRefStruct = StructType({
  delta: primitiveTypes.float32,
});
const RotateAbsolute_ResultRefArray = ArrayType(RotateAbsolute_ResultRefStruct);
const RotateAbsolute_ResultRefStructArray = StructType({
  data: RotateAbsolute_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RotateAbsolute_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RotateAbsolute_ResultRefStruct(other._refObject.toObject());
      this._deltaIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RotateAbsolute_ResultRefStruct();
    this._deltaIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RotateAbsolute_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RotateAbsolute_ResultArrayWrapper;
  }
  static get ArrayType() {
    return RotateAbsolute_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return RotateAbsolute_ResultRefStructArray
  }
  static get refObjectType() {
    return RotateAbsolute_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._deltaIntialized) {
        throw new TypeError('Invalid argument: delta in RotateAbsolute_Result');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._deltaIntialized = true;
    this._refObject.delta = refObject.delta;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    RotateAbsolute_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'turtlesim', subFolder: 'action', interfaceName: 'RotateAbsolute_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get delta() {
    if (!this._deltaIntialized) {
      return undefined;
    }
    return this._refObject.delta;
  }
  set delta(value) {
    this._deltaIntialized = true;
    this._refObject.delta = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RotateAbsolute_ResultRefStruct(refObject.toObject());
    this._deltaIntialized = true;
  }
  copy(other) {
    this._refObject = new RotateAbsolute_ResultRefStruct(other._refObject.toObject());
    this._deltaIntialized = true;
  }
  static get classType() {
    return RotateAbsolute_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"delta","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"turtlesim","type":"RotateAbsolute_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RotateAbsolute_Result"};
  }
  hasMember(name) {
    let memberNames = ["delta"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RotateAbsolute_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RotateAbsolute_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RotateAbsolute_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RotateAbsolute_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RotateAbsolute_ResultRefArray(size);
    this._refObject = new RotateAbsolute_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RotateAbsolute_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RotateAbsolute_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RotateAbsolute_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RotateAbsolute_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RotateAbsolute_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RotateAbsolute_ResultArrayWrapper;
  }
}
module.exports = RotateAbsolute_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "delta",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "turtlesim",
    "type": "RotateAbsolute_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RotateAbsolute_Result"
}
*/
