// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServiceType_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServiceType_RequestRefStruct = StructType({
  service: primitiveTypes.string,
});
const ServiceType_RequestRefArray = ArrayType(ServiceType_RequestRefStruct);
const ServiceType_RequestRefStructArray = StructType({
  data: ServiceType_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ServiceType_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServiceType_RequestRefStruct(other._refObject.toObject());
      this._serviceIntialized = true;
      this._wrapperFields.service =  new StringWrapper(other._wrapperFields.service);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ServiceType_RequestRefStruct();
    this._serviceIntialized = false;
    this._wrapperFields.service = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ServiceType_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServiceType_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ServiceType_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServiceType_RequestRefStructArray
  }
  static get refObjectType() {
    return ServiceType_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._serviceIntialized) {
        throw new TypeError('Invalid argument: service in ServiceType_Request');
      }
    }
    if (own) {
      this._wrapperFields.service.freeze(own, checkConsistency);
    }
    this._refObject.service = this._wrapperFields.service.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._serviceIntialized = true;
    this._wrapperFields.service.data = refObject.service.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.service);
  }
  static destoryRawROS(msg) {
    ServiceType_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServiceType_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get service() {
    if (!this._serviceIntialized) {
      return undefined;
    }
    return this._wrapperFields.service.data;
  }
  set service(value) {
    this._serviceIntialized = true;
    this._wrapperFields.service.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServiceType_RequestRefStruct(refObject.toObject());
    this._serviceIntialized = true;
    this._wrapperFields.service.copyRefObject(this._refObject.service);
  }
  copy(other) {
    this._refObject = new ServiceType_RequestRefStruct(other._refObject.toObject());
    this._serviceIntialized = true;
    this._wrapperFields.service.copy(other._wrapperFields.service);
  }
  static get classType() {
    return ServiceType_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"service","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServiceType_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServiceType_Request"};
  }
  hasMember(name) {
    let memberNames = ["service"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ServiceType_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServiceType_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServiceType_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceType_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceType_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ServiceType_RequestRefArray(size);
    this._refObject = new ServiceType_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServiceType_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServiceType_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServiceType_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServiceType_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServiceType_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServiceType_RequestArrayWrapper;
  }
}
module.exports = ServiceType_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "service",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "ServiceType_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ServiceType_Request"
}
*/
