// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServiceProviders_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServiceProviders_ResponseRefStruct = StructType({
  providers: StringWrapper.refObjectArrayType,
});
const ServiceProviders_ResponseRefArray = ArrayType(ServiceProviders_ResponseRefStruct);
const ServiceProviders_ResponseRefStructArray = StructType({
  data: ServiceProviders_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ServiceProviders_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._providersArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServiceProviders_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.providers = StringWrapper.createArray();
      this._wrapperFields.providers.copy(other._wrapperFields.providers);
      this.providers = other.providers;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ServiceProviders_ResponseRefStruct();
    this._wrapperFields.providers = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ServiceProviders_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServiceProviders_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ServiceProviders_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServiceProviders_ResponseRefStructArray
  }
  static get refObjectType() {
    return ServiceProviders_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.providers.fill(this._providersArray);
      this._wrapperFields.providers.freeze(own, checkConsistency);
      this._refObject.providers = this._wrapperFields.providers.refObject;
    } else {
      this._wrapperFields.providers.fill([]);
      this._wrapperFields.providers.freeze(own, checkConsistency);
      this._refObject.providers = this._wrapperFields.providers.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.providers.data.length = refObject.providers.size;
    for (let index = 0; index < refObject.providers.size; index++) {
      this._providersArray[index] = refObject.providers.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.providers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.providers);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.providers, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ServiceProviders_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServiceProviders_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get providers() {
    return this._providersArray;
  }
  set providers(value) {
    this._providersArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServiceProviders_ResponseRefStruct(refObject.toObject());
    refObject.providers.data.length = refObject.providers.size;
    for (let index = 0; index < refObject.providers.size; index++) {
      this._providersArray[index] = refObject.providers.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ServiceProviders_ResponseRefStruct(other._refObject.toObject());
    this._providersArray = other._providersArray.slice();
  }
  static get classType() {
    return ServiceProviders_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"providers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServiceProviders_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServiceProviders_Response"};
  }
  hasMember(name) {
    let memberNames = ["providers"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ServiceProviders_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServiceProviders_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServiceProviders_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceProviders_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServiceProviders_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ServiceProviders_ResponseRefArray(size);
    this._refObject = new ServiceProviders_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServiceProviders_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServiceProviders_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServiceProviders_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServiceProviders_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServiceProviders_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServiceProviders_ResponseArrayWrapper;
  }
}
module.exports = ServiceProviders_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "providers",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "ServiceProviders_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ServiceProviders_Response"
}
*/
