// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__GetParam_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetParam_ResponseRefStruct = StructType({
  value: primitiveTypes.string,
});
const GetParam_ResponseRefArray = ArrayType(GetParam_ResponseRefStruct);
const GetParam_ResponseRefStructArray = StructType({
  data: GetParam_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetParam_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetParam_ResponseRefStruct(other._refObject.toObject());
      this._valueIntialized = true;
      this._wrapperFields.value =  new StringWrapper(other._wrapperFields.value);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetParam_ResponseRefStruct();
    this._valueIntialized = false;
    this._wrapperFields.value = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetParam_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetParam_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetParam_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetParam_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetParam_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._valueIntialized) {
        throw new TypeError('Invalid argument: value in GetParam_Response');
      }
    }
    if (own) {
      this._wrapperFields.value.freeze(own, checkConsistency);
    }
    this._refObject.value = this._wrapperFields.value.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._valueIntialized = true;
    this._wrapperFields.value.data = refObject.value.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.value);
  }
  static destoryRawROS(msg) {
    GetParam_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'GetParam_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._wrapperFields.value.data;
  }
  set value(value) {
    this._valueIntialized = true;
    this._wrapperFields.value.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetParam_ResponseRefStruct(refObject.toObject());
    this._valueIntialized = true;
    this._wrapperFields.value.copyRefObject(this._refObject.value);
  }
  copy(other) {
    this._refObject = new GetParam_ResponseRefStruct(other._refObject.toObject());
    this._valueIntialized = true;
    this._wrapperFields.value.copy(other._wrapperFields.value);
  }
  static get classType() {
    return GetParam_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"GetParam_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetParam_Response"};
  }
  hasMember(name) {
    let memberNames = ["value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetParam_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetParam_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetParam_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParam_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParam_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetParam_ResponseRefArray(size);
    this._refObject = new GetParam_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetParam_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetParam_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetParam_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetParam_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetParam_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetParam_ResponseArrayWrapper;
  }
}
module.exports = GetParam_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "GetParam_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetParam_Response"
}
*/
