// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__GetParamNames_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const GetParamNames_ResponseRefStruct = StructType({
  names: StringWrapper.refObjectArrayType,
});
const GetParamNames_ResponseRefArray = ArrayType(GetParamNames_ResponseRefStruct);
const GetParamNames_ResponseRefStructArray = StructType({
  data: GetParamNames_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetParamNames_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._namesArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetParamNames_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.names = StringWrapper.createArray();
      this._wrapperFields.names.copy(other._wrapperFields.names);
      this.names = other.names;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetParamNames_ResponseRefStruct();
    this._wrapperFields.names = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetParamNames_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetParamNames_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetParamNames_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetParamNames_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetParamNames_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.names.fill(this._namesArray);
      this._wrapperFields.names.freeze(own, checkConsistency);
      this._refObject.names = this._wrapperFields.names.refObject;
    } else {
      this._wrapperFields.names.fill([]);
      this._wrapperFields.names.freeze(own, checkConsistency);
      this._refObject.names = this._wrapperFields.names.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.names.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.names);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.names, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetParamNames_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'GetParamNames_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get names() {
    return this._namesArray;
  }
  set names(value) {
    this._namesArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetParamNames_ResponseRefStruct(refObject.toObject());
    refObject.names.data.length = refObject.names.size;
    for (let index = 0; index < refObject.names.size; index++) {
      this._namesArray[index] = refObject.names.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new GetParamNames_ResponseRefStruct(other._refObject.toObject());
    this._namesArray = other._namesArray.slice();
  }
  static get classType() {
    return GetParamNames_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"names","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"GetParamNames_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetParamNames_Response"};
  }
  hasMember(name) {
    let memberNames = ["names"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetParamNames_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetParamNames_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetParamNames_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParamNames_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetParamNames_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetParamNames_ResponseRefArray(size);
    this._refObject = new GetParamNames_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetParamNames_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetParamNames_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetParamNames_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetParamNames_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetParamNames_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetParamNames_ResponseArrayWrapper;
  }
}
module.exports = GetParamNames_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "names",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "GetParamNames_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetParamNames_Response"
}
*/
