// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__ServicesForType_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ServicesForType_RequestRefStruct = StructType({
  type: primitiveTypes.string,
});
const ServicesForType_RequestRefArray = ArrayType(ServicesForType_RequestRefStruct);
const ServicesForType_RequestRefStructArray = StructType({
  data: ServicesForType_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ServicesForType_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ServicesForType_RequestRefStruct(other._refObject.toObject());
      this._typeIntialized = true;
      this._wrapperFields.type =  new StringWrapper(other._wrapperFields.type);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ServicesForType_RequestRefStruct();
    this._typeIntialized = false;
    this._wrapperFields.type = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ServicesForType_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ServicesForType_RequestArrayWrapper;
  }
  static get ArrayType() {
    return ServicesForType_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return ServicesForType_RequestRefStructArray
  }
  static get refObjectType() {
    return ServicesForType_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in ServicesForType_Request');
      }
    }
    if (own) {
      this._wrapperFields.type.freeze(own, checkConsistency);
    }
    this._refObject.type = this._wrapperFields.type.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._typeIntialized = true;
    this._wrapperFields.type.data = refObject.type.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.type);
  }
  static destoryRawROS(msg) {
    ServicesForType_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'ServicesForType_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._wrapperFields.type.data;
  }
  set type(value) {
    this._typeIntialized = true;
    this._wrapperFields.type.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ServicesForType_RequestRefStruct(refObject.toObject());
    this._typeIntialized = true;
    this._wrapperFields.type.copyRefObject(this._refObject.type);
  }
  copy(other) {
    this._refObject = new ServicesForType_RequestRefStruct(other._refObject.toObject());
    this._typeIntialized = true;
    this._wrapperFields.type.copy(other._wrapperFields.type);
  }
  static get classType() {
    return ServicesForType_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"ServicesForType_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ServicesForType_Request"};
  }
  hasMember(name) {
    let memberNames = ["type"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ServicesForType_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ServicesForType_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ServicesForType_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServicesForType_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ServicesForType_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ServicesForType_RequestRefArray(size);
    this._refObject = new ServicesForType_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ServicesForType_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ServicesForType_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ServicesForType_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ServicesForType_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ServicesForType_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ServicesForType_RequestArrayWrapper;
  }
}
module.exports = ServicesForType_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "ServicesForType_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ServicesForType_Request"
}
*/
