// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__Subscribers_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const Subscribers_ResponseRefStruct = StructType({
  subscribers: StringWrapper.refObjectArrayType,
});
const Subscribers_ResponseRefArray = ArrayType(Subscribers_ResponseRefStruct);
const Subscribers_ResponseRefStructArray = StructType({
  data: Subscribers_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Subscribers_ResponseWrapper {
  constructor(other) {
    this._wrapperFields = {};
    this._subscribersArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Subscribers_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.subscribers = StringWrapper.createArray();
      this._wrapperFields.subscribers.copy(other._wrapperFields.subscribers);
      this.subscribers = other.subscribers;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Subscribers_ResponseRefStruct();
    this._wrapperFields.subscribers = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Subscribers_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Subscribers_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return Subscribers_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return Subscribers_ResponseRefStructArray
  }
  static get refObjectType() {
    return Subscribers_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    if (!own) {
      this._wrapperFields.subscribers.fill(this._subscribersArray);
      this._wrapperFields.subscribers.freeze(own, checkConsistency);
      this._refObject.subscribers = this._wrapperFields.subscribers.refObject;
    } else {
      this._wrapperFields.subscribers.fill([]);
      this._wrapperFields.subscribers.freeze(own, checkConsistency);
      this._refObject.subscribers = this._wrapperFields.subscribers.refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.subscribers.data.length = refObject.subscribers.size;
    for (let index = 0; index < refObject.subscribers.size; index++) {
      this._subscribersArray[index] = refObject.subscribers.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.subscribers.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.subscribers);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.subscribers, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Subscribers_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'Subscribers_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get subscribers() {
    return this._subscribersArray;
  }
  set subscribers(value) {
    this._subscribersArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new Subscribers_ResponseRefStruct(refObject.toObject());
    refObject.subscribers.data.length = refObject.subscribers.size;
    for (let index = 0; index < refObject.subscribers.size; index++) {
      this._subscribersArray[index] = refObject.subscribers.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new Subscribers_ResponseRefStruct(other._refObject.toObject());
    this._subscribersArray = other._subscribersArray.slice();
  }
  static get classType() {
    return Subscribers_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"subscribers","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"Subscribers_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Subscribers_Response"};
  }
  hasMember(name) {
    let memberNames = ["subscribers"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Subscribers_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Subscribers_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Subscribers_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Subscribers_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Subscribers_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Subscribers_ResponseRefArray(size);
    this._refObject = new Subscribers_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Subscribers_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Subscribers_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Subscribers_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Subscribers_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Subscribers_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Subscribers_ResponseArrayWrapper;
  }
}
module.exports = Subscribers_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "subscribers",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "Subscribers_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Subscribers_Response"
}
*/
