// rclnodejs/generated/rosapi_msgs/rosapi_msgs__srv__NodeDetails_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('ref-napi');
const StructType = require('ref-struct-di')(ref);
const ArrayType = require('ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const NodeDetails_RequestRefStruct = StructType({
  node: primitiveTypes.string,
});
const NodeDetails_RequestRefArray = ArrayType(NodeDetails_RequestRefStruct);
const NodeDetails_RequestRefStructArray = StructType({
  data: NodeDetails_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NodeDetails_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NodeDetails_RequestRefStruct(other._refObject.toObject());
      this._nodeIntialized = true;
      this._wrapperFields.node =  new StringWrapper(other._wrapperFields.node);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NodeDetails_RequestRefStruct();
    this._nodeIntialized = false;
    this._wrapperFields.node = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new NodeDetails_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NodeDetails_RequestArrayWrapper;
  }
  static get ArrayType() {
    return NodeDetails_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return NodeDetails_RequestRefStructArray
  }
  static get refObjectType() {
    return NodeDetails_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._nodeIntialized) {
        throw new TypeError('Invalid argument: node in NodeDetails_Request');
      }
    }
    if (own) {
      this._wrapperFields.node.freeze(own, checkConsistency);
    }
    this._refObject.node = this._wrapperFields.node.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._nodeIntialized = true;
    this._wrapperFields.node.data = refObject.node.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.node);
  }
  static destoryRawROS(msg) {
    NodeDetails_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosapi_msgs', subFolder: 'srv', interfaceName: 'NodeDetails_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node() {
    if (!this._nodeIntialized) {
      return undefined;
    }
    return this._wrapperFields.node.data;
  }
  set node(value) {
    this._nodeIntialized = true;
    this._wrapperFields.node.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NodeDetails_RequestRefStruct(refObject.toObject());
    this._nodeIntialized = true;
    this._wrapperFields.node.copyRefObject(this._refObject.node);
  }
  copy(other) {
    this._refObject = new NodeDetails_RequestRefStruct(other._refObject.toObject());
    this._nodeIntialized = true;
    this._wrapperFields.node.copy(other._wrapperFields.node);
  }
  static get classType() {
    return NodeDetails_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosapi_msgs","type":"NodeDetails_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NodeDetails_Request"};
  }
  hasMember(name) {
    let memberNames = ["node"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NodeDetails_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NodeDetails_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NodeDetails_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeDetails_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NodeDetails_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NodeDetails_RequestRefArray(size);
    this._refObject = new NodeDetails_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NodeDetails_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NodeDetails_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NodeDetails_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NodeDetails_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NodeDetails_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NodeDetails_RequestArrayWrapper;
  }
}
module.exports = NodeDetails_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosapi_msgs",
    "type": "NodeDetails_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NodeDetails_Request"
}
*/
